// @mathjax/src/cjs/a11y/explorer/Highlighter.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ATTR = void 0;
exports.getHighlighter = getHighlighter;
var Region_js_1 = require("./Region.js");
var DEFAULT_BACKGROUND = { color: 'blue', alpha: 0.2 };
var DEFAULT_FOREGROUND = { color: 'black', alpha: 1 };
exports.ATTR = {
    ENCLOSED: 'data-sre-enclosed',
    BBOX: 'data-sre-highlighter-bbox',
    ADDED: 'data-sre-highlighter-added',
};
var AbstractHighlighter = (function () {
    function AbstractHighlighter(priority) {
        this.currentHighlights = [];
        this.priority = priority;
        this.ATTR = 'data-sre-highlight-' + priority;
    }
    AbstractHighlighter.prototype.highlight = function (nodes) {
        var e_1, _a;
        this.currentHighlights.push(nodes);
        try {
            for (var nodes_1 = __values(nodes), nodes_1_1 = nodes_1.next(); !nodes_1_1.done; nodes_1_1 = nodes_1.next()) {
                var node = nodes_1_1.value;
                this.highlightNode(node);
                this.setHighlighted(node);
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (nodes_1_1 && !nodes_1_1.done && (_a = nodes_1.return)) _a.call(nodes_1);
            }
            finally { if (e_1) throw e_1.error; }
        }
    };
    AbstractHighlighter.prototype.highlightAll = function (node) {
        var e_2, _a;
        var mactions = this.getMactionNodes(node);
        try {
            for (var mactions_1 = __values(mactions), mactions_1_1 = mactions_1.next(); !mactions_1_1.done; mactions_1_1 = mactions_1.next()) {
                var maction = mactions_1_1.value;
                this.highlight([maction]);
            }
        }
        catch (e_2_1) { e_2 = { error: e_2_1 }; }
        finally {
            try {
                if (mactions_1_1 && !mactions_1_1.done && (_a = mactions_1.return)) _a.call(mactions_1);
            }
            finally { if (e_2) throw e_2.error; }
        }
    };
    AbstractHighlighter.prototype.unhighlight = function () {
        var _this = this;
        var nodes = this.currentHighlights.pop();
        if (!nodes) {
            return;
        }
        nodes.forEach(function (node) {
            if (_this.isHighlighted(node)) {
                _this.unhighlightNode(node);
                _this.unsetHighlighted(node);
            }
        });
    };
    AbstractHighlighter.prototype.unhighlightAll = function () {
        while (this.currentHighlights.length > 0) {
            this.unhighlight();
        }
    };
    AbstractHighlighter.prototype.encloseNodes = function (parts, node) {
        var e_3, _a, e_4, _b, e_5, _c;
        if (parts.length === 1) {
            return parts;
        }
        var CLASS = this.constructor;
        var selector = CLASS.lineSelector;
        var lineno = CLASS.lineAttr;
        var lines = new Map();
        try {
            for (var parts_1 = __values(parts), parts_1_1 = parts_1.next(); !parts_1_1.done; parts_1_1 = parts_1.next()) {
                var part = parts_1_1.value;
                var line = part.closest(selector);
                var n = line ? line.getAttribute(lineno) : '';
                if (!lines.has(n)) {
                    lines.set(n, []);
                }
                lines.get(n).push(part);
            }
        }
        catch (e_3_1) { e_3 = { error: e_3_1 }; }
        finally {
            try {
                if (parts_1_1 && !parts_1_1.done && (_a = parts_1.return)) _a.call(parts_1);
            }
            finally { if (e_3) throw e_3.error; }
        }
        try {
            for (var _d = __values(lines.values()), _e = _d.next(); !_e.done; _e = _d.next()) {
                var list = _e.value;
                if (list.length > 1) {
                    var _f = __read([Infinity, Infinity, -Infinity, -Infinity], 4), L = _f[0], T = _f[1], R = _f[2], B = _f[3];
                    try {
                        for (var list_1 = (e_5 = void 0, __values(list)), list_1_1 = list_1.next(); !list_1_1.done; list_1_1 = list_1.next()) {
                            var part = list_1_1.value;
                            part.setAttribute(exports.ATTR.ENCLOSED, 'true');
                            var _g = part.getBoundingClientRect(), left = _g.left, top_1 = _g.top, right = _g.right, bottom = _g.bottom;
                            if (top_1 === bottom && left === right)
                                continue;
                            if (left < L)
                                L = left;
                            if (top_1 < T)
                                T = top_1;
                            if (bottom > B)
                                B = bottom;
                            if (right > R)
                                R = right;
                        }
                    }
                    catch (e_5_1) { e_5 = { error: e_5_1 }; }
                    finally {
                        try {
                            if (list_1_1 && !list_1_1.done && (_c = list_1.return)) _c.call(list_1);
                        }
                        finally { if (e_5) throw e_5.error; }
                    }
                    var enclosure = this.createEnclosure(L, B, R - L, B - T, node, list[0]);
                    parts.push(enclosure);
                }
            }
        }
        catch (e_4_1) { e_4 = { error: e_4_1 }; }
        finally {
            try {
                if (_e && !_e.done && (_b = _d.return)) _b.call(_d);
            }
            finally { if (e_4) throw e_4.error; }
        }
        return parts;
    };
    AbstractHighlighter.prototype.setColorCSS = function (type, color, def) {
        var _a, _b;
        var name = (_a = color.color) !== null && _a !== void 0 ? _a : def.color;
        var alpha = (_b = color.alpha) !== null && _b !== void 0 ? _b : def.alpha;
        Region_js_1.LiveRegion.setColor(type, this.priority, name, alpha);
    };
    AbstractHighlighter.prototype.setColor = function (background, foreground) {
        this.setColorCSS('fg', foreground, DEFAULT_FOREGROUND);
        this.setColorCSS('bg', background, DEFAULT_BACKGROUND);
    };
    AbstractHighlighter.prototype.isHighlighted = function (node) {
        return node.hasAttribute(this.ATTR);
    };
    AbstractHighlighter.prototype.setHighlighted = function (node) {
        node.setAttribute(this.ATTR, 'true');
    };
    AbstractHighlighter.prototype.unsetHighlighted = function (node) {
        node.removeAttribute(this.ATTR);
        node.removeAttribute(exports.ATTR.ENCLOSED);
    };
    return AbstractHighlighter;
}());
var SvgHighlighter = (function (_super) {
    __extends(SvgHighlighter, _super);
    function SvgHighlighter() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    SvgHighlighter.prototype.highlightNode = function (node) {
        if (this.isHighlighted(node) ||
            node.tagName === 'svg' ||
            node.tagName === 'MJX-CONTAINER' ||
            node.hasAttribute(exports.ATTR.BBOX) ||
            node.hasAttribute(exports.ATTR.ENCLOSED)) {
            return;
        }
        var _a = node.getBBox(), x = _a.x, y = _a.y, width = _a.width, height = _a.height;
        var rect = this.createRect(x, y, width, height, node.getAttribute('transform'));
        this.setHighlighted(rect);
        node.parentNode.insertBefore(rect, node);
    };
    SvgHighlighter.prototype.unhighlightNode = function (node) {
        if (node.hasAttribute(exports.ATTR.BBOX)) {
            node.remove();
            return;
        }
        var previous = node.previousSibling;
        if (previous === null || previous === void 0 ? void 0 : previous.hasAttribute(exports.ATTR.ADDED)) {
            previous.remove();
        }
    };
    SvgHighlighter.prototype.createEnclosure = function (x, y, w, h, _node, part) {
        var _a = __read(this.screen2svg(x, y, part), 2), x1 = _a[0], y1 = _a[1];
        var _b = __read(this.screen2svg(x + w, y - h, part), 2), x2 = _b[0], y2 = _b[1];
        var rect = this.createRect(x1, y1, x2 - x1, y2 - y1, part.getAttribute('transform'));
        rect.setAttribute(exports.ATTR.BBOX, 'true');
        part.parentNode.insertBefore(rect, part);
        return rect;
    };
    SvgHighlighter.prototype.screen2svg = function (x, y, part) {
        var node = part;
        var P = DOMPoint.fromPoint({ x: x, y: y }).matrixTransform(node.getScreenCTM().inverse());
        return [P.x, P.y];
    };
    SvgHighlighter.prototype.createRect = function (x, y, w, h, transform) {
        var padding = 40;
        var rect = document.createElementNS('http://www.w3.org/2000/svg', 'rect');
        rect.setAttribute(exports.ATTR.ADDED, 'true');
        rect.setAttribute('x', String(x - padding));
        rect.setAttribute('y', String(y - padding));
        rect.setAttribute('width', String(w + 2 * padding));
        rect.setAttribute('height', String(h + 2 * padding));
        if (transform) {
            rect.setAttribute('transform', transform);
        }
        return rect;
    };
    SvgHighlighter.prototype.isMactionNode = function (node) {
        return node.getAttribute('data-mml-node') === 'maction';
    };
    SvgHighlighter.prototype.getMactionNodes = function (node) {
        return Array.from(node.querySelectorAll('[data-mml-node="maction"]'));
    };
    SvgHighlighter.lineSelector = '[data-mjx-linebox]';
    SvgHighlighter.lineAttr = 'data-mjx-lineno';
    return SvgHighlighter;
}(AbstractHighlighter));
var ChtmlHighlighter = (function (_super) {
    __extends(ChtmlHighlighter, _super);
    function ChtmlHighlighter() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ChtmlHighlighter.prototype.highlightNode = function (_node) { };
    ChtmlHighlighter.prototype.unhighlightNode = function (node) {
        if (node.tagName.toLowerCase() === 'mjx-bbox') {
            node.remove();
        }
    };
    ChtmlHighlighter.prototype.createEnclosure = function (x, y, w, h, node) {
        var base = node.getBoundingClientRect();
        var enclosure = document.createElement('mjx-bbox');
        enclosure.style.width = w + 'px';
        enclosure.style.height = h + 'px';
        enclosure.style.left = x - base.left + 'px';
        enclosure.style.top = y - h - base.top + 'px';
        enclosure.style.position = 'absolute';
        node.prepend(enclosure);
        return enclosure;
    };
    ChtmlHighlighter.prototype.isMactionNode = function (node) {
        var _a;
        return ((_a = node.tagName) === null || _a === void 0 ? void 0 : _a.toLowerCase()) === 'mjx-maction';
    };
    ChtmlHighlighter.prototype.getMactionNodes = function (node) {
        return Array.from(node.querySelectorAll('mjx-maction'));
    };
    ChtmlHighlighter.lineSelector = 'mjx-linebox';
    ChtmlHighlighter.lineAttr = 'lineno';
    return ChtmlHighlighter;
}(AbstractHighlighter));
function getHighlighter(priority, back, fore, renderer) {
    var highlighter = new highlighterMapping[renderer](priority);
    highlighter.setColor(back, fore);
    return highlighter;
}
var highlighterMapping = {
    SVG: SvgHighlighter,
    CHTML: ChtmlHighlighter,
    generic: ChtmlHighlighter,
};
