// speech-rule-engine/cjs/common/debugger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Debugger = void 0;
const system_external_js_1 = require("./system_external.js");
class Debugger {
    static getInstance() {
        Debugger.instance = Debugger.instance || new Debugger();
        return Debugger.instance;
    }
    init(opt_file) {
        if (opt_file) {
            this.startDebugFile_(opt_file);
        }
        this.isActive_ = true;
        return this.fileHandle;
    }
    output(...args) {
        if (this.isActive_) {
            this.output_(args);
        }
    }
    generate(func) {
        if (this.isActive_) {
            this.output_(func.apply(func, []));
        }
    }
    exit(callback = () => { }) {
        if (this.stream_) {
            this.fileHandle.then(() => {
                if (this.isActive_ && this.stream_) {
                    this.stream_.end('', '', callback);
                    this.stream_ = null;
                }
            });
        }
        this.isActive_ = false;
    }
    constructor() {
        this.isActive_ = false;
        this.fileHandle = Promise.resolve();
        this.stream_ = null;
    }
    startDebugFile_(filename) {
        this.fileHandle = system_external_js_1.SystemExternal.fs.promises.open(filename, 'w');
        this.fileHandle = this.fileHandle.then((handle) => {
            this.stream_ = handle.createWriteStream(filename);
            this.outputFunction = function (...args) {
                this.stream_.write(args.join(' '));
                this.stream_.write('\n');
            }.bind(this);
            this.stream_.on('error', function (_error) {
                console.info('Invalid log file. Debug information sent to console.');
                this.outputFunction = console.info;
            }.bind(this));
            this.stream_.on('finish', function () {
                console.info('Finalizing debug file.');
            });
        });
    }
    output_(outputList) {
        if (this.stream_) {
            this.fileHandle.then(() => this.outputFunction.apply(this.outputFunction, ['Speech Rule Engine Debugger:'].concat(outputList)));
            return;
        }
        console.info(...['Speech Rule Engine Debugger:'].concat(outputList));
    }
}
exports.Debugger = Debugger;
