// @mathjax/src/mjs/core/MmlTree/OperatorDictionary.js
import { TEXCLASS } from './MmlNode.js';
export function OPDEF(lspace, rspace, texClass = TEXCLASS.BIN, properties = null) {
    return [lspace, rspace, texClass, properties];
}
export const MO = {
    REL: OPDEF(5, 5, TEXCLASS.REL),
    WIDEREL: OPDEF(5, 5, TEXCLASS.REL, { accent: true, stretchy: true }),
    BIN4: OPDEF(4, 4, TEXCLASS.BIN),
    RELSTRETCH: OPDEF(5, 5, TEXCLASS.REL, { stretchy: true }),
    ORD: OPDEF(0, 0, TEXCLASS.ORD),
    BIN3: OPDEF(3, 3, TEXCLASS.BIN),
    OPEN: OPDEF(0, 0, TEXCLASS.OPEN, {
        fence: true,
        stretchy: true,
        symmetric: true,
    }),
    CLOSE: OPDEF(0, 0, TEXCLASS.CLOSE, {
        fence: true,
        stretchy: true,
        symmetric: true,
    }),
    INTEGRAL: OPDEF(3, 3, TEXCLASS.OP, { largeop: true, symmetric: true }),
    ACCENT: OPDEF(0, 0, TEXCLASS.ORD, { accent: true }),
    WIDEACCENT: OPDEF(0, 0, TEXCLASS.ORD, { accent: true, stretchy: true }),
    OP: OPDEF(3, 3, TEXCLASS.OP, {
        largeop: true,
        movablelimits: true,
        symmetric: true,
    }),
    RELACCENT: OPDEF(5, 5, TEXCLASS.REL, { accent: true }),
    BIN0: OPDEF(0, 0, TEXCLASS.BIN),
    BIN5: OPDEF(5, 5, TEXCLASS.BIN),
    FENCE: OPDEF(0, 0, TEXCLASS.ORD, {
        fence: true,
        stretchy: true,
        symmetric: true,
    }),
    INNER: OPDEF(1, 1, TEXCLASS.INNER),
    ORD30: OPDEF(3, 0, TEXCLASS.ORD),
    NONE: OPDEF(0, 0, TEXCLASS.NONE),
    ORDSTRETCH0: OPDEF(0, 0, TEXCLASS.ORD, { stretchy: true }),
    BINSTRETCH0: OPDEF(0, 0, TEXCLASS.BIN, { stretchy: true }),
    RELSTRETCH0: OPDEF(0, 0, TEXCLASS.REL, { stretchy: true }),
    CLOSE0: OPDEF(0, 0, TEXCLASS.CLOSE, { fence: true }),
    ORD3: OPDEF(3, 3, TEXCLASS.ORD),
    PUNCT03: OPDEF(0, 3, TEXCLASS.PUNCT, { linebreakstyle: 'after' }),
    OPEN0: OPDEF(0, 0, TEXCLASS.OPEN, { fence: true }),
    STRETCH4: OPDEF(4, 4, TEXCLASS.BIN, { stretchy: true }),
};
export const RANGES = [
    [0x0020, 0x007f, TEXCLASS.REL, 'mo'],
    [0x00a0, 0x00bf, TEXCLASS.ORD, 'mo'],
    [0x00c0, 0x024f, TEXCLASS.ORD, 'mi'],
    [0x02b0, 0x036f, TEXCLASS.ORD, 'mo'],
    [0x0370, 0x1a20, TEXCLASS.ORD, 'mi'],
    [0x1ab0, 0x1aff, TEXCLASS.ORD, 'mo'],
    [0x1b00, 0x1dbf, TEXCLASS.ORD, 'mi'],
    [0x1dc0, 0x1dff, TEXCLASS.ORD, 'mo'],
    [0x1e00, 0x1fff, TEXCLASS.ORD, 'mi'],
    [0x2000, 0x206f, TEXCLASS.ORD, 'mo'],
    [0x2070, 0x209f, TEXCLASS.ORD, 'mo'],
    [0x2100, 0x214f, TEXCLASS.ORD, 'mi'],
    [0x2150, 0x218f, TEXCLASS.ORD, 'mn'],
    [0x2190, 0x21ff, TEXCLASS.REL, 'mo'],
    [0x2200, 0x22ff, TEXCLASS.BIN, 'mo'],
    [0x2300, 0x23ff, TEXCLASS.ORD, 'mo'],
    [0x2460, 0x24ff, TEXCLASS.ORD, 'mn'],
    [0x2500, 0x27ef, TEXCLASS.ORD, 'mo'],
    [0x27f0, 0x27ff, TEXCLASS.REL, 'mo'],
    [0x2800, 0x28ff, TEXCLASS.ORD, 'mtext'],
    [0x2900, 0x297f, TEXCLASS.REL, 'mo'],
    [0x2980, 0x29ff, TEXCLASS.ORD, 'mo'],
    [0x2a00, 0x2aff, TEXCLASS.BIN, 'mo'],
    [0x2b00, 0x2b2f, TEXCLASS.ORD, 'mo'],
    [0x2b30, 0x2b4f, TEXCLASS.REL, 'mo'],
    [0x2b50, 0x2bff, TEXCLASS.ORD, 'mo'],
    [0x2c00, 0x2de0, TEXCLASS.ORD, 'mi'],
    [0x2e00, 0x2e7f, TEXCLASS.ORD, 'mo'],
    [0x2e80, 0x2fdf, TEXCLASS.ORD, 'mi', 'normal'],
    [0x2ff0, 0x303f, TEXCLASS.ORD, 'mo'],
    [0x3040, 0xa49f, TEXCLASS.ORD, 'mi', 'normal'],
    [0xa4d0, 0xa82f, TEXCLASS.ORD, 'mi'],
    [0xa830, 0xa83f, TEXCLASS.ORD, 'mn'],
    [0xa840, 0xd7ff, TEXCLASS.ORD, 'mi'],
    [0xf900, 0xfaff, TEXCLASS.ORD, 'mi', 'normal'],
    [0xfb00, 0xfdff, TEXCLASS.ORD, 'mi'],
    [0xfe00, 0xfe6f, TEXCLASS.ORD, 'mo'],
    [0xfe70, 0x100ff, TEXCLASS.ORD, 'mi'],
    [0x10100, 0x1018f, TEXCLASS.ORD, 'mn'],
    [0x10190, 0x123ff, TEXCLASS.ORD, 'mi', 'normal'],
    [0x12400, 0x1247f, TEXCLASS.ORD, 'mn'],
    [0x12480, 0x1bc9f, TEXCLASS.ORD, 'mi', 'normal'],
    [0x1bca0, 0x1d25f, TEXCLASS.ORD, 'mo'],
    [0x1d360, 0x1d37f, TEXCLASS.ORD, 'mn'],
    [0x1d400, 0x1d7cd, TEXCLASS.ORD, 'mi'],
    [0x1d7ce, 0x1d7ff, TEXCLASS.ORD, 'mn'],
    [0x1df00, 0x1f7ff, TEXCLASS.ORD, 'mo'],
    [0x1f800, 0x1f8ff, TEXCLASS.REL, 'mo'],
    [0x1f900, 0x1f9ff, TEXCLASS.ORD, 'mo'],
    [0x20000, 0x2fa1f, TEXCLASS.ORD, 'mi', 'normal'],
];
export function getRange(text) {
    const def = OPTABLE.infix[text] || OPTABLE.prefix[text] || OPTABLE.postfix[text];
    if (def) {
        return [0, 0, def[2], 'mo'];
    }
    const n = text.codePointAt(0);
    for (const range of RANGES) {
        if (n <= range[1]) {
            if (n >= range[0]) {
                return range;
            }
            break;
        }
    }
    return [0, 0, TEXCLASS.REL, 'mo'];
}
export const MMLSPACING = [
    [0, 0],
    [1, 2],
    [3, 3],
    [4, 4],
    [0, 0],
    [0, 0],
    [0, 3],
    [1, 1],
];
export const OPTABLE = {
    prefix: {
        '!': MO.ORD,
        '(': MO.OPEN,
        '+': MO.BIN0,
        '-': MO.BIN0,
        '[': MO.OPEN,
        '{': MO.OPEN,
        '|': MO.OPEN,
        '||': MO.BIN0,
        '\u00AC': MO.ORD,
        '\u00B1': MO.BIN0,
        '\u2016': MO.FENCE,
        '\u2018': MO.OPEN0,
        '\u201C': MO.OPEN0,
        '\u2145': MO.ORD30,
        '\u2146': MO.ORD30,
        '\u2200': MO.ORD,
        '\u2201': MO.ORD,
        '\u2202': MO.ORD30,
        '\u2203': MO.ORD,
        '\u2204': MO.ORD,
        '\u2207': MO.ORD,
        '\u220F': MO.OP,
        '\u2210': MO.OP,
        '\u2211': MO.OP,
        '\u2212': MO.BIN0,
        '\u2213': MO.BIN0,
        '\u221A': [3, 0, TEXCLASS.ORD, { stretchy: true }],
        '\u221B': MO.ORD30,
        '\u221C': MO.ORD30,
        '\u221F': MO.ORD,
        '\u2220': MO.ORD,
        '\u2221': MO.ORD,
        '\u2222': MO.ORD,
        '\u222B': MO.INTEGRAL,
        '\u222C': MO.INTEGRAL,
        '\u222D': MO.INTEGRAL,
        '\u222E': MO.INTEGRAL,
        '\u222F': MO.INTEGRAL,
        '\u2230': MO.INTEGRAL,
        '\u2231': MO.INTEGRAL,
        '\u2232': MO.INTEGRAL,
        '\u2233': MO.INTEGRAL,
        '\u2234': MO.REL,
        '\u2235': MO.REL,
        '\u223C': [0, 0, TEXCLASS.REL, {}],
        '\u22BE': MO.ORD,
        '\u22BF': MO.ORD,
        '\u22C0': MO.OP,
        '\u22C1': MO.OP,
        '\u22C2': MO.OP,
        '\u22C3': MO.OP,
        '\u2308': MO.OPEN,
        '\u230A': MO.OPEN,
        '\u2310': MO.ORD,
        '\u2319': MO.ORD,
        '\u2772': MO.OPEN,
        '\u2795': MO.ORD,
        '\u2796': MO.ORD,
        '\u27C0': MO.ORD,
        '\u27E6': MO.OPEN,
        '\u27E8': MO.OPEN,
        '\u27EA': MO.OPEN,
        '\u27EC': MO.OPEN,
        '\u27EE': MO.OPEN,
        '\u2980': MO.FENCE,
        '\u2983': MO.OPEN,
        '\u2985': MO.OPEN,
        '\u2987': MO.OPEN,
        '\u2989': MO.OPEN,
        '\u298B': MO.OPEN,
        '\u298D': MO.OPEN,
        '\u298F': MO.OPEN,
        '\u2991': MO.OPEN,
        '\u2993': MO.OPEN,
        '\u2995': MO.OPEN,
        '\u2997': MO.OPEN,
        '\u2999': MO.FENCE,
        '\u299B': MO.ORD,
        '\u299C': MO.ORD,
        '\u299D': MO.ORD,
        '\u299E': MO.ORD,
        '\u299F': MO.ORD,
        '\u29A0': MO.ORD,
        '\u29A1': MO.ORD,
        '\u29A2': MO.ORD,
        '\u29A3': MO.ORD,
        '\u29A4': MO.ORD,
        '\u29A5': MO.ORD,
        '\u29A6': MO.ORD,
        '\u29A7': MO.ORD,
        '\u29A8': MO.ORD,
        '\u29A9': MO.ORD,
        '\u29AA': MO.ORD,
        '\u29AB': MO.ORD,
        '\u29AC': MO.ORD,
        '\u29AD': MO.ORD,
        '\u29AE': MO.ORD,
        '\u29AF': MO.ORD,
        '\u29D8': MO.OPEN,
        '\u29DA': MO.OPEN,
        '\u29FC': MO.OPEN,
        '\u2A00': MO.OP,
        '\u2A01': MO.OP,
        '\u2A02': MO.OP,
        '\u2A03': MO.OP,
        '\u2A04': MO.OP,
        '\u2A05': MO.OP,
        '\u2A06': MO.OP,
        '\u2A07': MO.OP,
        '\u2A08': MO.OP,
        '\u2A09': MO.OP,
        '\u2A0A': MO.OP,
        '\u2A0B': MO.INTEGRAL,
        '\u2A0C': MO.INTEGRAL,
        '\u2A0D': MO.INTEGRAL,
        '\u2A0E': MO.INTEGRAL,
        '\u2A0F': MO.INTEGRAL,
        '\u2A10': MO.INTEGRAL,
        '\u2A11': MO.INTEGRAL,
        '\u2A12': MO.INTEGRAL,
        '\u2A13': MO.INTEGRAL,
        '\u2A14': MO.INTEGRAL,
        '\u2A15': MO.INTEGRAL,
        '\u2A16': MO.INTEGRAL,
        '\u2A17': MO.INTEGRAL,
        '\u2A18': MO.INTEGRAL,
        '\u2A19': MO.INTEGRAL,
        '\u2A1A': MO.INTEGRAL,
        '\u2A1B': MO.INTEGRAL,
        '\u2A1C': MO.INTEGRAL,
        '\u2A1D': MO.OP,
        '\u2A1E': MO.OP,
        '\u2AEC': MO.ORD,
        '\u2AED': MO.ORD,
        '\u2AFC': MO.OP,
        '\u2AFF': MO.OP,
        '\u3008': MO.OPEN,
    },
    postfix: {
        '!!': MO.BIN0,
        '!': MO.CLOSE0,
        '"': MO.ORD,
        '%': MO.ORD,
        '&': MO.ORD,
        '\'': MO.ACCENT,
        ')': MO.CLOSE,
        '++': MO.BIN0,
        '--': MO.BIN0,
        ']': MO.CLOSE,
        '^': MO.WIDEACCENT,
        '_': MO.WIDEACCENT,
        '`': MO.ACCENT,
        '|': MO.CLOSE,
        '||': MO.BIN0,
        '}': MO.CLOSE,
        '~': MO.WIDEACCENT,
        '\u00A8': MO.ACCENT,
        '\u00AF': MO.WIDEACCENT,
        '\u00B0': MO.ACCENT,
        '\u00B2': MO.ORD,
        '\u00B3': MO.ORD,
        '\u00B4': MO.ACCENT,
        '\u00B8': MO.ACCENT,
        '\u00B9': MO.ORD,
        '\u02C6': MO.WIDEACCENT,
        '\u02C7': MO.WIDEACCENT,
        '\u02C9': MO.WIDEACCENT,
        '\u02CA': MO.ACCENT,
        '\u02CB': MO.ACCENT,
        '\u02CD': MO.WIDEACCENT,
        '\u02D8': MO.ACCENT,
        '\u02D9': MO.ACCENT,
        '\u02DA': MO.ACCENT,
        '\u02DC': MO.WIDEACCENT,
        '\u02DD': MO.ACCENT,
        '\u02F7': MO.WIDEACCENT,
        '\u0302': MO.WIDEACCENT,
        '\u0311': MO.ACCENT,
        '\u2016': MO.FENCE,
        '\u2019': MO.CLOSE0,
        '\u201A': MO.ORD,
        '\u201B': MO.ORD,
        '\u201D': MO.CLOSE0,
        '\u201E': MO.ORD,
        '\u201F': MO.ORD,
        '\u2032': MO.ORD,
        '\u2033': MO.ORD,
        '\u2034': MO.ORD,
        '\u2035': MO.ORD,
        '\u2036': MO.ORD,
        '\u2037': MO.ORD,
        '\u203E': MO.WIDEACCENT,
        '\u2057': MO.ORD,
        '\u20DB': MO.ACCENT,
        '\u20DC': MO.ACCENT,
        '\u2309': MO.CLOSE,
        '\u230B': MO.CLOSE,
        '\u2322': MO.RELSTRETCH0,
        '\u2323': MO.RELSTRETCH0,
        '\u23B4': MO.WIDEACCENT,
        '\u23B5': MO.WIDEACCENT,
        '\u23CD': MO.ORD,
        '\u23DC': MO.WIDEACCENT,
        '\u23DD': MO.WIDEACCENT,
        '\u23DE': MO.WIDEACCENT,
        '\u23DF': MO.WIDEACCENT,
        '\u23E0': MO.WIDEACCENT,
        '\u23E1': MO.WIDEACCENT,
        '\u2773': MO.CLOSE,
        '\u27E7': MO.CLOSE,
        '\u27E9': MO.CLOSE,
        '\u27EB': MO.CLOSE,
        '\u27ED': MO.CLOSE,
        '\u27EF': MO.CLOSE,
        '\u2980': MO.FENCE,
        '\u2984': MO.CLOSE,
        '\u2986': MO.CLOSE,
        '\u2988': MO.CLOSE,
        '\u298A': MO.CLOSE,
        '\u298C': MO.CLOSE,
        '\u298E': MO.CLOSE,
        '\u2990': MO.CLOSE,
        '\u2992': MO.CLOSE,
        '\u2994': MO.CLOSE,
        '\u2996': MO.CLOSE,
        '\u2998': MO.CLOSE,
        '\u2999': MO.FENCE,
        '\u29D9': MO.CLOSE,
        '\u29DB': MO.CLOSE,
        '\u29FD': MO.CLOSE,
        '\u3009': MO.CLOSE,
        '\u{1EEF0}': MO.BINSTRETCH0,
        '\u{1EEF1}': MO.BINSTRETCH0,
    },
    infix: {
        '!': MO.ORD,
        '!=': MO.BIN5,
        '#': MO.ORD,
        '$': MO.ORD,
        '%': MO.ORD3,
        '&&': MO.BIN4,
        '**': MO.BIN3,
        '*': MO.BIN3,
        '*=': MO.BIN5,
        '+': MO.BIN4,
        '+=': MO.BIN5,
        ',': MO.PUNCT03,
        '': MO.ORD,
        '-': MO.BIN4,
        '-=': MO.BIN5,
        '->': MO.BIN5,
        '.': MO.ORD3,
        '..': MO.BIN3,
        '...': MO.INNER,
        '/': [4, 4, TEXCLASS.ORD, {}],
        '//': MO.BIN5,
        '/=': MO.BIN5,
        ':': [0, 3, TEXCLASS.REL, {}],
        ':=': MO.BIN5,
        ';': MO.PUNCT03,
        '<': MO.REL,
        '<=': MO.REL,
        '<>': [3, 3, TEXCLASS.REL, {}],
        '=': MO.REL,
        '==': MO.REL,
        '>': MO.REL,
        '>=': MO.REL,
        '?': [3, 3, TEXCLASS.CLOSE, { fence: true }],
        '@': MO.ORD3,
        '\\': MO.ORD,
        '^': [3, 3, TEXCLASS.ORD, { accent: true, stretchy: true }],
        '_': MO.WIDEACCENT,
        '|': [5, 5, TEXCLASS.ORD, {}],
        '||': MO.BIN5,
        '\u00B1': MO.BIN4,
        '\u00B7': MO.BIN3,
        '\u00D7': MO.BIN3,
        '\u00F7': MO.BIN4,
        '\u02B9': MO.ORD,
        '\u0300': MO.ACCENT,
        '\u0301': MO.ACCENT,
        '\u0303': MO.WIDEACCENT,
        '\u0304': MO.ACCENT,
        '\u0306': MO.ACCENT,
        '\u0307': MO.ACCENT,
        '\u0308': MO.ACCENT,
        '\u030C': MO.ACCENT,
        '\u0332': MO.WIDEACCENT,
        '\u0338': MO.REL,
        '\u03F6': MO.REL,
        '\u2015': MO.ORDSTRETCH0,
        '\u2017': MO.ORDSTRETCH0,
        '\u2020': MO.BIN3,
        '\u2021': MO.BIN3,
        '\u2022': MO.BIN3,
        '\u2026': MO.INNER,
        '\u2043': MO.BIN3,
        '\u2044': MO.STRETCH4,
        '\u2061': MO.NONE,
        '\u2062': MO.NONE,
        '\u2063': [0, 0, TEXCLASS.NONE, { linebreakstyle: 'after' }],
        '\u2064': MO.NONE,
        '\u20D7': MO.ACCENT,
        '\u2111': MO.ORD,
        '\u2113': MO.ORD,
        '\u2118': MO.ORD,
        '\u211C': MO.ORD,
        '\u2190': MO.WIDEREL,
        '\u2191': MO.RELSTRETCH,
        '\u2192': MO.WIDEREL,
        '\u2193': MO.RELSTRETCH,
        '\u2194': MO.WIDEREL,
        '\u2195': MO.RELSTRETCH,
        '\u2196': MO.REL,
        '\u2197': MO.REL,
        '\u2198': MO.REL,
        '\u2199': MO.REL,
        '\u219A': MO.WIDEREL,
        '\u219B': MO.WIDEREL,
        '\u219C': MO.WIDEREL,
        '\u219D': MO.WIDEREL,
        '\u219E': MO.WIDEREL,
        '\u219F': MO.RELSTRETCH,
        '\u21A0': MO.WIDEREL,
        '\u21A1': MO.RELSTRETCH,
        '\u21A2': MO.WIDEREL,
        '\u21A3': MO.WIDEREL,
        '\u21A4': MO.WIDEREL,
        '\u21A5': MO.RELSTRETCH,
        '\u21A6': MO.WIDEREL,
        '\u21A7': MO.RELSTRETCH,
        '\u21A8': MO.RELSTRETCH,
        '\u21A9': MO.WIDEREL,
        '\u21AA': MO.WIDEREL,
        '\u21AB': MO.WIDEREL,
        '\u21AC': MO.WIDEREL,
        '\u21AD': MO.WIDEREL,
        '\u21AE': MO.WIDEREL,
        '\u21AF': MO.REL,
        '\u21B0': MO.RELSTRETCH,
        '\u21B1': MO.RELSTRETCH,
        '\u21B2': MO.RELSTRETCH,
        '\u21B3': MO.RELSTRETCH,
        '\u21B4': MO.RELSTRETCH,
        '\u21B5': MO.RELSTRETCH,
        '\u21B6': MO.REL,
        '\u21B7': MO.REL,
        '\u21B8': MO.REL,
        '\u21B9': MO.WIDEREL,
        '\u21BA': MO.REL,
        '\u21BB': MO.REL,
        '\u21BC': MO.WIDEREL,
        '\u21BD': MO.WIDEREL,
        '\u21BE': MO.RELSTRETCH,
        '\u21BF': MO.RELSTRETCH,
        '\u21C0': MO.WIDEREL,
        '\u21C1': MO.WIDEREL,
        '\u21C2': MO.RELSTRETCH,
        '\u21C3': MO.RELSTRETCH,
        '\u21C4': MO.WIDEREL,
        '\u21C5': MO.RELSTRETCH,
        '\u21C6': MO.WIDEREL,
        '\u21C7': MO.WIDEREL,
        '\u21C8': MO.RELSTRETCH,
        '\u21C9': MO.WIDEREL,
        '\u21CA': MO.RELSTRETCH,
        '\u21CB': MO.WIDEREL,
        '\u21CC': MO.WIDEREL,
        '\u21CD': MO.WIDEREL,
        '\u21CE': MO.WIDEREL,
        '\u21CF': MO.WIDEREL,
        '\u21D0': MO.WIDEREL,
        '\u21D1': MO.RELSTRETCH,
        '\u21D2': MO.WIDEREL,
        '\u21D3': MO.RELSTRETCH,
        '\u21D4': MO.WIDEREL,
        '\u21D5': MO.RELSTRETCH,
        '\u21D6': MO.REL,
        '\u21D7': MO.REL,
        '\u21D8': MO.REL,
        '\u21D9': MO.REL,
        '\u21DA': MO.WIDEREL,
        '\u21DB': MO.WIDEREL,
        '\u21DC': MO.WIDEREL,
        '\u21DD': MO.WIDEREL,
        '\u21DE': MO.RELSTRETCH,
        '\u21DF': MO.RELSTRETCH,
        '\u21E0': MO.WIDEREL,
        '\u21E1': MO.RELSTRETCH,
        '\u21E2': MO.WIDEREL,
        '\u21E3': MO.RELSTRETCH,
        '\u21E4': MO.WIDEREL,
        '\u21E5': MO.WIDEREL,
        '\u21E6': MO.WIDEREL,
        '\u21E7': MO.RELSTRETCH,
        '\u21E8': MO.WIDEREL,
        '\u21E9': MO.RELSTRETCH,
        '\u21EA': MO.RELSTRETCH,
        '\u21EB': MO.RELSTRETCH,
        '\u21EC': MO.RELSTRETCH,
        '\u21ED': MO.RELSTRETCH,
        '\u21EE': MO.RELSTRETCH,
        '\u21EF': MO.RELSTRETCH,
        '\u21F0': MO.WIDEREL,
        '\u21F1': MO.REL,
        '\u21F2': MO.REL,
        '\u21F3': MO.RELSTRETCH,
        '\u21F4': MO.WIDEREL,
        '\u21F5': MO.RELSTRETCH,
        '\u21F6': MO.WIDEREL,
        '\u21F7': MO.WIDEREL,
        '\u21F8': MO.WIDEREL,
        '\u21F9': MO.WIDEREL,
        '\u21FA': MO.WIDEREL,
        '\u21FB': MO.WIDEREL,
        '\u21FC': MO.WIDEREL,
        '\u21FD': MO.WIDEREL,
        '\u21FE': MO.WIDEREL,
        '\u21FF': MO.WIDEREL,
        '\u2205': MO.ORD,
        '\u2206': MO.ORD,
        '\u2208': MO.REL,
        '\u2209': MO.REL,
        '\u220A': MO.REL,
        '\u220B': MO.REL,
        '\u220C': MO.REL,
        '\u220D': MO.REL,
        '\u2212': MO.BIN4,
        '\u2213': MO.BIN4,
        '\u2214': MO.BIN4,
        '\u2215': MO.STRETCH4,
        '\u2216': MO.BIN4,
        '\u2217': MO.BIN3,
        '\u2218': MO.BIN3,
        '\u2219': MO.BIN3,
        '\u221D': MO.REL,
        '\u221E': MO.ORD,
        '\u2223': MO.REL,
        '\u2224': MO.REL,
        '\u2225': MO.REL,
        '\u2226': MO.REL,
        '\u2227': MO.BIN4,
        '\u2228': MO.BIN4,
        '\u2229': MO.BIN4,
        '\u222A': MO.BIN4,
        '\u2236': MO.BIN4,
        '\u2237': MO.REL,
        '\u2238': MO.BIN4,
        '\u2239': MO.REL,
        '\u223A': MO.REL,
        '\u223B': MO.REL,
        '\u223C': MO.REL,
        '\u223D': MO.REL,
        '\u223E': MO.REL,
        '\u2240': MO.BIN3,
        '\u2241': MO.REL,
        '\u2242': MO.REL,
        '\u2242\u0338': MO.REL,
        '\u2243': MO.REL,
        '\u2244': MO.REL,
        '\u2245': MO.REL,
        '\u2246': MO.REL,
        '\u2247': MO.REL,
        '\u2248': MO.REL,
        '\u2249': MO.REL,
        '\u224A': MO.REL,
        '\u224B': MO.REL,
        '\u224C': MO.REL,
        '\u224D': MO.REL,
        '\u224E': MO.REL,
        '\u224F': MO.REL,
        '\u2250': MO.REL,
        '\u2251': MO.REL,
        '\u2252': MO.REL,
        '\u2253': MO.REL,
        '\u2254': MO.REL,
        '\u2255': MO.REL,
        '\u2256': MO.REL,
        '\u2257': MO.REL,
        '\u2258': MO.REL,
        '\u2259': MO.REL,
        '\u225A': MO.REL,
        '\u225B': MO.REL,
        '\u225C': MO.REL,
        '\u225D': MO.REL,
        '\u225E': MO.REL,
        '\u225F': MO.REL,
        '\u2260': MO.REL,
        '\u2261': MO.REL,
        '\u2262': MO.REL,
        '\u2263': MO.REL,
        '\u2264': MO.REL,
        '\u2265': MO.REL,
        '\u2266': MO.REL,
        '\u2266\u0338': MO.REL,
        '\u2267': MO.REL,
        '\u2267\u0338': MO.REL,
        '\u2268': MO.REL,
        '\u2269': MO.REL,
        '\u226A': MO.REL,
        '\u226A\u0338': MO.REL,
        '\u226B': MO.REL,
        '\u226B\u0338': MO.REL,
        '\u226C': MO.REL,
        '\u226D': MO.REL,
        '\u226E': MO.REL,
        '\u226F': MO.REL,
        '\u2270': MO.REL,
        '\u2271': MO.REL,
        '\u2272': MO.REL,
        '\u2273': MO.REL,
        '\u2274': MO.REL,
        '\u2275': MO.REL,
        '\u2276': MO.REL,
        '\u2277': MO.REL,
        '\u2278': MO.REL,
        '\u2279': MO.REL,
        '\u227A': MO.REL,
        '\u227B': MO.REL,
        '\u227C': MO.REL,
        '\u227D': MO.REL,
        '\u227E': MO.REL,
        '\u227E\u0338': MO.REL,
        '\u227F': MO.REL,
        '\u227F\u0338': MO.REL,
        '\u2280': MO.REL,
        '\u2281': MO.REL,
        '\u2282': MO.REL,
        '\u2283': MO.REL,
        '\u2284': MO.REL,
        '\u2285': MO.REL,
        '\u2286': MO.REL,
        '\u2287': MO.REL,
        '\u2288': MO.REL,
        '\u2289': MO.REL,
        '\u228A': MO.REL,
        '\u228B': MO.REL,
        '\u228C': MO.BIN4,
        '\u228D': MO.BIN4,
        '\u228E': MO.BIN4,
        '\u228F': MO.REL,
        '\u228F\u0338': MO.REL,
        '\u2290': MO.REL,
        '\u2290\u0338': MO.REL,
        '\u2291': MO.REL,
        '\u2292': MO.REL,
        '\u2293': MO.BIN4,
        '\u2294': MO.BIN4,
        '\u2295': MO.BIN4,
        '\u2296': MO.BIN4,
        '\u2297': MO.BIN3,
        '\u2298': MO.BIN4,
        '\u2299': MO.BIN3,
        '\u229A': MO.BIN3,
        '\u229B': MO.BIN3,
        '\u229C': MO.REL,
        '\u229D': MO.BIN4,
        '\u229E': MO.BIN4,
        '\u229F': MO.BIN4,
        '\u22A0': MO.BIN3,
        '\u22A1': MO.BIN3,
        '\u22A2': MO.REL,
        '\u22A3': MO.REL,
        '\u22A4': MO.ORD,
        '\u22A5': MO.ORD,
        '\u22A6': MO.REL,
        '\u22A7': MO.REL,
        '\u22A8': MO.REL,
        '\u22A9': MO.REL,
        '\u22AA': MO.REL,
        '\u22AB': MO.REL,
        '\u22AC': MO.REL,
        '\u22AD': MO.REL,
        '\u22AE': MO.REL,
        '\u22AF': MO.REL,
        '\u22B0': MO.REL,
        '\u22B1': MO.REL,
        '\u22B2': MO.REL,
        '\u22B3': MO.REL,
        '\u22B4': MO.REL,
        '\u22B5': MO.REL,
        '\u22B6': MO.REL,
        '\u22B7': MO.REL,
        '\u22B8': MO.REL,
        '\u22BA': MO.BIN3,
        '\u22BB': MO.BIN4,
        '\u22BC': MO.BIN4,
        '\u22BD': MO.BIN4,
        '\u22C4': MO.BIN3,
        '\u22C5': MO.BIN3,
        '\u22C6': MO.BIN3,
        '\u22C7': MO.BIN3,
        '\u22C8': MO.REL,
        '\u22C9': MO.BIN3,
        '\u22CA': MO.BIN3,
        '\u22CB': MO.BIN3,
        '\u22CC': MO.BIN3,
        '\u22CD': MO.REL,
        '\u22CE': MO.BIN4,
        '\u22CF': MO.BIN4,
        '\u22D0': MO.REL,
        '\u22D1': MO.REL,
        '\u22D2': MO.BIN4,
        '\u22D3': MO.BIN4,
        '\u22D4': MO.REL,
        '\u22D5': MO.REL,
        '\u22D6': MO.REL,
        '\u22D7': MO.REL,
        '\u22D8': MO.REL,
        '\u22D9': MO.REL,
        '\u22DA': MO.REL,
        '\u22DB': MO.REL,
        '\u22DC': MO.REL,
        '\u22DD': MO.REL,
        '\u22DE': MO.REL,
        '\u22DF': MO.REL,
        '\u22E0': MO.REL,
        '\u22E1': MO.REL,
        '\u22E2': MO.REL,
        '\u22E3': MO.REL,
        '\u22E4': MO.REL,
        '\u22E5': MO.REL,
        '\u22E6': MO.REL,
        '\u22E7': MO.REL,
        '\u22E8': MO.REL,
        '\u22E9': MO.REL,
        '\u22EA': MO.REL,
        '\u22EB': MO.REL,
        '\u22EC': MO.REL,
        '\u22ED': MO.REL,
        '\u22EE': MO.ORD,
        '\u22EF': MO.INNER,
        '\u22F0': MO.INNER,
        '\u22F1': MO.INNER,
        '\u22F2': MO.REL,
        '\u22F3': MO.REL,
        '\u22F4': MO.REL,
        '\u22F5': MO.REL,
        '\u22F6': MO.REL,
        '\u22F7': MO.REL,
        '\u22F8': MO.REL,
        '\u22F9': MO.REL,
        '\u22FA': MO.REL,
        '\u22FB': MO.REL,
        '\u22FC': MO.REL,
        '\u22FD': MO.REL,
        '\u22FE': MO.REL,
        '\u22FF': MO.REL,
        '\u2301': MO.REL,
        '\u2305': MO.BIN3,
        '\u2306': MO.BIN3,
        '\u2329': MO.OPEN,
        '\u232A': MO.CLOSE,
        '\u237C': MO.REL,
        '\u238B': MO.REL,
        '\u23AA': MO.ORD,
        '\u23AF': MO.ORDSTRETCH0,
        '\u23B0': MO.OPEN,
        '\u23B1': MO.CLOSE,
        '\u2500': MO.ORD,
        '\u25B3': MO.BIN3,
        '\u25B5': MO.BIN3,
        '\u25B9': MO.BIN3,
        '\u25BD': MO.BIN3,
        '\u25BF': MO.BIN3,
        '\u25C3': MO.BIN3,
        '\u25EF': MO.BIN3,
        '\u2660': MO.ORD,
        '\u2661': MO.ORD,
        '\u2662': MO.ORD,
        '\u2663': MO.ORD,
        '\u266D': MO.ORD,
        '\u266E': MO.ORD,
        '\u266F': MO.ORD,
        '\u2758': [5, 5, TEXCLASS.REL, { stretchy: true, symmetric: true }],
        '\u2794': MO.WIDEREL,
        '\u2795': MO.BIN4,
        '\u2796': MO.BIN4,
        '\u2797': MO.BIN4,
        '\u2798': MO.REL,
        '\u2799': MO.WIDEREL,
        '\u279A': MO.REL,
        '\u279B': MO.WIDEREL,
        '\u279C': MO.WIDEREL,
        '\u279D': MO.WIDEREL,
        '\u279E': MO.WIDEREL,
        '\u279F': MO.WIDEREL,
        '\u27A0': MO.WIDEREL,
        '\u27A1': MO.WIDEREL,
        '\u27A5': MO.WIDEREL,
        '\u27A6': MO.WIDEREL,
        '\u27A7': MO.RELACCENT,
        '\u27A8': MO.WIDEREL,
        '\u27A9': MO.WIDEREL,
        '\u27AA': MO.WIDEREL,
        '\u27AB': MO.WIDEREL,
        '\u27AC': MO.WIDEREL,
        '\u27AD': MO.WIDEREL,
        '\u27AE': MO.WIDEREL,
        '\u27AF': MO.WIDEREL,
        '\u27B1': MO.WIDEREL,
        '\u27B2': MO.RELACCENT,
        '\u27B3': MO.WIDEREL,
        '\u27B4': MO.REL,
        '\u27B5': MO.WIDEREL,
        '\u27B6': MO.REL,
        '\u27B7': MO.REL,
        '\u27B8': MO.WIDEREL,
        '\u27B9': MO.REL,
        '\u27BA': MO.WIDEREL,
        '\u27BB': MO.WIDEREL,
        '\u27BC': MO.WIDEREL,
        '\u27BD': MO.WIDEREL,
        '\u27BE': MO.WIDEREL,
        '\u27C2': MO.REL,
        '\u27C2\u0338': MO.REL,
        '\u27CB': MO.BIN3,
        '\u27CD': MO.BIN3,
        '\u27F0': MO.RELSTRETCH,
        '\u27F1': MO.RELSTRETCH,
        '\u27F2': MO.REL,
        '\u27F3': MO.REL,
        '\u27F4': MO.RELSTRETCH,
        '\u27F5': MO.WIDEREL,
        '\u27F6': MO.WIDEREL,
        '\u27F7': MO.WIDEREL,
        '\u27F8': MO.WIDEREL,
        '\u27F9': MO.WIDEREL,
        '\u27FA': MO.WIDEREL,
        '\u27FB': MO.WIDEREL,
        '\u27FC': MO.WIDEREL,
        '\u27FD': MO.WIDEREL,
        '\u27FE': MO.WIDEREL,
        '\u27FF': MO.WIDEREL,
        '\u2900': MO.WIDEREL,
        '\u2901': MO.WIDEREL,
        '\u2902': MO.WIDEREL,
        '\u2903': MO.WIDEREL,
        '\u2904': MO.WIDEREL,
        '\u2905': MO.WIDEREL,
        '\u2906': MO.WIDEREL,
        '\u2907': MO.WIDEREL,
        '\u2908': MO.RELSTRETCH,
        '\u2909': MO.RELSTRETCH,
        '\u290A': MO.RELSTRETCH,
        '\u290B': MO.RELSTRETCH,
        '\u290C': MO.WIDEREL,
        '\u290D': MO.WIDEREL,
        '\u290E': MO.WIDEREL,
        '\u290F': MO.WIDEREL,
        '\u2910': MO.WIDEREL,
        '\u2911': MO.WIDEREL,
        '\u2912': MO.RELSTRETCH,
        '\u2913': MO.RELSTRETCH,
        '\u2914': MO.WIDEREL,
        '\u2915': MO.WIDEREL,
        '\u2916': MO.WIDEREL,
        '\u2917': MO.WIDEREL,
        '\u2918': MO.WIDEREL,
        '\u2919': MO.WIDEREL,
        '\u291A': MO.WIDEREL,
        '\u291B': MO.WIDEREL,
        '\u291C': MO.WIDEREL,
        '\u291D': MO.WIDEREL,
        '\u291E': MO.WIDEREL,
        '\u291F': MO.WIDEREL,
        '\u2920': MO.WIDEREL,
        '\u2921': MO.REL,
        '\u2922': MO.REL,
        '\u2923': MO.REL,
        '\u2924': MO.REL,
        '\u2925': MO.REL,
        '\u2926': MO.REL,
        '\u2927': MO.REL,
        '\u2928': MO.REL,
        '\u2929': MO.REL,
        '\u292A': MO.REL,
        '\u292B': MO.REL,
        '\u292C': MO.REL,
        '\u292D': MO.REL,
        '\u292E': MO.REL,
        '\u292F': MO.REL,
        '\u2930': MO.REL,
        '\u2931': MO.REL,
        '\u2932': MO.REL,
        '\u2933': MO.RELACCENT,
        '\u2934': MO.RELSTRETCH,
        '\u2935': MO.RELSTRETCH,
        '\u2936': MO.RELSTRETCH,
        '\u2937': MO.RELSTRETCH,
        '\u2938': MO.REL,
        '\u2939': MO.REL,
        '\u293A': MO.RELACCENT,
        '\u293B': MO.RELACCENT,
        '\u293C': MO.RELACCENT,
        '\u293D': MO.RELACCENT,
        '\u293E': MO.REL,
        '\u293F': MO.REL,
        '\u2940': MO.REL,
        '\u2941': MO.REL,
        '\u2942': MO.WIDEREL,
        '\u2943': MO.WIDEREL,
        '\u2944': MO.WIDEREL,
        '\u2945': MO.RELSTRETCH,
        '\u2946': MO.RELSTRETCH,
        '\u2947': MO.WIDEREL,
        '\u2948': MO.WIDEREL,
        '\u2949': MO.RELSTRETCH,
        '\u294A': MO.WIDEREL,
        '\u294B': MO.WIDEREL,
        '\u294C': MO.RELSTRETCH,
        '\u294D': MO.RELSTRETCH,
        '\u294E': MO.WIDEREL,
        '\u294F': MO.RELSTRETCH,
        '\u2950': MO.WIDEREL,
        '\u2951': MO.RELSTRETCH,
        '\u2952': MO.WIDEREL,
        '\u2953': MO.WIDEREL,
        '\u2954': MO.RELSTRETCH,
        '\u2955': MO.RELSTRETCH,
        '\u2956': MO.WIDEREL,
        '\u2957': MO.WIDEREL,
        '\u2958': MO.RELSTRETCH,
        '\u2959': MO.RELSTRETCH,
        '\u295A': MO.WIDEREL,
        '\u295B': MO.WIDEREL,
        '\u295C': MO.RELSTRETCH,
        '\u295D': MO.RELSTRETCH,
        '\u295E': MO.WIDEREL,
        '\u295F': MO.WIDEREL,
        '\u2960': MO.RELSTRETCH,
        '\u2961': MO.RELSTRETCH,
        '\u2962': MO.WIDEREL,
        '\u2963': MO.RELSTRETCH,
        '\u2964': MO.WIDEREL,
        '\u2965': MO.RELSTRETCH,
        '\u2966': MO.WIDEREL,
        '\u2967': MO.WIDEREL,
        '\u2968': MO.WIDEREL,
        '\u2969': MO.WIDEREL,
        '\u296A': MO.WIDEREL,
        '\u296B': MO.WIDEREL,
        '\u296C': MO.WIDEREL,
        '\u296D': MO.WIDEREL,
        '\u296E': MO.RELSTRETCH,
        '\u296F': MO.RELSTRETCH,
        '\u2970': MO.WIDEREL,
        '\u2971': MO.WIDEREL,
        '\u2972': MO.WIDEREL,
        '\u2973': MO.WIDEREL,
        '\u2974': MO.WIDEREL,
        '\u2975': MO.WIDEREL,
        '\u2976': MO.RELACCENT,
        '\u2977': MO.RELACCENT,
        '\u2978': MO.RELACCENT,
        '\u2979': MO.RELACCENT,
        '\u297A': MO.RELACCENT,
        '\u297B': MO.RELACCENT,
        '\u297C': MO.WIDEREL,
        '\u297D': MO.WIDEREL,
        '\u297E': MO.RELSTRETCH,
        '\u297F': MO.RELSTRETCH,
        '\u2981': MO.REL,
        '\u2982': MO.REL,
        '\u29B6': MO.REL,
        '\u29B7': MO.REL,
        '\u29B8': MO.BIN4,
        '\u29B9': MO.REL,
        '\u29BC': MO.BIN4,
        '\u29C0': MO.REL,
        '\u29C1': MO.REL,
        '\u29C4': MO.BIN4,
        '\u29C5': MO.BIN4,
        '\u29C6': MO.BIN3,
        '\u29C7': MO.BIN3,
        '\u29C8': MO.BIN3,
        '\u29CE': MO.REL,
        '\u29CF': MO.REL,
        '\u29D0': MO.REL,
        '\u29D1': MO.REL,
        '\u29D2': MO.REL,
        '\u29D3': MO.REL,
        '\u29D4': MO.BIN3,
        '\u29D5': MO.BIN3,
        '\u29D6': MO.BIN3,
        '\u29D7': MO.BIN3,
        '\u29DF': MO.REL,
        '\u29E1': MO.REL,
        '\u29E2': MO.BIN3,
        '\u29E3': MO.REL,
        '\u29E4': MO.REL,
        '\u29E5': MO.REL,
        '\u29E6': MO.REL,
        '\u29F4': MO.REL,
        '\u29F5': MO.BIN4,
        '\u29F6': MO.BIN4,
        '\u29F7': MO.BIN4,
        '\u29F8': MO.BIN4,
        '\u29F9': MO.BIN4,
        '\u29FA': MO.BIN4,
        '\u29FB': MO.BIN4,
        '\u2A1D': MO.BIN3,
        '\u2A1E': MO.BIN3,
        '\u2A1F': MO.BIN4,
        '\u2A20': MO.BIN4,
        '\u2A21': MO.BIN4,
        '\u2A22': MO.BIN4,
        '\u2A23': MO.BIN4,
        '\u2A24': MO.BIN4,
        '\u2A25': MO.BIN4,
        '\u2A26': MO.BIN4,
        '\u2A27': MO.BIN4,
        '\u2A28': MO.BIN4,
        '\u2A29': MO.BIN4,
        '\u2A2A': MO.BIN4,
        '\u2A2B': MO.BIN4,
        '\u2A2C': MO.BIN4,
        '\u2A2D': MO.BIN4,
        '\u2A2E': MO.BIN4,
        '\u2A2F': MO.BIN3,
        '\u2A30': MO.BIN3,
        '\u2A31': MO.BIN3,
        '\u2A32': MO.BIN3,
        '\u2A33': MO.BIN3,
        '\u2A34': MO.BIN3,
        '\u2A35': MO.BIN3,
        '\u2A36': MO.BIN3,
        '\u2A37': MO.BIN3,
        '\u2A38': MO.BIN4,
        '\u2A39': MO.BIN4,
        '\u2A3A': MO.BIN4,
        '\u2A3B': MO.BIN3,
        '\u2A3C': MO.BIN3,
        '\u2A3D': MO.BIN3,
        '\u2A3E': MO.BIN4,
        '\u2A3F': MO.BIN3,
        '\u2A40': MO.BIN4,
        '\u2A41': MO.BIN4,
        '\u2A42': MO.BIN4,
        '\u2A43': MO.BIN4,
        '\u2A44': MO.BIN4,
        '\u2A45': MO.BIN4,
        '\u2A46': MO.BIN4,
        '\u2A47': MO.BIN4,
        '\u2A48': MO.BIN4,
        '\u2A49': MO.BIN4,
        '\u2A4A': MO.BIN4,
        '\u2A4B': MO.BIN4,
        '\u2A4C': MO.BIN4,
        '\u2A4D': MO.BIN4,
        '\u2A4E': MO.BIN4,
        '\u2A4F': MO.BIN4,
        '\u2A50': MO.BIN3,
        '\u2A51': MO.BIN4,
        '\u2A52': MO.BIN4,
        '\u2A53': MO.BIN4,
        '\u2A54': MO.BIN4,
        '\u2A55': MO.BIN4,
        '\u2A56': MO.BIN4,
        '\u2A57': MO.BIN4,
        '\u2A58': MO.BIN4,
        '\u2A59': MO.BIN4,
        '\u2A5A': MO.BIN4,
        '\u2A5B': MO.BIN4,
        '\u2A5C': MO.BIN4,
        '\u2A5D': MO.BIN4,
        '\u2A5E': MO.BIN4,
        '\u2A5F': MO.BIN4,
        '\u2A60': MO.BIN4,
        '\u2A61': MO.BIN4,
        '\u2A62': MO.BIN4,
        '\u2A63': MO.BIN4,
        '\u2A64': MO.BIN3,
        '\u2A65': MO.BIN3,
        '\u2A66': MO.REL,
        '\u2A67': MO.REL,
        '\u2A68': MO.REL,
        '\u2A69': MO.REL,
        '\u2A6A': MO.REL,
        '\u2A6B': MO.REL,
        '\u2A6C': MO.REL,
        '\u2A6D': MO.REL,
        '\u2A6E': MO.REL,
        '\u2A6F': MO.REL,
        '\u2A70': MO.REL,
        '\u2A71': MO.REL,
        '\u2A72': MO.REL,
        '\u2A73': MO.REL,
        '\u2A74': MO.REL,
        '\u2A75': MO.REL,
        '\u2A76': MO.REL,
        '\u2A77': MO.REL,
        '\u2A78': MO.REL,
        '\u2A79': MO.REL,
        '\u2A7A': MO.REL,
        '\u2A7B': MO.REL,
        '\u2A7C': MO.REL,
        '\u2A7D': MO.REL,
        '\u2A7D\u0338': MO.REL,
        '\u2A7E': MO.REL,
        '\u2A7E\u0338': MO.REL,
        '\u2A7F': MO.REL,
        '\u2A80': MO.REL,
        '\u2A81': MO.REL,
        '\u2A82': MO.REL,
        '\u2A83': MO.REL,
        '\u2A84': MO.REL,
        '\u2A85': MO.REL,
        '\u2A86': MO.REL,
        '\u2A87': MO.REL,
        '\u2A88': MO.REL,
        '\u2A89': MO.REL,
        '\u2A8A': MO.REL,
        '\u2A8B': MO.REL,
        '\u2A8C': MO.REL,
        '\u2A8D': MO.REL,
        '\u2A8E': MO.REL,
        '\u2A8F': MO.REL,
        '\u2A90': MO.REL,
        '\u2A91': MO.REL,
        '\u2A92': MO.REL,
        '\u2A93': MO.REL,
        '\u2A94': MO.REL,
        '\u2A95': MO.REL,
        '\u2A96': MO.REL,
        '\u2A97': MO.REL,
        '\u2A98': MO.REL,
        '\u2A99': MO.REL,
        '\u2A9A': MO.REL,
        '\u2A9B': MO.REL,
        '\u2A9C': MO.REL,
        '\u2A9D': MO.REL,
        '\u2A9E': MO.REL,
        '\u2A9F': MO.REL,
        '\u2AA0': MO.REL,
        '\u2AA1': MO.REL,
        '\u2AA2': MO.REL,
        '\u2AA3': MO.REL,
        '\u2AA4': MO.REL,
        '\u2AA5': MO.REL,
        '\u2AA6': MO.REL,
        '\u2AA7': MO.REL,
        '\u2AA8': MO.REL,
        '\u2AA9': MO.REL,
        '\u2AAA': MO.REL,
        '\u2AAB': MO.REL,
        '\u2AAC': MO.REL,
        '\u2AAD': MO.REL,
        '\u2AAE': MO.REL,
        '\u2AAF': MO.REL,
        '\u2AAF\u0338': MO.REL,
        '\u2AB0': MO.REL,
        '\u2AB0\u0338': MO.REL,
        '\u2AB1': MO.REL,
        '\u2AB2': MO.REL,
        '\u2AB3': MO.REL,
        '\u2AB4': MO.REL,
        '\u2AB5': MO.REL,
        '\u2AB6': MO.REL,
        '\u2AB7': MO.REL,
        '\u2AB8': MO.REL,
        '\u2AB9': MO.REL,
        '\u2ABA': MO.REL,
        '\u2ABB': MO.REL,
        '\u2ABC': MO.REL,
        '\u2ABD': MO.REL,
        '\u2ABE': MO.REL,
        '\u2ABF': MO.REL,
        '\u2AC0': MO.REL,
        '\u2AC1': MO.REL,
        '\u2AC2': MO.REL,
        '\u2AC3': MO.REL,
        '\u2AC4': MO.REL,
        '\u2AC5': MO.REL,
        '\u2AC6': MO.REL,
        '\u2AC7': MO.REL,
        '\u2AC8': MO.REL,
        '\u2AC9': MO.REL,
        '\u2ACA': MO.REL,
        '\u2ACB': MO.REL,
        '\u2ACC': MO.REL,
        '\u2ACD': MO.REL,
        '\u2ACE': MO.REL,
        '\u2ACF': MO.REL,
        '\u2AD0': MO.REL,
        '\u2AD1': MO.REL,
        '\u2AD2': MO.REL,
        '\u2AD3': MO.REL,
        '\u2AD4': MO.REL,
        '\u2AD5': MO.REL,
        '\u2AD6': MO.REL,
        '\u2AD7': MO.REL,
        '\u2AD8': MO.REL,
        '\u2AD9': MO.REL,
        '\u2ADA': MO.REL,
        '\u2ADB': MO.BIN4,
        '\u2ADD': MO.BIN3,
        '\u2ADD\u0338': MO.REL,
        '\u2ADE': MO.REL,
        '\u2ADF': MO.REL,
        '\u2AE0': MO.REL,
        '\u2AE1': MO.REL,
        '\u2AE2': MO.REL,
        '\u2AE3': MO.REL,
        '\u2AE4': MO.REL,
        '\u2AE5': MO.REL,
        '\u2AE6': MO.REL,
        '\u2AE7': MO.REL,
        '\u2AE8': MO.REL,
        '\u2AE9': MO.REL,
        '\u2AEA': MO.REL,
        '\u2AEB': MO.REL,
        '\u2AEE': MO.REL,
        '\u2AF2': MO.REL,
        '\u2AF3': MO.REL,
        '\u2AF4': MO.REL,
        '\u2AF5': MO.REL,
        '\u2AF6': MO.BIN4,
        '\u2AF7': MO.REL,
        '\u2AF8': MO.REL,
        '\u2AF9': MO.REL,
        '\u2AFA': MO.REL,
        '\u2AFB': MO.BIN4,
        '\u2AFD': MO.BIN4,
        '\u2AFE': MO.BIN3,
        '\u2B00': MO.REL,
        '\u2B01': MO.REL,
        '\u2B02': MO.REL,
        '\u2B03': MO.REL,
        '\u2B04': MO.WIDEREL,
        '\u2B05': MO.WIDEREL,
        '\u2B06': MO.RELSTRETCH,
        '\u2B07': MO.RELSTRETCH,
        '\u2B08': MO.REL,
        '\u2B09': MO.REL,
        '\u2B0A': MO.REL,
        '\u2B0B': MO.REL,
        '\u2B0C': MO.WIDEREL,
        '\u2B0D': MO.RELSTRETCH,
        '\u2B0E': MO.RELSTRETCH,
        '\u2B0F': MO.RELSTRETCH,
        '\u2B10': MO.RELSTRETCH,
        '\u2B11': MO.RELSTRETCH,
        '\u2B30': MO.WIDEREL,
        '\u2B31': MO.WIDEREL,
        '\u2B32': MO.RELSTRETCH,
        '\u2B33': MO.WIDEREL,
        '\u2B34': MO.WIDEREL,
        '\u2B35': MO.WIDEREL,
        '\u2B36': MO.WIDEREL,
        '\u2B37': MO.WIDEREL,
        '\u2B38': MO.WIDEREL,
        '\u2B39': MO.WIDEREL,
        '\u2B3A': MO.WIDEREL,
        '\u2B3B': MO.WIDEREL,
        '\u2B3C': MO.WIDEREL,
        '\u2B3D': MO.WIDEREL,
        '\u2B3E': MO.WIDEREL,
        '\u2B3F': MO.RELACCENT,
        '\u2B40': MO.WIDEREL,
        '\u2B41': MO.WIDEREL,
        '\u2B42': MO.WIDEREL,
        '\u2B43': MO.WIDEREL,
        '\u2B44': MO.WIDEREL,
        '\u2B45': MO.WIDEREL,
        '\u2B46': MO.WIDEREL,
        '\u2B47': MO.WIDEREL,
        '\u2B48': MO.WIDEREL,
        '\u2B49': MO.WIDEREL,
        '\u2B4A': MO.WIDEREL,
        '\u2B4B': MO.WIDEREL,
        '\u2B4C': MO.WIDEREL,
        '\u2B4D': MO.REL,
        '\u2B4E': MO.REL,
        '\u2B4F': MO.REL,
        '\u2B5A': MO.REL,
        '\u2B5B': MO.REL,
        '\u2B5C': MO.REL,
        '\u2B5D': MO.REL,
        '\u2B5E': MO.REL,
        '\u2B5F': MO.REL,
        '\u2B60': MO.WIDEREL,
        '\u2B61': MO.RELSTRETCH,
        '\u2B62': MO.WIDEREL,
        '\u2B63': MO.RELSTRETCH,
        '\u2B64': MO.WIDEREL,
        '\u2B65': MO.RELSTRETCH,
        '\u2B66': MO.REL,
        '\u2B67': MO.REL,
        '\u2B68': MO.REL,
        '\u2B69': MO.REL,
        '\u2B6A': MO.WIDEREL,
        '\u2B6B': MO.RELSTRETCH,
        '\u2B6C': MO.WIDEREL,
        '\u2B6D': MO.RELSTRETCH,
        '\u2B6E': MO.REL,
        '\u2B6F': MO.REL,
        '\u2B70': MO.WIDEREL,
        '\u2B71': MO.RELSTRETCH,
        '\u2B72': MO.WIDEREL,
        '\u2B73': MO.RELSTRETCH,
        '\u2B76': MO.REL,
        '\u2B77': MO.REL,
        '\u2B78': MO.REL,
        '\u2B79': MO.REL,
        '\u2B7A': MO.WIDEREL,
        '\u2B7B': MO.RELSTRETCH,
        '\u2B7C': MO.WIDEREL,
        '\u2B7D': MO.RELSTRETCH,
        '\u2B80': MO.WIDEREL,
        '\u2B81': MO.RELSTRETCH,
        '\u2B82': MO.WIDEREL,
        '\u2B83': MO.RELSTRETCH,
        '\u2B84': MO.WIDEREL,
        '\u2B85': MO.RELSTRETCH,
        '\u2B86': MO.WIDEREL,
        '\u2B87': MO.RELSTRETCH,
        '\u2B88': MO.RELACCENT,
        '\u2B89': MO.REL,
        '\u2B8A': MO.RELACCENT,
        '\u2B8B': MO.REL,
        '\u2B8C': MO.REL,
        '\u2B8D': MO.REL,
        '\u2B8E': MO.REL,
        '\u2B8F': MO.REL,
        '\u2B94': MO.REL,
        '\u2B95': MO.WIDEREL,
        '\u2BA0': MO.RELSTRETCH,
        '\u2BA1': MO.RELSTRETCH,
        '\u2BA2': MO.RELSTRETCH,
        '\u2BA3': MO.RELSTRETCH,
        '\u2BA4': MO.RELSTRETCH,
        '\u2BA5': MO.RELSTRETCH,
        '\u2BA6': MO.RELSTRETCH,
        '\u2BA7': MO.RELSTRETCH,
        '\u2BA8': MO.WIDEREL,
        '\u2BA9': MO.WIDEREL,
        '\u2BAA': MO.WIDEREL,
        '\u2BAB': MO.WIDEREL,
        '\u2BAC': MO.RELSTRETCH,
        '\u2BAD': MO.RELSTRETCH,
        '\u2BAE': MO.RELSTRETCH,
        '\u2BAF': MO.RELSTRETCH,
        '\u2BB0': MO.REL,
        '\u2BB1': MO.REL,
        '\u2BB2': MO.REL,
        '\u2BB3': MO.REL,
        '\u2BB4': MO.REL,
        '\u2BB5': MO.REL,
        '\u2BB6': MO.REL,
        '\u2BB7': MO.REL,
        '\u2BB8': MO.RELSTRETCH,
        '\u2BD1': MO.REL,
        '\u3ADC': MO.BIN3,
        '\uFE37': MO.WIDEACCENT,
        '\uFE38': MO.WIDEACCENT,
    },
};
