// @mathjax/src/cjs/input/tex/fontsizev3/FontSizeV3Configuration.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.FontSizeV3Configuration = void 0;
var HandlerTypes_js_1 = require("../HandlerTypes.js");
var TokenMap_js_1 = require("../TokenMap.js");
var Configuration_js_1 = require("../Configuration.js");
var BaseMethods_js_1 = __importDefault(require("../base/BaseMethods.js"));
new TokenMap_js_1.CommandMap('fontsizev3-macros', {
    tiny: [BaseMethods_js_1.default.SetSize, 0.5],
    Tiny: [BaseMethods_js_1.default.SetSize, 0.6],
    scriptsize: [BaseMethods_js_1.default.SetSize, 0.7],
    small: [BaseMethods_js_1.default.SetSize, 0.85],
    normalsize: [BaseMethods_js_1.default.SetSize, 1.0],
    large: [BaseMethods_js_1.default.SetSize, 1.2],
    Large: [BaseMethods_js_1.default.SetSize, 1.44],
    LARGE: [BaseMethods_js_1.default.SetSize, 1.73],
    huge: [BaseMethods_js_1.default.SetSize, 2.07],
    Huge: [BaseMethods_js_1.default.SetSize, 2.49],
});
exports.FontSizeV3Configuration = Configuration_js_1.Configuration.create('fontsizev3', (_a = {},
    _a[HandlerTypes_js_1.ConfigurationType.HANDLER] = (_b = {}, _b[HandlerTypes_js_1.HandlerType.MACRO] = ['fontsizev3-macros'], _b),
    _a));
