// @mathjax/src/mjs/a11y/explorer/KeyExplorer.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { STATE } from '../../core/MathItem.js';
import { AbstractExplorer } from './Explorer.js';
import { honk, SemAttr } from '../speech/SpeechUtil.js';
import { context } from '../../util/context.js';
import { InfoDialog } from '../../ui/dialog/InfoDialog.js';
const isWindows = context.os === 'Windows';
const BRAILLE_PADDING = Array(40).fill('\u2800').join('');
const nav = '[data-speech-node]';
export function isContainer(el) {
    return el.matches('mjx-container');
}
export function hasModifiers(event, shift = true) {
    return ((event.shiftKey && shift) || event.metaKey || event.altKey || event.ctrlKey);
}
export class SpeechExplorer extends AbstractExplorer {
    static helpMessage(title, select, braille) {
        return `
      <h2 role="heading" aria-level="2">Exploring expressions ${title}</h2>

      <p>The mathematics on this page is being rendered by <a
      href="https://www.mathjax.org/" target="_blank">MathJax</a>, which
      generates both the text spoken by screen readers, as well as the
      visual layout for sighted users.</p>

      <p>Expressions typeset by MathJax can be explored interactively, and
      are focusable.  You can use the <kbd>Tab</kbd> key to move to a typeset
      expression${select}.  Initially, the expression will be read in full,
      but you can use the following keys to explore the expression
      further:</p>

      <ul>

      <li><kbd>Down Arrow</kbd> moves one level deeper into the
      expression to allow you to explore the current subexpression term by
      term.</li>

      <li><kbd>Up Arrow</kbd> moves back up a level within the
      expression.</li>

      <li><kbd>Right Arrow</kbd> moves to the next term in the
      current subexpression.</li>

      <li><kbd>Left Arrow</kbd> moves to the next term in the
      current subexpression.</li>

      <li><kbd>Shift</kbd>+<kbd>Arrow</kbd> moves to a
      neighboring cell within a table.</li>

      <li><kbd>0-9</kbd>+<kbd>0-9</kbd> jumps to a cell
      by its index in the table, where 0 = 10.</li>

      <li><kbd>Home</kbd> takes you to the top of the
      expression.</li>

      <li><kbd>Enter</kbd> or <kbd>Return</kbd> clicks a
      link or activates an active subexpression.</li>

      <li><kbd>Space</kbd> opens the MathJax contextual menu
      where you can view or copy the source format of the expression, or
      modify MathJax's settings.</li>

      <li><kbd>Escape</kbd> exits the expression
      explorer.</li>

      <li><kbd>x</kbd> gives a summary of the current
      subexpression.</li>

      <li><kbd>z</kbd> gives the full text of a collapsed
      expression.</li>

      <li><kbd>d</kbd> gives the current depth within the
      expression.</li>

      <li><kbd>s</kbd> starts or stops auto-voicing with
      synchronized highlighting.</li>

      <li><kbd>v</kbd> marks the current position in the
      expression.</li>

      <li><kbd>p</kbd> cycles through the marked positions in
      the expression.</li>

      <li><kbd>u</kbd> clears all marked positions and returns
      to the starting position.</li>

      <li><kbd>&gt;</kbd> cycles through the available speech
      rule sets (MathSpeak, ClearSpeak).</li>

      <li><kbd>&lt;</kbd> cycles through the verbosity levels
      for the current rule set.</li>

      <li><kbd>b</kbd> toggles whether Braille notation is combined
      with speech text for tactile Braille devices, as discussed
      below.

      <li><kbd>h</kbd> produces this help listing.</li>
      </ul>

      <p>The MathJax contextual menu allows you to enable or disable speech
      or Braille generation for mathematical expressions, the language to
      use for the spoken mathematics, and other features of MathJax.  In
      particular, the Explorer submenu allows you to specify how the
      mathematics should be identified in the page (e.g., by saying "math"
      when the expression is spoken), and whether or not to include a
      message about the letter "h" bringing up this dialog box.  Turning off
      speech and Braille will disable the expression explorer, its
      highlighting, and its help icon.</p>

      <p>Support for tactile Braille devices varies across screen readers,
      browsers, and operative systems.  If you are using a Braille output
      device, you may need to select the "Combine with Speech" option in the
      contextual menu's Braille submenu in order to obtain Nemeth or Euro
      Braille output rather than the speech text on your Braille
      device. ${braille}</p>

      <p>The contextual menu also provides options for viewing or copying a
      MathML version of the expression or its original source format,
      creating an SVG version of the expression, and viewing various other
      information.</p>

      <p>Finally, selecting the "Insert Hidden MathML" item from the options
      submenu will turn of MathJax's speech and Braille generation and
      instead use visually hidden MathML that some screen readers can voice,
      though support for this is not universal across all screen readers and
      operating systems.  Selecting speech or Braille generation in their
      submenus will remove the hidden MathML again.</p>

      <p>For more help, see the <a
      href="https://docs.mathjax.org/en/latest/basic/accessibility.html"
      target="_blank">MathJax accessibility documentation.</a></p>
    `;
    }
    get generators() {
        var _a;
        return (_a = this.item) === null || _a === void 0 ? void 0 : _a.generatorPool;
    }
    get role() {
        return this.item.ariaRole;
    }
    get description() {
        return this.item.roleDescription;
    }
    get none() {
        return this.document.options.a11y.brailleSpeech
            ? this.item.brailleNone
            : this.item.none;
    }
    get brailleNone() {
        return this.item.brailleNone;
    }
    FocusIn(event) {
        if (event.target.closest('mjx-html'))
            return;
        if (this.item.outputData.nofocus) {
            this.item.outputData.nofocus = false;
            return;
        }
        if (!this.clicked) {
            this.Start();
            this.backTab = event.target === this.img;
        }
        this.clicked = null;
    }
    FocusOut(_event) {
        if (this.current && !this.focusSpeech) {
            if (!this.document.options.keepRegions) {
                this.setCurrent(null);
                this.Stop();
            }
            if (!document.hasFocus()) {
                this.focusTop();
            }
        }
    }
    KeyDown(event) {
        this.pendingIndex.shift();
        this.region.cancelVoice();
        if (hasModifiers(event, false))
            return;
        const CLASS = this.constructor;
        const key = event.key.length === 1 ? event.key.toLowerCase() : event.key;
        const [action, value] = CLASS.keyMap.get(key) || [];
        const result = action
            ? value === undefined || this.active
                ? action(this, event)
                : value
            : this.undefinedKey(event);
        if (result)
            return;
        this.stopEvent(event);
        if (result === false && this.sound) {
            this.NoMove();
        }
    }
    MouseDown(event) {
        var _a;
        this.pendingIndex = [];
        this.region.cancelVoice();
        if (hasModifiers(event) || event.buttons === 2) {
            this.item.outputData.nofocus = true;
            return;
        }
        const clicked = this.findClicked(event.target, event.x, event.y);
        if (clicked === this.document.infoIcon) {
            this.stopEvent(event);
            return;
        }
        (_a = document.getSelection()) === null || _a === void 0 ? void 0 : _a.removeAllRanges();
        if (event.target.getAttribute('sre-highlighter-added')) {
            this.refocus = clicked;
        }
        else {
            this.clicked = clicked;
        }
    }
    Click(event) {
        if (hasModifiers(event) ||
            event.buttons === 2 ||
            document.getSelection().type === 'Range') {
            this.FocusOut(null);
            return;
        }
        const clicked = this.findClicked(event.target, event.x, event.y);
        if (clicked === this.document.infoIcon) {
            this.stopEvent(event);
            this.help();
            return;
        }
        if (!this.clicked && (!clicked || this.node.contains(clicked))) {
            this.refocus = clicked;
            if (!this.triggerLinkMouse()) {
                this.Start();
            }
        }
    }
    DblClick(event) {
        var _a;
        const direction = (_a = document.getSelection().direction) !== null && _a !== void 0 ? _a : 'none';
        if (hasModifiers(event) || event.buttons === 2 || direction !== 'none') {
            this.FocusOut(null);
        }
        else {
            this.refocus = this.rootNode();
            this.Start();
        }
    }
    spaceKey() {
        this.refocus = this.current;
        return true;
    }
    hKey() {
        if (!this.document.options.enableExplorerHelp) {
            return true;
        }
        this.refocus = this.current;
        this.help();
    }
    escapeKey(event) {
        if (event.target.closest('mjx-html')) {
            this.refocus = event.target.closest(nav);
            this.Start();
        }
        else {
            this.Stop();
            this.focusTop();
            this.setCurrent(null);
        }
        return true;
    }
    tabKey(event) {
        var _a;
        const active = (_a = this.current) !== null && _a !== void 0 ? _a : (this.node.contains(document.activeElement)
            ? document.activeElement
            : null);
        if (this.tabs.length === 0 || !active)
            return true;
        if (this.backTab) {
            if (!event.shiftKey)
                return true;
            this.tabTo(this.tabs[this.tabs.length - 1]);
            return;
        }
        const [tabs, position, current] = event.shiftKey
            ? [
                this.tabs.slice(0).reverse(),
                Node.DOCUMENT_POSITION_PRECEDING,
                this.current && this.isLink() ? this.getAnchor() : active,
            ]
            : [this.tabs, Node.DOCUMENT_POSITION_FOLLOWING, active];
        for (const tab of tabs) {
            if (current.compareDocumentPosition(tab) & position) {
                this.tabTo(tab);
                return;
            }
        }
        if (event.shiftKey && this.current === this.rootNode()) {
            this.tabOut();
        }
        return true;
    }
    tabTo(node) {
        if (node.getAttribute('data-mjx-href')) {
            this.setCurrent(this.linkFor(node));
        }
        else {
            node.focus();
        }
    }
    tabOut() {
        const html = Array.from(this.node.querySelectorAll('mjx-html'));
        if (html.length) {
            html.forEach((node) => {
                node.style.display = 'none';
            });
            setTimeout(() => {
                html.forEach((node) => {
                    node.style.display = '';
                });
            }, 0);
        }
    }
    enterKey(event) {
        if (this.active) {
            if (this.triggerLinkKeyboard(event)) {
                this.Stop();
            }
            else {
                const expandable = this.actionable(this.current);
                if (expandable) {
                    this.refocus = expandable;
                    expandable.dispatchEvent(new Event('click'));
                    return;
                }
                const tabs = this.getInternalTabs(this.current).filter((node) => !node.getAttribute('data-mjx-href'));
                if (tabs.length) {
                    tabs[0].focus();
                    return;
                }
            }
        }
        else {
            this.Start();
        }
    }
    homeKey() {
        this.setCurrent(this.rootNode());
    }
    moveDown(shift) {
        return shift
            ? this.moveToNeighborCell(1, 0)
            : this.moveTo(this.firstNode(this.current));
    }
    moveUp(shift) {
        return shift
            ? this.moveToNeighborCell(-1, 0)
            : this.moveTo(this.getParent(this.current));
    }
    moveRight(shift) {
        return shift
            ? this.moveToNeighborCell(0, 1)
            : this.moveTo(this.nextSibling(this.current));
    }
    moveLeft(shift) {
        return shift
            ? this.moveToNeighborCell(0, -1)
            : this.moveTo(this.prevSibling(this.current));
    }
    moveTo(node) {
        if (!node)
            return false;
        this.setCurrent(node);
    }
    moveToNeighborCell(di, dj) {
        const cell = this.tableCell(this.current);
        if (!cell)
            return false;
        const [i, j] = this.cellPosition(cell);
        if (i == null)
            return false;
        const move = this.cellAt(this.cellTable(cell), i + di, j + dj);
        if (!move)
            return false;
        this.setCurrent(move);
    }
    undefinedKey(event) {
        return !this.active || hasModifiers(event);
    }
    addMark() {
        if (this.current === this.marks[this.marks.length - 1]) {
            this.setCurrent(this.current);
        }
        else {
            this.currentMark = this.marks.length - 1;
            this.marks.push(this.current);
            this.speak('Position marked');
        }
    }
    prevMark() {
        if (this.currentMark < 0) {
            if (this.marks.length === 0) {
                this.setCurrent(this.lastMark || this.rootNode());
                return;
            }
            this.currentMark = this.marks.length - 1;
        }
        const current = this.currentMark;
        this.setCurrent(this.marks[current]);
        this.currentMark = current - 1;
    }
    clearMarks() {
        this.marks = [];
        this.currentMark = -1;
        this.prevMark();
    }
    autoVoice() {
        const value = !this.document.options.a11y.voicing;
        if (this.document.menu) {
            this.document.menu.menu.pool.lookup('voicing').setValue(value);
        }
        else {
            this.document.options.a11y.voicing = value;
        }
        this.Update();
    }
    toggleBraille() {
        const value = !this.document.options.a11y.brailleCombine;
        if (this.document.menu) {
            this.document.menu.menu.pool.lookup('brailleCombine').setValue(value);
        }
        else {
            this.document.options.a11y.brailleCombine = value;
        }
    }
    numberKey(n) {
        if (!this.tableCell(this.current))
            return false;
        if (n === 0) {
            n = 10;
        }
        if (this.pendingIndex.length) {
            const table = this.cellTable(this.tableCell(this.current));
            const cell = this.cellAt(table, this.pendingIndex[0] - 1, n - 1);
            this.pendingIndex = [];
            this.speak(String(n));
            if (!cell)
                return false;
            setTimeout(() => this.setCurrent(cell), 500);
        }
        else {
            this.pendingIndex = [null, n];
            this.speak(`Jump to row ${n} and column`);
        }
    }
    depth() {
        var _a, _b, _c;
        if (this.speechType === 'd') {
            this.setCurrent(this.current);
            return;
        }
        this.speechType = 'd';
        const parts = [
            [
                (_a = this.node.getAttribute('data-semantic-level')) !== null && _a !== void 0 ? _a : 'Level',
                (_b = this.current.getAttribute('data-semantic-level-number')) !== null && _b !== void 0 ? _b : '0',
            ]
                .join(' ')
                .trim(),
        ];
        const action = this.actionable(this.current);
        if (action) {
            parts.unshift((_c = this.node.getAttribute(action.getAttribute('toggle') === '1'
                ? 'data-semantic-expandable'
                : 'data-semantic-collapsible')) !== null && _c !== void 0 ? _c : '');
        }
        this.speak(parts.join(' '), this.current.getAttribute(SemAttr.BRAILLE));
    }
    summary() {
        if (this.speechType === 'x') {
            this.setCurrent(this.current);
            return;
        }
        this.speechType = 'x';
        const summary = this.current.getAttribute(SemAttr.SUMMARY);
        this.speak(summary, this.current.getAttribute(SemAttr.BRAILLE), this.SsmlAttributes(this.current, SemAttr.SUMMARY_SSML));
    }
    nextRules() {
        this.node.removeAttribute('data-speech-attached');
        this.restartAfter(this.generators.nextRules(this.item));
    }
    nextStyle() {
        this.node.removeAttribute('data-speech-attached');
        this.restartAfter(this.generators.nextStyle(this.current, this.item));
    }
    details() {
        const action = this.actionable(this.current);
        if (!action ||
            !action.getAttribute('data-collapsible') ||
            action.getAttribute('toggle') !== '1' ||
            this.speechType === 'z') {
            this.setCurrent(this.current);
            return;
        }
        this.speechType = 'z';
        const id = this.nodeId(this.current);
        let current;
        this.item.root.walkTree((node) => {
            if (node.attributes.get('data-semantic-id') === id) {
                current = node;
            }
        });
        let mml = this.item.toMathML(current, this.item);
        if (!current.isKind('math')) {
            mml = `<math>${mml}</math>`;
        }
        mml = mml.replace(/ (?:data-semantic-|aria-|data-speech-|data-latex).*?=".*?"/g, '');
        this.item
            .speechFor(mml)
            .then(([speech, braille]) => this.speak(speech, braille));
    }
    help() {
        if (!this.document.options.enableExplorerHelp) {
            return;
        }
        const CLASS = this.constructor;
        const [title, select, braille] = CLASS.helpData.get(context.os);
        InfoDialog.post({
            title: 'MathJax Expression Explorer Help',
            message: CLASS.helpMessage(title, select, braille),
            node: this.node,
            adaptor: this.document.adaptor,
            styles: {
                '.mjx-dialog': {
                    'max-height': 'calc(min(35em, 90%))',
                },
                'mjx-dialog mjx-title': {
                    'font-size': '133%',
                    margin: '.5em 1.75em',
                },
                'mjx-dialog h2': {
                    'font-size': '20px',
                    margin: '.5em 0',
                },
                'mjx-dialog ul': {
                    'list-style-type': 'none',
                },
                'mjx-dialog li': {
                    'margin-bottom': '.5em',
                },
            },
        });
    }
    setCurrent(node, addDescription = false) {
        this.backTab = false;
        this.speechType = '';
        if (!document.hasFocus()) {
            this.refocus = this.current;
        }
        this.node.setAttribute('aria-busy', 'true');
        if (this.current) {
            this.pool.unhighlight();
            for (const part of Array.from(this.node.querySelectorAll('.mjx-selected'))) {
                part.classList.remove('mjx-selected');
            }
            if (this.document.options.a11y.tabSelects === 'last') {
                this.refocus = this.current;
            }
            if (!node) {
                this.lastMark = this.current;
                this.removeSpeech();
            }
            this.current = null;
        }
        this.current = node;
        this.currentMark = -1;
        if (this.current) {
            const parts = [...this.getSplitNodes(this.current)];
            this.highlighter.encloseNodes(parts, this.node);
            for (const part of parts) {
                if (!part.getAttribute('data-sre-enclosed')) {
                    part.classList.add('mjx-selected');
                }
            }
            this.pool.highlight(parts);
            this.addSpeech(node, addDescription);
            this.node.setAttribute('tabindex', '-1');
            this.Update();
        }
        this.node.removeAttribute('aria-busy');
    }
    getSplitNodes(node) {
        const id = this.nodeId(node);
        if (!id) {
            return [node];
        }
        if (this.cacheParts.has(id)) {
            return this.cacheParts.get(id);
        }
        const parts = Array.from(this.node.querySelectorAll(`[data-semantic-id="${id}"]`));
        const subtree = this.subtree(id, parts);
        this.cacheParts.set(id, [...parts, ...subtree]);
        return this.cacheParts.get(id);
    }
    subtree(id, nodes) {
        const sub = this.subtrees.get(id);
        const children = new Set();
        for (const node of nodes) {
            Array.from(node.querySelectorAll(`[data-semantic-id]`)).forEach((x) => children.add(this.nodeId(x)));
        }
        const rest = setdifference(sub, children);
        return [...rest]
            .map((child) => this.getNode(child))
            .filter((node) => node !== null);
    }
    addSpeech(node, describe) {
        var _a;
        if (!this.document.options.enableSpeech &&
            !this.document.options.enableBraille) {
            return;
        }
        if (this.anchors.length) {
            setTimeout(() => { var _a; return (_a = this.img) === null || _a === void 0 ? void 0 : _a.remove(); }, 10);
        }
        else {
            (_a = this.img) === null || _a === void 0 ? void 0 : _a.remove();
        }
        let speech = this.addComma([
            node.getAttribute(SemAttr.PREFIX),
            node.getAttribute(SemAttr.SPEECH),
            node.getAttribute(SemAttr.POSTFIX),
        ])
            .join(' ')
            .trim();
        if (describe) {
            let description = this.description === this.none ? '' : ', ' + this.description;
            if (this.document.options.a11y.help &&
                this.document.options.enableExplorerHelp) {
                description += ', press h for help';
            }
            speech += description;
        }
        this.speak(speech, node.getAttribute(SemAttr.BRAILLE), this.SsmlAttributes(node, SemAttr.SPEECH_SSML));
    }
    addComma(words) {
        if (words[2] && (words[1] || words[0])) {
            words[1] += ',';
        }
        return words;
    }
    removeSpeech() {
        if (this.speech) {
            this.unspeak(this.speech);
            this.speech = null;
            if (this.img) {
                this.node.append(this.img);
            }
            this.node.setAttribute('tabindex', this.tabIndex);
        }
    }
    speak(speech, braille = '', ssml = null, description = this.none) {
        const oldspeech = this.speech;
        const speechNode = (this.speech = document.createElement('mjx-speech'));
        speechNode.setAttribute('role', this.role);
        speechNode.setAttribute('aria-label', speech || this.none);
        speechNode.setAttribute('aria-roledescription', description || this.none);
        speechNode.setAttribute(SemAttr.SPEECH, speech);
        if (ssml) {
            speechNode.setAttribute(SemAttr.PREFIX_SSML, ssml[0] || '');
            speechNode.setAttribute(SemAttr.SPEECH_SSML, ssml[1] || '');
            speechNode.setAttribute(SemAttr.POSTFIX_SSML, ssml[2] || '');
        }
        if (braille) {
            if (this.document.options.a11y.brailleSpeech) {
                speechNode.setAttribute('aria-label', braille);
                speechNode.setAttribute('aria-roledescription', this.brailleNone);
            }
            speechNode.setAttribute('aria-braillelabel', braille);
            speechNode.setAttribute('aria-brailleroledescription', this.brailleNone);
            if (this.document.options.a11y.brailleCombine) {
                speechNode.setAttribute('aria-label', braille + BRAILLE_PADDING + speech);
            }
        }
        speechNode.setAttribute('tabindex', '0');
        if (isWindows) {
            const container = document.createElement('mjx-speech-container');
            container.setAttribute('role', 'application');
            container.setAttribute('aria-roledescription', this.none);
            container.setAttribute('aria-brailleroledescription', this.brailleNone);
            container.append(speechNode);
            this.node.append(container);
            speechNode.setAttribute('role', 'img');
        }
        else {
            this.node.append(speechNode);
        }
        this.focusSpeech = true;
        speechNode.focus();
        this.focusSpeech = false;
        this.Update();
        if (oldspeech) {
            setTimeout(() => this.unspeak(oldspeech), 100);
        }
    }
    unspeak(node) {
        if (isWindows) {
            node = node.parentElement;
        }
        node.remove();
    }
    attachSpeech() {
        var _a;
        const item = this.item;
        const container = this.node;
        if (!container.hasAttribute('has-speech')) {
            for (const child of Array.from(container.childNodes)) {
                child.setAttribute('aria-hidden', 'true');
            }
            container.setAttribute('has-speech', 'true');
        }
        const description = item.roleDescription;
        const speech = (container.getAttribute(SemAttr.SPEECH) || '') +
            (description ? ', ' + description : '');
        (_a = this.img) === null || _a === void 0 ? void 0 : _a.remove();
        this.img = this.document.adaptor.node('mjx-speech', {
            'aria-label': speech,
            role: 'img',
            'aria-roledescription': item.none,
        });
        const braille = container.getAttribute(SemAttr.BRAILLE);
        if (braille) {
            if (this.document.options.a11y.brailleSpeech) {
                this.img.setAttribute('aria-label', braille);
                this.img.setAttribute('aria-roledescription', this.brailleNone);
            }
            this.img.setAttribute('aria-braillelabel', braille);
            this.img.setAttribute('aria-brailleroledescription', this.brailleNone);
            if (this.document.options.a11y.brailleCombine) {
                this.img.setAttribute('aria-label', braille + BRAILLE_PADDING + speech);
            }
        }
        container.appendChild(this.img);
        this.adjustAnchors();
        this.getTabs();
    }
    detachSpeech() {
        var _a;
        const container = this.node;
        (_a = this.img) === null || _a === void 0 ? void 0 : _a.remove();
        container.removeAttribute('has-speech');
        for (const child of Array.from(container.childNodes)) {
            child.removeAttribute('aria-hidden');
        }
        this.restoreAnchors();
    }
    adjustAnchors() {
        this.anchors = Array.from(this.node.querySelectorAll('a[href]'));
        for (const anchor of this.anchors) {
            const href = anchor.getAttribute('href');
            anchor.setAttribute('data-mjx-href', href);
            anchor.removeAttribute('href');
        }
        if (this.anchors.length) {
            this.img.setAttribute('tabindex', '0');
        }
    }
    restoreAnchors() {
        for (const anchor of this.anchors) {
            anchor.setAttribute('href', anchor.getAttribute('data-mjx-href'));
            anchor.removeAttribute('data-mjx-href');
        }
        this.anchors = [];
    }
    getTabs() {
        this.tabs = this.getInternalTabs(this.node);
    }
    getInternalTabs(node) {
        return Array.from(node.querySelectorAll('button, [data-mjx-href], input, select, textarea, [tabindex]:not([tabindex="-1"],mjx-speech)'));
    }
    focus() {
        this.node.focus();
    }
    nodeId(node) {
        return node.getAttribute('data-semantic-id');
    }
    parentId(node) {
        return node.getAttribute('data-semantic-parent');
    }
    getNode(id) {
        return id ? this.node.querySelector(`[data-semantic-id="${id}"]`) : null;
    }
    getParent(node) {
        return this.getNode(this.parentId(node));
    }
    childArray(node) {
        return node ? node.getAttribute('data-semantic-children').split(/,/) : [];
    }
    isCell(node) {
        return (!!node && this.cellTypes.includes(node.getAttribute('data-semantic-type')));
    }
    isRow(node) {
        return !!node && node.getAttribute('data-semantic-type') === 'row';
    }
    tableCell(node) {
        while (node && node !== this.node) {
            if (this.isCell(node)) {
                return node;
            }
            node = node.parentNode;
        }
        return null;
    }
    cellTable(cell) {
        const row = this.getParent(cell);
        return this.isRow(row) ? this.getParent(row) : row;
    }
    cellPosition(cell) {
        const row = this.getParent(cell);
        const j = this.childArray(row).indexOf(this.nodeId(cell));
        if (!this.isRow(row)) {
            return [j, 1];
        }
        const table = this.getParent(row);
        const i = this.childArray(table).indexOf(this.nodeId(row));
        return [i, j];
    }
    cellAt(table, i, j) {
        const row = this.getNode(this.childArray(table)[i]);
        if (!this.isRow(row)) {
            return j === 1 ? row : null;
        }
        const cell = this.getNode(this.childArray(row)[j]);
        return cell;
    }
    firstNode(node) {
        const owns = node.getAttribute('data-semantic-owns');
        if (!owns) {
            return node.querySelector(nav);
        }
        const ownsList = owns.split(/ /);
        for (const id of ownsList) {
            const node = this.getNode(id);
            if (node === null || node === void 0 ? void 0 : node.hasAttribute('data-speech-node')) {
                return node;
            }
        }
        return node.querySelector(nav);
    }
    rootNode() {
        const base = this.node.querySelector('[data-semantic-structure]');
        if (!base) {
            return this.node.querySelector(nav);
        }
        const id = base
            .getAttribute('data-semantic-structure')
            .split(/ /)[0]
            .replace('(', '');
        return this.getNode(id);
    }
    nextSibling(node) {
        var _a;
        const id = this.parentId(node);
        if (!id)
            return null;
        const owns = (_a = this.getNode(id)
            .getAttribute('data-semantic-owns')) === null || _a === void 0 ? void 0 : _a.split(/ /);
        if (!owns)
            return null;
        let i = owns.indexOf(this.nodeId(node));
        let next;
        do {
            next = this.getNode(owns[++i]);
        } while (next && !next.hasAttribute('data-speech-node'));
        return next;
    }
    prevSibling(node) {
        var _a;
        const id = this.parentId(node);
        if (!id)
            return null;
        const owns = (_a = this.getNode(id)
            .getAttribute('data-semantic-owns')) === null || _a === void 0 ? void 0 : _a.split(/ /);
        if (!owns)
            return null;
        let i = owns.indexOf(this.nodeId(node));
        let prev;
        do {
            prev = this.getNode(owns[--i]);
        } while (prev && !prev.hasAttribute('data-speech-node'));
        return prev;
    }
    findClicked(node, x, y) {
        const icon = this.document.infoIcon;
        if (icon === node || icon.contains(node)) {
            return icon;
        }
        if (this.node.getAttribute('jax') !== 'SVG') {
            return node.closest(nav);
        }
        let found = null;
        let clicked = this.node;
        while (clicked) {
            if (clicked.matches(nav)) {
                found = clicked;
            }
            const nodes = Array.from(clicked.childNodes);
            clicked = null;
            for (const child of nodes) {
                if (child !== this.speech &&
                    child !== this.img &&
                    child.tagName &&
                    child.tagName.toLowerCase() !== 'rect') {
                    const { left, right, top, bottom } = child.getBoundingClientRect();
                    if (left <= x && x <= right && top <= y && y <= bottom) {
                        clicked = child;
                        break;
                    }
                }
            }
        }
        return found;
    }
    isLink(node = this.current) {
        var _a;
        return !!((_a = node === null || node === void 0 ? void 0 : node.getAttribute('data-semantic-attributes')) === null || _a === void 0 ? void 0 : _a.includes('href:'));
    }
    getAnchor(node = this.current) {
        const anchor = node.closest('a');
        return anchor && this.node.contains(anchor) ? anchor : null;
    }
    linkFor(anchor) {
        return anchor === null || anchor === void 0 ? void 0 : anchor.querySelector('[data-semantic-attributes*="href:"]');
    }
    parentLink(node) {
        const link = node === null || node === void 0 ? void 0 : node.closest('[data-semantic-attributes*="href:"]');
        return link && this.node.contains(link) ? link : null;
    }
    focusTop() {
        this.focusSpeech = true;
        this.node.focus();
        this.focusSpeech = false;
    }
    SsmlAttributes(node, center) {
        return [
            node.getAttribute(SemAttr.PREFIX_SSML),
            node.getAttribute(center),
            node.getAttribute(SemAttr.POSTFIX_SSML),
        ];
    }
    restartAfter(promise) {
        return __awaiter(this, void 0, void 0, function* () {
            yield promise;
            this.attachSpeech();
            const current = this.current;
            this.current = null;
            this.pool.unhighlight();
            this.setCurrent(current);
        });
    }
    constructor(document, pool, region, node, brailleRegion, magnifyRegion, _mml, item) {
        super(document, pool, null, node);
        this.document = document;
        this.pool = pool;
        this.region = region;
        this.node = node;
        this.brailleRegion = brailleRegion;
        this.magnifyRegion = magnifyRegion;
        this.item = item;
        this.sound = false;
        this.current = null;
        this.clicked = null;
        this.refocus = null;
        this.focusSpeech = false;
        this.restarted = null;
        this.speech = null;
        this.speechType = '';
        this.img = null;
        this.attached = false;
        this.eventsAttached = false;
        this.marks = [];
        this.currentMark = -1;
        this.lastMark = null;
        this.pendingIndex = [];
        this.cellTypes = ['cell', 'line'];
        this.backTab = false;
        this.events = super.Events().concat([
            ['focusin', this.FocusIn.bind(this)],
            ['focusout', this.FocusOut.bind(this)],
            ['keydown', this.KeyDown.bind(this)],
            ['mousedown', this.MouseDown.bind(this)],
            ['click', this.Click.bind(this)],
            ['dblclick', this.DblClick.bind(this)],
        ]);
        this.subtrees = null;
        this.cacheParts = new Map();
    }
    findStartNode() {
        let node = this.refocus || this.current;
        if (!node && this.restarted) {
            node = this.node.querySelector(this.restarted);
        }
        this.refocus = this.restarted = null;
        return node;
    }
    Start() {
        const _super = Object.create(null, {
            Start: { get: () => super.Start }
        });
        return __awaiter(this, void 0, void 0, function* () {
            if (!this.subtrees) {
                this.subtrees = new Map();
                this.getSubtrees();
            }
            if (!this.attached || this.active)
                return;
            this.document.activeItem = this.item;
            if (this.item.state() < STATE.ATTACHSPEECH) {
                this.item.attachSpeech(this.document);
                yield this.generators.promise;
            }
            if (this.focusSpeech)
                return;
            this.node.classList.add('mjx-explorer-active');
            if (this.document.options.enableExplorerHelp) {
                this.node.append(this.document.infoIcon);
            }
            const node = this.findStartNode();
            this.setCurrent(node || this.rootNode(), !node);
            _super.Start.call(this);
            const options = this.document.options;
            const a11y = options.a11y;
            if (a11y.subtitles && a11y.speech && options.enableSpeech) {
                this.region.Show(this.node);
            }
            if (a11y.viewBraille && a11y.braille && options.enableBraille) {
                this.brailleRegion.Show(this.node);
            }
            if (a11y.keyMagnifier) {
                this.magnifyRegion.Show(this.current);
            }
            this.Update();
        });
    }
    Stop() {
        if (this.active) {
            this.node.classList.remove('mjx-explorer-active');
            if (this.document.options.enableExplorerHelp) {
                this.document.infoIcon.remove();
            }
            this.pool.unhighlight();
            this.magnifyRegion.Hide();
            this.region.Hide();
            this.brailleRegion.Hide();
        }
        super.Stop();
    }
    Update() {
        if (!this.active)
            return;
        this.region.node = this.node;
        if (this.document.options.enableSpeech ||
            this.document.options.enableBraille) {
            this.generators.updateRegions(this.speech || this.node, this.region, this.brailleRegion);
        }
        this.magnifyRegion.Update(this.current);
    }
    get tabIndex() {
        return this.document.options.a11y.inTabOrder ? '0' : '-1';
    }
    Attach() {
        if (this.attached)
            return;
        super.Attach();
        this.node.setAttribute('tabindex', this.tabIndex);
        this.attached = true;
    }
    Detach() {
        var _a;
        super.RemoveEvents();
        this.node.removeAttribute('role');
        this.node.removeAttribute('aria-roledescription');
        this.node.removeAttribute('aria-label');
        (_a = this.img) === null || _a === void 0 ? void 0 : _a.remove();
        if (this.active) {
            this.node.setAttribute('tabindex', this.tabIndex);
        }
        this.attached = false;
    }
    NoMove() {
        honk();
    }
    AddEvents() {
        if (!this.eventsAttached) {
            super.AddEvents();
            this.addHtmlEvents();
            this.eventsAttached = true;
        }
    }
    addHtmlEvents() {
        for (const html of Array.from(this.node.querySelectorAll('mjx-html'))) {
            const stop = (event) => {
                if (html.contains(document.activeElement)) {
                    if (event instanceof KeyboardEvent) {
                        this.clicked = null;
                        if (event.key !== 'Tab' && event.key !== 'Escape') {
                            event.stopPropagation();
                        }
                    }
                    else {
                        this.clicked = event.target;
                    }
                }
            };
            html.addEventListener('mousedown', stop);
            html.addEventListener('click', stop);
            html.addEventListener('keydown', stop);
            html.addEventListener('dblclick', stop);
        }
    }
    actionable(node) {
        const parent = node === null || node === void 0 ? void 0 : node.parentNode;
        return parent && this.highlighter.isMactionNode(parent) ? parent : null;
    }
    triggerLinkKeyboard(event) {
        if (!this.current) {
            if (event.target instanceof HTMLAnchorElement) {
                event.target.dispatchEvent(new MouseEvent('click'));
                return true;
            }
            return false;
        }
        return this.triggerLink(this.current);
    }
    triggerLink(node) {
        if (this.isLink(node)) {
            const anchor = this.getAnchor(node);
            anchor.classList.add('mjx-visited');
            setTimeout(() => this.FocusOut(null), 50);
            window.location.href = anchor.getAttribute('data-mjx-href');
            return true;
        }
        return false;
    }
    triggerLinkMouse() {
        const link = this.parentLink(this.refocus);
        if (this.triggerLink(link)) {
            return true;
        }
        return false;
    }
    semanticFocus() {
        const focus = [];
        let name = 'data-semantic-id';
        let node = this.current || this.refocus || this.node;
        const action = this.actionable(node);
        if (action) {
            name = action.hasAttribute('data-maction-id') ? 'data-maction-id' : 'id';
            node = action;
            focus.push(nav);
        }
        const attr = node.getAttribute(name);
        if (attr) {
            focus.unshift(`[${name}="${attr}"]`);
        }
        return focus.join(' ');
    }
    getSubtrees() {
        const node = this.node.querySelector('[data-semantic-structure]');
        if (!node)
            return;
        const sexp = node.getAttribute('data-semantic-structure');
        const tokens = tokenize(sexp);
        const tree = parse(tokens);
        buildMap(tree, this.subtrees);
    }
}
SpeechExplorer.helpData = new Map([
    [
        'MacOS',
        [
            'on MacOS and iOS using VoiceOver',
            ', or the VoiceOver arrow keys to select an expression',
            '',
        ],
    ],
    [
        'Windows',
        [
            'in Windows using NVDA or JAWS',
            `. The screen reader should enter focus or forms mode automatically
        when the expression gets the browser focus, but if not, you can toggle
        focus mode using NVDA+space in NVDA; for JAWS, Enter should start
        forms mode while Numpad Plus leaves it.  Also note that you can use
        the NVDA or JAWS key plus the arrow keys to explore the expression
        even in browse mode, and you can use NVDA+shift+arrow keys to
        navigate out of an expression that has the focus in NVDA`,
            `NVDA users need to select this option, while JAWS users should be able
        to get Braille output without changing this setting.`,
        ],
    ],
    [
        'Unix',
        [
            'in Unix using Orca',
            `, and Orca should enter focus mode automatically.  If not, use the
        Orca+a key to toggle focus mode on or off.  Also note that you can use
        Orca+arrow keys to explore expressions even in browse mode`,
            '',
        ],
    ],
    ['unknown', ['with a Screen Reader.', '', '']],
]);
SpeechExplorer.keyMap = new Map([
    ['Tab', [(explorer, event) => explorer.tabKey(event)]],
    ['Escape', [(explorer, event) => explorer.escapeKey(event)]],
    ['Enter', [(explorer, event) => explorer.enterKey(event)]],
    ['Home', [(explorer) => explorer.homeKey()]],
    [
        'ArrowDown',
        [(explorer, event) => explorer.moveDown(event.shiftKey), true],
    ],
    ['ArrowUp', [(explorer, event) => explorer.moveUp(event.shiftKey), true]],
    [
        'ArrowLeft',
        [(explorer, event) => explorer.moveLeft(event.shiftKey), true],
    ],
    [
        'ArrowRight',
        [(explorer, event) => explorer.moveRight(event.shiftKey), true],
    ],
    [' ', [(explorer) => explorer.spaceKey()]],
    ['h', [(explorer) => explorer.hKey()]],
    ['>', [(explorer) => explorer.nextRules(), false]],
    ['<', [(explorer) => explorer.nextStyle(), false]],
    ['x', [(explorer) => explorer.summary(), false]],
    ['z', [(explorer) => explorer.details(), false]],
    ['d', [(explorer) => explorer.depth(), false]],
    ['v', [(explorer) => explorer.addMark(), false]],
    ['p', [(explorer) => explorer.prevMark(), false]],
    ['u', [(explorer) => explorer.clearMarks(), false]],
    ['s', [(explorer) => explorer.autoVoice(), false]],
    ['b', [(explorer) => explorer.toggleBraille(), false]],
    ...[...'0123456789'].map((n) => [
        n,
        [(explorer) => explorer.numberKey(parseInt(n)), false],
    ]),
]);
function tokenize(str) {
    return str.replace(/\(/g, ' ( ').replace(/\)/g, ' ) ').trim().split(/\s+/);
}
function parse(tokens) {
    const stack = [[]];
    for (const token of tokens) {
        if (token === '(') {
            const newNode = [];
            stack[stack.length - 1].push(newNode);
            stack.push(newNode);
        }
        else if (token === ')') {
            stack.pop();
        }
        else {
            stack[stack.length - 1].push(token);
        }
    }
    return stack[0][0];
}
function buildMap(tree, map) {
    if (typeof tree === 'string') {
        if (!map.has(tree))
            map.set(tree, new Set());
        return new Set();
    }
    const [root, ...children] = tree;
    const rootId = root;
    const descendants = new Set();
    for (const child of children) {
        const childRoot = typeof child === 'string' ? child : child[0];
        const childDescendants = buildMap(child, map);
        descendants.add(childRoot);
        childDescendants.forEach((d) => descendants.add(d));
    }
    map.set(rootId, descendants);
    return descendants;
}
function setdifference(a, b) {
    if (!a) {
        return new Set();
    }
    if (!b) {
        return a;
    }
    return new Set([...a].filter((x) => !b.has(x)));
}
