// @mathjax/src/mjs/a11y/explorer/Region.js
import { StyleJsonSheet } from '../../util/StyleJson.js';
import { buildSpeech } from '../speech/SpeechUtil.js';
export class AbstractRegion {
    constructor(document) {
        this.document = document;
        this.CLASS = this.constructor;
        this.AddStyles();
    }
    static get sheetId() {
        return 'MJX-' + this.name + '-styles';
    }
    static get styleSheet() {
        return document.head.querySelector('#' + this.sheetId);
    }
    AddStyles() {
        const id = this.CLASS.sheetId;
        if (!this.CLASS.style ||
            this.document.adaptor.head().querySelector('#' + id)) {
            return;
        }
        const node = this.document.adaptor.node('style', { id });
        node.innerHTML = this.CLASS.style.cssText;
        this.document.adaptor.head().appendChild(node);
    }
    AddElement() {
        if (this.div)
            return;
        const element = this.document.adaptor.node('div');
        element.classList.add(this.CLASS.className);
        this.div = element;
        this.inner = this.document.adaptor.node('div');
        this.div.appendChild(this.inner);
        this.document.adaptor
            .body(this.document.adaptor.document)
            .appendChild(this.div);
    }
    Show(node) {
        this.AddElement();
        this.position(node);
        this.div.classList.add(this.CLASS.className + '_Show');
    }
    Hide() {
        if (!this.div)
            return;
        this.div.remove();
        this.div = null;
        this.inner = null;
    }
    stackRegions(node) {
        const rect = node.getBoundingClientRect();
        let baseBottom = 0;
        let baseLeft = Number.POSITIVE_INFINITY;
        const regions = this.document.adaptor.document.getElementsByClassName(this.CLASS.className + '_Show');
        for (let i = 0, region; (region = regions[i]); i++) {
            if (region !== this.div) {
                baseBottom = Math.max(region.getBoundingClientRect().bottom, baseBottom);
                baseLeft = Math.min(region.getBoundingClientRect().left, baseLeft);
            }
        }
        const bot = (baseBottom ? baseBottom : rect.bottom + 10) + window.scrollY;
        const left = (baseLeft < Number.POSITIVE_INFINITY ? baseLeft : rect.left) +
            window.scrollX;
        this.div.style.top = bot + 'px';
        this.div.style.left = left + 'px';
    }
}
export class DummyRegion extends AbstractRegion {
    Clear() { }
    Update() { }
    Hide() { }
    Show() { }
    AddElement() { }
    AddStyles() { }
    position() { }
}
export class StringRegion extends AbstractRegion {
    Clear() {
        if (!this.div)
            return;
        this.Update('');
        this.inner.style.top = '';
        this.inner.style.backgroundColor = '';
    }
    Update(speech) {
        if (speech) {
            this.AddElement();
        }
        if (this.inner) {
            this.inner.textContent = '';
            this.inner.textContent = speech || '\u00a0';
        }
    }
    position(node) {
        this.stackRegions(node);
    }
}
export class ToolTip extends StringRegion {
}
ToolTip.className = 'MJX_ToolTip';
ToolTip.style = new StyleJsonSheet({
    [`.${ToolTip.className}`]: {
        width: 'auto',
        height: 'auto',
        opacity: 1,
        'text-align': 'center',
        'border-radius': '4px',
        padding: 0,
        'border-bottom': '1px dotted black',
        position: 'absolute',
        display: 'inline-block',
        'background-color': 'white',
        'z-index': 202,
    },
    [`.${ToolTip.className} > div`]: {
        'border-radius': 'inherit',
        padding: '0 2px',
    },
    '@media (prefers-color-scheme: dark)': {
        ['.' + ToolTip.className]: {
            'background-color': '#222025',
            'box-shadow': '0px 5px 20px #000',
            border: '1px solid #7C7C7C',
        },
    },
});
export class LiveRegion extends StringRegion {
    static setColor(type, priority, color, opacity) {
        const style = this.styleSheet;
        if (style) {
            const css = style.sheet.cssRules[0].style;
            const alpha = opacity === 1 ? 1 : `var(--mjx-${type}${priority}-alpha)`;
            const name = `--mjx-${type}${priority}-color`;
            const value = `rgba(var(--mjx-${type}-${color}), ${alpha})`;
            if (css.getPropertyValue(name) !== value) {
                css.setProperty(name, value);
            }
            const oname = `--mjx-${type}${priority}-alpha`;
            if (css.getPropertyValue(oname) !== String(opacity)) {
                css.setProperty(oname, opacity);
                style.sheet.cssRules[1].cssRules[0].style.setProperty(oname, Math.pow(opacity, 0.7071));
            }
        }
    }
}
LiveRegion.className = 'MJX_LiveRegion';
LiveRegion.priority = {
    primary: 1,
    secondary: 2,
};
LiveRegion.style = new StyleJsonSheet({
    ':root': {
        '--mjx-fg-red': '255, 0, 0',
        '--mjx-fg-green': '0, 255, 0',
        '--mjx-fg-blue': '0, 0, 255',
        '--mjx-fg-yellow': '255, 255, 0',
        '--mjx-fg-cyan': '0, 255, 255',
        '--mjx-fg-magenta': '255, 0, 255',
        '--mjx-fg-white': '255, 255, 255',
        '--mjx-fg-black': '0, 0, 0',
        '--mjx-bg-red': '255, 0, 0',
        '--mjx-bg-green': '0, 255, 0',
        '--mjx-bg-blue': '0, 0, 255',
        '--mjx-bg-yellow': '255, 255, 0',
        '--mjx-bg-cyan': '0, 255, 255',
        '--mjx-bg-magenta': '255, 0, 255',
        '--mjx-bg-white': '255, 255, 255',
        '--mjx-bg-black': '0, 0, 0',
        '--mjx-live-bg-color': 'white',
        '--mjx-live-shadow-color': '#888',
        '--mjx-live-border-color': '#CCCCCC',
        '--mjx-bg1-color': 'rgba(var(--mjx-bg-blue), var(--mjx-bg1-alpha))',
        '--mjx-fg1-color': 'rgba(var(--mjx-fg-black), 1)',
        '--mjx-bg2-color': 'rgba(var(--mjx-bg-red), 1)',
        '--mjx-fg2-color': 'rgba(var(--mjx-fg-black), 1)',
        '--mjx-bg1-alpha': 0.2,
        '--mjx-fg1-alpha': 1,
        '--mjx-bg2-alpha': 1,
        '--mjx-fg2-alpha': 1,
    },
    '@media (prefers-color-scheme: dark)': {
        ':root': {
            '--mjx-bg-blue': '132, 132, 255',
            '--mjx-bg-white': '0, 0, 0',
            '--mjx-bg-black': '255, 255, 255',
            '--mjx-fg-white': '0, 0, 0',
            '--mjx-fg-black': '255, 255, 255',
            '--mjx-live-bg-color': '#222025',
            '--mjx-live-shadow-color': 'black',
            '--mjx-live-border-color': '#7C7C7C',
            '--mjx-bg1-alpha': 0.3,
            '--mjx-fg1-alpha': 1,
            '--mjx-bg2-alpha': 1,
            '--mjx-fg2-alpha': 1,
        },
    },
    [`.${LiveRegion.className}`]: {
        position: 'absolute',
        top: 0,
        display: 'none',
        width: 'auto',
        height: 'auto',
        padding: 0,
        opacity: 1,
        'z-index': '202',
        left: 0,
        right: 0,
        margin: '0 auto',
        'background-color': 'var(--mjx-live-bg-color)',
        'box-shadow': '0px 5px 20px var(--mjx-live-shadow-color)',
        border: '2px solid var(--mjx-live-border-color)',
    },
    [`.${LiveRegion.className}_Show`]: {
        display: 'block',
    },
    [`.${LiveRegion.className} > div`]: {
        color: 'var(--mjx-fg1-color)',
        'background-color': 'var(--mjx-bg1-color)',
    },
    'mjx-container [data-sre-highlight-1]:not([data-mjx-collapsed], rect)': {
        color: 'var(--mjx-fg1-color) ! important',
        fill: 'var(--mjx-fg1-color) ! important',
    },
    [[
        'mjx-container:not([data-mjx-clone-container])',
        '[data-sre-highlight-1]:not([data-sre-enclosed], rect)',
    ].join(' ')]: {
        'background-color': 'var(--mjx-bg1-color) ! important',
    },
    'mjx-container rect[data-sre-highlight-1]:not([data-sre-enclosed])': {
        fill: 'var(--mjx-bg1-color) ! important',
    },
    'mjx-container [data-sre-highlight-2]': {
        color: 'var(--mjx-fg2-color) ! important',
        'background-color': 'var(--mjx-bg2-color) ! important',
        fill: 'var(--mjx-fg2-color) ! important',
    },
    'mjx-container rect[data-sre-highlight-2]': {
        fill: 'var(--mjx-bg2-color) ! important',
    },
});
export class SpeechRegion extends LiveRegion {
    constructor() {
        super(...arguments);
        this.active = false;
        this.node = null;
        this.clear = false;
        this.voiceRequest = false;
        this.voiceCancelled = false;
    }
    Show(node) {
        super.Update('\u00a0');
        this.node = node;
        super.Show(node);
    }
    Update(speech) {
        if (this.voiceRequest) {
            this.makeVoice(speech);
            return;
        }
        speechSynthesis.onvoiceschanged = (() => (this.voiceRequest = true)).bind(this);
        const promise = new Promise((resolve) => {
            setTimeout(() => {
                if (this.voiceRequest) {
                    resolve(true);
                }
                else {
                    setTimeout(() => {
                        this.voiceRequest = true;
                        resolve(true);
                    }, 100);
                }
            }, 100);
        });
        promise.then(() => this.makeVoice(speech));
    }
    makeVoice(speech) {
        this.active =
            this.document.options.a11y.voicing &&
                !!speechSynthesis.getVoices().length;
        speechSynthesis.cancel();
        this.clear = true;
        const [text, ssml] = buildSpeech(speech, this.document.options.sre.locale, this.document.options.sre.rate);
        super.Update(text);
        if (this.active && text) {
            this.makeUtterances(ssml, this.document.options.sre.locale);
        }
    }
    makeUtterances(ssml, locale) {
        this.voiceCancelled = false;
        let utterance = null;
        for (const utter of ssml) {
            if (utter.mark) {
                if (!utterance) {
                    this.highlightNode(utter.mark, true);
                    continue;
                }
                utterance.addEventListener('end', (_event) => {
                    if (!this.voiceCancelled) {
                        this.highlightNode(utter.mark);
                    }
                });
                continue;
            }
            if (utter.pause) {
                const time = parseInt(utter.pause.match(/^[0-9]+/)[0]);
                if (isNaN(time) || !utterance) {
                    continue;
                }
                utterance.addEventListener('end', (_event) => {
                    speechSynthesis.pause();
                    setTimeout(() => {
                        speechSynthesis.resume();
                    }, time);
                });
                continue;
            }
            utterance = new SpeechSynthesisUtterance(utter.text);
            if (utter.rate) {
                utterance.rate = utter.rate;
            }
            if (utter.pitch) {
                utterance.pitch = utter.pitch;
            }
            utterance.lang = locale;
            speechSynthesis.speak(utterance);
        }
        if (utterance) {
            utterance.addEventListener('end', (_event) => {
                this.highlighter.unhighlight();
            });
        }
    }
    Hide() {
        this.cancelVoice();
        super.Hide();
    }
    cancelVoice() {
        this.voiceCancelled = true;
        speechSynthesis.cancel();
        this.highlighter.unhighlight();
    }
    highlightNode(id, init = false) {
        this.highlighter.unhighlight();
        const nodes = Array.from(this.node.querySelectorAll(`[data-semantic-id="${id}"]`));
        if (!this.clear || init) {
            this.highlighter.highlight(nodes);
        }
        this.clear = false;
    }
}
SpeechRegion.style = null;
export class HoverRegion extends AbstractRegion {
    position(node) {
        const nodeRect = node.getBoundingClientRect();
        const divRect = this.div.getBoundingClientRect();
        const xCenter = nodeRect.left + nodeRect.width / 2;
        let left = xCenter - divRect.width / 2;
        left = left < 0 ? 0 : left;
        left = left + window.scrollX;
        let top;
        switch (this.document.options.a11y.align) {
            case 'top':
                top = nodeRect.top - divRect.height - 10;
                break;
            case 'bottom':
                top = nodeRect.bottom + 10;
                break;
            case 'center':
            default: {
                const yCenter = nodeRect.top + nodeRect.height / 2;
                top = yCenter - divRect.height / 2;
            }
        }
        top = top + window.scrollY;
        top = top < 0 ? 0 : top;
        this.div.style.top = top + 'px';
        this.div.style.left = left + 'px';
    }
    Show(node) {
        this.AddElement();
        this.div.style.fontSize = this.document.options.a11y.magnify;
        this.Update(node);
        super.Show(node);
    }
    Clear() {
        if (!this.div)
            return;
        this.inner.textContent = '';
        this.inner.style.top = '';
        this.inner.style.backgroundColor = '';
    }
    Update(node) {
        if (!this.div)
            return;
        this.Clear();
        const mjx = this.cloneNode(node);
        const selected = mjx.querySelector('[data-mjx-clone]');
        this.inner.style.backgroundColor = node.style.backgroundColor;
        selected.style.backgroundColor = '';
        selected.classList.remove('mjx-selected');
        this.inner.appendChild(mjx);
        this.position(node);
    }
    cloneNode(node) {
        let mjx = node.cloneNode(true);
        mjx.setAttribute('data-mjx-clone', 'true');
        if (mjx.nodeName !== 'MJX-CONTAINER') {
            if (mjx.nodeName !== 'g') {
                mjx.style.marginLeft = mjx.style.marginRight = '0';
            }
            const container = node.closest('mjx-container');
            if (mjx.nodeName !== 'MJX-MATH' && mjx.nodeName !== 'svg') {
                let math = container.firstChild;
                if (math.nodeName === 'MJX-BBOX') {
                    math = math.nextSibling;
                }
                mjx = math.cloneNode(false).appendChild(mjx).parentElement;
                const enclosed = Array.from(container.querySelectorAll('[data-sre-enclosed]'));
                math.nodeName === 'svg'
                    ? this.svgClone(node, enclosed, mjx, container)
                    : this.chtmlClone(node, enclosed, mjx);
            }
            mjx = container.cloneNode(false).appendChild(mjx).parentElement;
            mjx.style.margin = '0';
            mjx.style.minWidth = '';
        }
        mjx.setAttribute('data-mjx-clone-container', 'true');
        return mjx;
    }
    chtmlClone(node, enclosed, mjx) {
        for (const child of enclosed) {
            if (child !== node) {
                const id = child.getAttribute('data-semantic-id');
                if (!id || !mjx.querySelector(`[data-semantic-id="${id}"]`)) {
                    mjx.appendChild(child.cloneNode(true));
                }
            }
        }
    }
    svgClone(node, enclosed, mjx, container) {
        var _a;
        let { x, y, width, height } = node.getBBox();
        if (enclosed.length) {
            mjx.firstChild.remove();
            const g = container.querySelector('g').cloneNode(false);
            for (const child of enclosed) {
                const clone = g.appendChild(child.cloneNode(true));
                if (child === node) {
                    clone.setAttribute('data-mjx-clone', 'true');
                }
                const [cx, cy] = this.xy(child);
                clone.setAttribute('transform', `translate(${cx}, ${cy})`);
            }
            mjx.appendChild(g);
            const rect = node.previousSibling;
            const bbox = rect.getBBox();
            width = bbox.width;
            height = bbox.height;
            const [X, Y] = this.xy(rect);
            x = X;
            y = Y + bbox.y;
        }
        const g = container.querySelector('g');
        if (container.getAttribute('width') === 'full' &&
            g.firstChild.lastChild === node) {
            mjx.innerHTML = '';
            mjx.appendChild(container.cloneNode(true).firstChild);
            mjx.querySelector('.mjx-selected').setAttribute('data-mjx-clone', 'true');
            (_a = mjx.querySelector('[data-sre-highlighter-added]')) === null || _a === void 0 ? void 0 : _a.remove();
            return;
        }
        mjx.firstChild.setAttribute('transform', 'scale(1, -1)');
        const W = parseFloat((mjx.getAttribute('viewBox') || mjx.getAttribute('data-mjx-viewBox')).split(/ /)[2]);
        const w = parseFloat(mjx.style.minWidth || mjx.getAttribute('width'));
        mjx.setAttribute('viewBox', [x, -(y + height), width, height].join(' '));
        mjx.removeAttribute('style');
        mjx.setAttribute('width', (w / W) * width + 'ex');
        mjx.setAttribute('height', (w / W) * height + 'ex');
    }
    xy(node) {
        const P = DOMPoint.fromPoint({ x: 0, y: 0 }).matrixTransform(node.getCTM().inverse());
        return [-P.x, -P.y];
    }
}
HoverRegion.className = 'MJX_HoverRegion';
HoverRegion.style = new StyleJsonSheet({
    [`.${HoverRegion.className}`]: {
        display: 'block',
        position: 'absolute',
        width: 'max-content',
        height: 'auto',
        padding: 0,
        opacity: 1,
        'z-index': '202',
        margin: '0 auto',
        'background-color': 'white',
        'line-height': 0,
        'box-shadow': '0px 10px 20px #888',
        border: '2px solid #CCCCCC',
    },
    [`.${HoverRegion.className} > div`]: {
        overflow: 'hidden',
        color: 'var(--mjx-fg1-color)',
        'background-color': 'var(--mjx-bg1-color)',
    },
    '@media (prefers-color-scheme: dark)': {
        ['.' + HoverRegion.className]: {
            'background-color': '#222025',
            'box-shadow': '0px 5px 20px #000',
            border: '1px solid #7C7C7C',
        },
    },
    'mjx-container[data-mjx-clone-container]': {
        padding: '2px ! important',
    },
    'mjx-math > mjx-mlabeledtr': {
        display: 'inline-block',
        'margin-right': '.5em ! important',
    },
    'mjx-math > mjx-mtd': {
        float: 'right',
    },
});
