// speech-rule-engine/cjs/api/control.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.localeLoader = void 0;
exports.setupEngine = setupEngine;
exports.engineSetup = engineSetup;
exports.resetEngine = resetEngine;
exports.engineReady = engineReady;
exports.exit = exit;
const engine_js_1 = require("../common/engine.js");
const engine_setup_js_1 = require("../common/engine_setup.js");
const math_map_js_1 = require("../speech_rules/math_map.js");
function setupEngine(feature) {
    return __awaiter(this, void 0, void 0, function* () {
        return (0, engine_setup_js_1.setupEngine)(feature);
    });
}
function engineSetup() {
    return (0, engine_setup_js_1.engineSetup)();
}
function resetEngine() {
    engine_js_1.Engine.getInstance().reset();
}
function engineReady() {
    return __awaiter(this, void 0, void 0, function* () {
        return setupEngine({}).then(() => engine_js_1.EnginePromise.getall());
    });
}
exports.localeLoader = math_map_js_1.standardLoader;
function exit(opt_value) {
    const value = opt_value || 0;
    engine_js_1.EnginePromise.getall().then(() => process.exit(value));
}
