// speech-rule-engine/cjs/api/string.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.toSpeech = toSpeech;
exports.toSemantic = toSemantic;
exports.toJson = toJson;
exports.toVis = toVis;
exports.toDescription = toDescription;
exports.toEnriched = toEnriched;
exports.toSpeechStructure = toSpeechStructure;
const processor_factory_js_1 = require("../common/processor_factory.js");
function toSpeech(expr) {
    return (0, processor_factory_js_1.processString)('speech', expr);
}
function toSemantic(expr) {
    return (0, processor_factory_js_1.processString)('semantic', expr);
}
function toJson(expr) {
    return (0, processor_factory_js_1.processString)('json', expr);
}
function toVis(expr) {
    return (0, processor_factory_js_1.processString)('vis', expr);
}
function toDescription(expr) {
    return (0, processor_factory_js_1.processString)('description', expr);
}
function toEnriched(expr) {
    return (0, processor_factory_js_1.processString)('enriched', expr);
}
function toSpeechStructure(expr) {
    return (0, processor_factory_js_1.processString)('speechStructure', expr);
}
