// speech-rule-engine/cjs/api/worker.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.speech = speech;
exports.nextRules = nextRules;
exports.nextStyle = nextStyle;
exports.localePreferences = localePreferences;
exports.relevantPreferences = relevantPreferences;
const engine_js_1 = require("../common/engine.js");
const SpeechGeneratorUtil = __importStar(require("../speech_generator/speech_generator_util.js"));
const ProcessorFactory = __importStar(require("../common/processor_factory.js"));
const rebuild_stree_js_1 = require("../walker/rebuild_stree.js");
const DomUtil = __importStar(require("../common/dom_util.js"));
const engine_setup_js_1 = require("../common/engine_setup.js");
const clearspeak_preferences_js_1 = require("../speech_rules/clearspeak_preferences.js");
function speech(expr, options) {
    return __awaiter(this, void 0, void 0, function* () {
        const mml = DomUtil.parseInput(expr);
        const rebuilt = new rebuild_stree_js_1.RebuildStree(mml);
        const styles = SpeechGeneratorUtil.toStyles(options);
        options.domain2style = SpeechGeneratorUtil.fromStyles(styles);
        return assembleWorkerStructure(mml, rebuilt.stree.xml(), options);
    });
}
function nextRules(expr, options) {
    return __awaiter(this, void 0, void 0, function* () {
        const mml = DomUtil.parseInput(expr);
        const rebuilt = new rebuild_stree_js_1.RebuildStree(mml);
        const styles = SpeechGeneratorUtil.toStyles(options);
        options = SpeechGeneratorUtil.nextRules(options, styles);
        options.domain2style = SpeechGeneratorUtil.fromStyles(styles);
        return assembleWorkerStructure(mml, rebuilt.stree.xml(), options);
    });
}
function nextStyle(expr, options, id) {
    return __awaiter(this, void 0, void 0, function* () {
        const mml = DomUtil.parseInput(expr);
        const rebuilt = new rebuild_stree_js_1.RebuildStree(mml);
        const styles = SpeechGeneratorUtil.toStyles(options);
        options.style = SpeechGeneratorUtil.nextStyle(rebuilt.nodeDict[id], options);
        styles[options.domain] = options.style;
        options.domain2style = SpeechGeneratorUtil.fromStyles(styles);
        return assembleWorkerStructure(mml, rebuilt.stree.xml(), options);
    });
}
function localePreferences(options) {
    return __awaiter(this, void 0, void 0, function* () {
        return clearspeak_preferences_js_1.ClearspeakPreferences.getLocalePreferences()[options.locale];
    });
}
function relevantPreferences(expr, id) {
    return __awaiter(this, void 0, void 0, function* () {
        var _a;
        const mml = DomUtil.parseInput(expr);
        const rebuilt = new rebuild_stree_js_1.RebuildStree(mml);
        const query = (_a = rebuilt.stree.root.querySelectorAll((x) => x.id.toString() === id)[0]) !== null && _a !== void 0 ? _a : rebuilt.stree.root;
        return clearspeak_preferences_js_1.ClearspeakPreferences.relevantPreferences(query);
    });
}
function assembleWorkerStructure(mml, sxml, options) {
    return __awaiter(this, void 0, void 0, function* () {
        var _a;
        yield (0, engine_setup_js_1.setupEngine)(options);
        engine_js_1.Engine.getInstance().options.automark = true;
        const json = {};
        ProcessorFactory.assembleSpeechStructure(json, mml, sxml, options);
        if (options.enableBraille === false) {
            return json;
        }
        yield (0, engine_setup_js_1.setupEngine)({
            modality: 'braille',
            locale: options.braille,
            domain: 'default',
            style: 'default'
        });
        const root = (_a = sxml.childNodes[0]) === null || _a === void 0 ? void 0 : _a.getAttribute('id');
        json.braille = SpeechGeneratorUtil.computeBrailleStructure(sxml);
        json.braillelabel = json.braille[root]['braille-none'];
        return json;
    });
}
