// speech-rule-engine/js/api/control.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { Engine, EnginePromise } from '../common/engine.js';
import { setupEngine as setup, engineSetup as engine } from '../common/engine_setup.js';
import { standardLoader } from '../speech_rules/math_map.js';
export function setupEngine(feature) {
    return __awaiter(this, void 0, void 0, function* () {
        return setup(feature);
    });
}
export function engineSetup() {
    return engine();
}
export function resetEngine() {
    Engine.getInstance().reset();
}
export function engineReady() {
    return __awaiter(this, void 0, void 0, function* () {
        return setupEngine({}).then(() => EnginePromise.getall());
    });
}
export const localeLoader = standardLoader;
export function exit(opt_value) {
    const value = opt_value || 0;
    EnginePromise.getall().then(() => process.exit(value));
}
