// speech-rule-engine/js/api/file.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import * as FileUtil from '../common/file_util.js';
import { Engine, SREError } from '../common/engine.js';
import * as ProcessorFactory from '../common/processor_factory.js';
import { SystemExternal } from '../common/system_external.js';
import * as EngineConst from '../common/engine_const.js';
export function toSpeech(input, opt_output) {
    return processFile('speech', input, opt_output);
}
;
export function toSemantic(input, opt_output) {
    return processFile('semantic', input, opt_output);
}
;
export function toJson(input, opt_output) {
    return processFile('json', input, opt_output);
}
;
export function toDescription(input, opt_output) {
    return processFile('description', input, opt_output);
}
;
export function toEnriched(input, opt_output) {
    return processFile('enriched', input, opt_output);
}
;
export function processFile(processor, input, opt_output) {
    switch (Engine.getInstance().mode) {
        case EngineConst.Mode.ASYNC:
            return processFileAsync(processor, input, opt_output);
        case EngineConst.Mode.SYNC:
            return processFileSync(processor, input, opt_output);
        default:
            throw new SREError(`Cannot process files in ${Engine.getInstance().mode} mode`);
    }
}
function processFileSync(processor, input, opt_output) {
    const expr = inputFileSync_(input);
    const result = ProcessorFactory.output(processor, expr);
    if (opt_output) {
        try {
            SystemExternal.fs.writeFileSync(opt_output, result);
        }
        catch (_err) {
            throw new SREError('Can not write to file: ' + opt_output);
        }
    }
    return result;
}
function inputFileSync_(file) {
    let expr;
    try {
        expr = SystemExternal.fs.readFileSync(file, { encoding: 'utf8' });
    }
    catch (_err) {
        throw new SREError('Can not open file: ' + file);
    }
    return expr;
}
function processFileAsync(processor, file, output) {
    return __awaiter(this, void 0, void 0, function* () {
        const expr = yield SystemExternal.fs.promises.readFile(file, {
            encoding: 'utf8'
        });
        const result = ProcessorFactory.output(processor, expr);
        if (output) {
            try {
                SystemExternal.fs.promises.writeFile(output, result);
            }
            catch (_err) {
                throw new SREError('Can not write to file: ' + output);
            }
        }
        return result;
    });
}
export const localePath = FileUtil.localePath;
