// speech-rule-engine/js/api/worker.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { Engine } from '../common/engine.js';
import * as SpeechGeneratorUtil from '../speech_generator/speech_generator_util.js';
import * as ProcessorFactory from '../common/processor_factory.js';
import { RebuildStree } from '../walker/rebuild_stree.js';
import * as DomUtil from '../common/dom_util.js';
import { setupEngine } from '../common/engine_setup.js';
import { ClearspeakPreferences } from '../speech_rules/clearspeak_preferences.js';
export function speech(expr, options) {
    return __awaiter(this, void 0, void 0, function* () {
        const mml = DomUtil.parseInput(expr);
        const rebuilt = new RebuildStree(mml);
        const styles = SpeechGeneratorUtil.toStyles(options);
        options.domain2style = SpeechGeneratorUtil.fromStyles(styles);
        return assembleWorkerStructure(mml, rebuilt.stree.xml(), options);
    });
}
export function nextRules(expr, options) {
    return __awaiter(this, void 0, void 0, function* () {
        const mml = DomUtil.parseInput(expr);
        const rebuilt = new RebuildStree(mml);
        const styles = SpeechGeneratorUtil.toStyles(options);
        options = SpeechGeneratorUtil.nextRules(options, styles);
        options.domain2style = SpeechGeneratorUtil.fromStyles(styles);
        return assembleWorkerStructure(mml, rebuilt.stree.xml(), options);
    });
}
export function nextStyle(expr, options, id) {
    return __awaiter(this, void 0, void 0, function* () {
        const mml = DomUtil.parseInput(expr);
        const rebuilt = new RebuildStree(mml);
        const styles = SpeechGeneratorUtil.toStyles(options);
        options.style = SpeechGeneratorUtil.nextStyle(rebuilt.nodeDict[id], options);
        styles[options.domain] = options.style;
        options.domain2style = SpeechGeneratorUtil.fromStyles(styles);
        return assembleWorkerStructure(mml, rebuilt.stree.xml(), options);
    });
}
export function localePreferences(options) {
    return __awaiter(this, void 0, void 0, function* () {
        return ClearspeakPreferences.getLocalePreferences()[options.locale];
    });
}
export function relevantPreferences(expr, id) {
    return __awaiter(this, void 0, void 0, function* () {
        var _a;
        const mml = DomUtil.parseInput(expr);
        const rebuilt = new RebuildStree(mml);
        const query = (_a = rebuilt.stree.root.querySelectorAll((x) => x.id.toString() === id)[0]) !== null && _a !== void 0 ? _a : rebuilt.stree.root;
        return ClearspeakPreferences.relevantPreferences(query);
    });
}
function assembleWorkerStructure(mml, sxml, options) {
    return __awaiter(this, void 0, void 0, function* () {
        var _a;
        yield setupEngine(options);
        Engine.getInstance().options.automark = true;
        const json = {};
        ProcessorFactory.assembleSpeechStructure(json, mml, sxml, options);
        if (options.enableBraille === false) {
            return json;
        }
        yield setupEngine({
            modality: 'braille',
            locale: options.braille,
            domain: 'default',
            style: 'default'
        });
        const root = (_a = sxml.childNodes[0]) === null || _a === void 0 ? void 0 : _a.getAttribute('id');
        json.braille = SpeechGeneratorUtil.computeBrailleStructure(sxml);
        json.braillelabel = json.braille[root]['braille-none'];
        return json;
    });
}
