// @mathjax/src/cjs/output/chtml/Wrappers/mo.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChtmlMo = void 0;
var Wrapper_js_1 = require("../Wrapper.js");
var FontData_js_1 = require("../FontData.js");
var mo_js_1 = require("../../common/Wrappers/mo.js");
var mo_js_2 = require("../../../core/MmlTree/MmlNodes/mo.js");
var FontData_js_2 = require("../FontData.js");
exports.ChtmlMo = (function () {
    var _a;
    var _b;
    var Base = (0, mo_js_1.CommonMoMixin)(Wrapper_js_1.ChtmlWrapper);
    return _b = (function (_super) {
            __extends(ChtmlMo, _super);
            function ChtmlMo() {
                return _super !== null && _super.apply(this, arguments) || this;
            }
            ChtmlMo.prototype.toCHTML = function (parents) {
                var _this = this;
                var adaptor = this.adaptor;
                var attributes = this.node.attributes;
                var symmetric = attributes.get('symmetric') &&
                    this.stretch.dir !== FontData_js_2.DIRECTION.Horizontal;
                var stretchy = this.stretch.dir !== FontData_js_2.DIRECTION.None;
                if (stretchy && this.size === null) {
                    this.getStretchedVariant([]);
                }
                if (parents.length > 1) {
                    parents.forEach(function (dom) {
                        return adaptor.append(dom, _this.html('mjx-linestrut'));
                    });
                }
                var chtml = this.standardChtmlNodes(parents);
                if (chtml.length > 1 && this.breakStyle !== 'duplicate') {
                    var i = this.breakStyle === 'after' ? 1 : 0;
                    adaptor.remove(chtml[i]);
                    chtml[i] = null;
                }
                if (stretchy && this.size < 0) {
                    this.stretchHTML(chtml);
                }
                else {
                    if (symmetric || attributes.get('largeop')) {
                        var u_1 = this.em(this.getCenterOffset());
                        if (u_1 !== '0') {
                            chtml.forEach(function (dom) { return dom && adaptor.setStyle(dom, 'verticalAlign', u_1); });
                        }
                    }
                    if (this.node.getProperty('mathaccent')) {
                        chtml.forEach(function (dom) {
                            adaptor.setStyle(dom, 'width', '0');
                            adaptor.setStyle(dom, 'margin-left', _this.em(_this.getAccentOffset()));
                        });
                    }
                    if (chtml[0]) {
                        this.addChildren([chtml[0]]);
                    }
                    if (chtml[1]) {
                        (this.multChar || this).addChildren([chtml[1]]);
                    }
                }
            };
            ChtmlMo.prototype.stretchHTML = function (chtml) {
                var c = this.getText().codePointAt(0);
                this.font.delimUsage.add(c);
                this.childNodes[0].markUsed();
                var delim = this.stretch;
                var stretch = delim.stretch;
                var stretchv = this.font.getStretchVariants(c);
                var dom = [];
                var parts = [];
                for (var i = 0; i < stretch.length; i++) {
                    if (stretch[i]) {
                        parts[i] = this.font.getChar(stretchv[i], stretch[i]);
                    }
                }
                var _a = this.bbox, h = _a.h, d = _a.d, w = _a.w;
                var styles = {};
                if (delim.dir === FontData_js_2.DIRECTION.Vertical) {
                    this.createAssembly(parts, stretch, stretchv, dom, h + d, FontData_js_1.VFUZZ, '\n');
                    dom.push(this.html('mjx-mark'));
                    styles.height = this.em(h + d);
                    styles.verticalAlign = this.em(-d);
                }
                else {
                    var ext = (delim.ext || 0) + FontData_js_1.HFUZZ;
                    this.createAssembly(parts, stretch, stretchv, dom, w, ext);
                    styles.width = this.em(w);
                }
                var properties = { class: this.char(delim.c || c), style: styles };
                var html = this.html('mjx-stretchy-' + delim.dir, properties, dom);
                var adaptor = this.adaptor;
                if (chtml[0]) {
                    adaptor.append(chtml[0], html);
                }
                if (chtml[1]) {
                    adaptor.append(chtml[1], chtml[0] ? adaptor.clone(html) : html);
                }
            };
            ChtmlMo.prototype.createAssembly = function (parts, sn, sv, dom, wh, ext, nl) {
                if (nl === void 0) { nl = ''; }
                parts = __spreadArray(__spreadArray([], __read(parts), false), [null, null, null], false).slice(0, 4);
                var _a = __read(parts.map(function (part) {
                    return part ? (nl ? part[0] + part[1] : part[2]) : 0;
                }), 4), WHb = _a[0], WHx = _a[1], WHe = _a[2], WHm = _a[3];
                WHx = Math.max(0, WHx - ext);
                var _c = __read(parts[3]
                    ? [(wh - WHm) / 2 - WHb, (wh - WHm) / 2 - WHe]
                    : [wh - WHb - WHe, 0], 2), WH1 = _c[0], WH2 = _c[1];
                this.createPart('mjx-beg', parts[0], sn[0], sv[0], dom);
                this.createPart('mjx-ext', parts[1], sn[1], sv[1], dom, WH1, WHx, nl, WHb, WHm / 2 || WHe);
                if (parts[3]) {
                    this.createPart('mjx-mid', parts[3], sn[3], sv[3], dom);
                    this.createPart('mjx-ext', parts[1], sn[1], sv[1], dom, WH2, WHx, nl, WHm / 2, WHe);
                }
                this.createPart('mjx-end', parts[2], sn[2], sv[2], dom);
            };
            ChtmlMo.prototype.createPart = function (part, data, n, v, dom, W, Wx, nl, Wb, We) {
                if (W === void 0) { W = 0; }
                if (Wx === void 0) { Wx = 0; }
                if (nl === void 0) { nl = ''; }
                if (Wb === void 0) { Wb = 0; }
                if (We === void 0) { We = 0; }
                if (n) {
                    var options = data[3];
                    var letter = options.f || (v === 'normal' ? '' : this.font.getVariant(v).letter);
                    var font = options.ff || (letter ? "".concat(this.font.cssFontPrefix, "-").concat(letter) : '');
                    var c = options.c || String.fromCodePoint(n);
                    var nodes = [];
                    if (part === 'mjx-ext' && (Wx || options.dx)) {
                        if (W < 0 && nl) {
                            dom.push(this.html(part, __assign(__assign({}, (font ? { class: font } : {})), { style: {
                                    'border-width': "".concat(this.em(Wb + W / 2), " 0 ").concat(this.em(We + W / 2)),
                                    'clip-path': 'none',
                                } })));
                            return;
                        }
                        if (!Wx) {
                            Wx = Math.max(FontData_js_1.HFUZZ, 2 * options.dx - FontData_js_1.HFUZZ);
                        }
                        var n_1 = Math.min(Math.ceil(W / Wx) + 1, 500);
                        if (options.cmb) {
                            nodes.push(this.html('mjx-spacer'));
                            for (var i = 0; i < n_1; i++) {
                                nodes.push(this.html('mjx-c', {}, [this.text(c)]));
                            }
                        }
                        else {
                            nodes = [
                                this.html('mjx-spacer', {}, [
                                    this.text(Array(n_1).fill(c).join(nl)),
                                ]),
                            ];
                        }
                    }
                    else {
                        nodes = [this.text(c)];
                    }
                    dom.push(this.html(part, font ? { class: font } : {}, nodes));
                }
            };
            return ChtmlMo;
        }(Base)),
        _b.kind = mo_js_2.MmlMo.prototype.kind,
        _b.styles = (_a = {
                'mjx-stretchy-h': {
                    display: 'inline-block',
                }
            },
            _a[['beg', 'ext', 'end', 'mid']
                .map(function (node) { return "mjx-stretchy-h > mjx-".concat(node); })
                .join(', ')] = {
                display: 'inline-block',
                width: 0,
                'text-align': 'right',
            },
            _a['mjx-stretchy-h > mjx-ext'] = {
                'clip-path': 'padding-box polygon(0 -1em, 100% -1em, 100% calc(100% + 1em), 0 calc(100% + 1em))',
                width: '100%',
                border: '0px solid transparent',
                'box-sizing': 'border-box',
                'text-align': 'left',
            },
            _a['mjx-stretchy-v'] = {
                display: 'inline-block',
                'text-align': 'center',
            },
            _a[['beg', 'ext', 'end', 'mid']
                .map(function (node) { return "mjx-stretchy-v > mjx-".concat(node); })
                .join(', ')] = {
                display: 'block',
                height: 0,
                margin: '0 auto',
            },
            _a['mjx-stretchy-v > mjx-ext > mjx-spacer'] = {
                display: 'block',
            },
            _a['mjx-stretchy-v > mjx-ext'] = {
                'clip-path': 'padding-box polygon(-1em 0, calc(100% + 1em) 0, calc(100% + 1em) 100%, -1em 100%)',
                height: '100%',
                border: '0.1px solid transparent',
                'box-sizing': 'border-box',
                'white-space': 'pre',
            },
            _a['mjx-mark'] = {
                display: 'inline-block',
                height: 0,
            },
            _a),
        _b;
})();
