// @mathjax/src/cjs/ui/dialog/DraggableDialog.js
"use strict";
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DraggableDialog = exports.isDialog = void 0;
var StyleJson_js_1 = require("../../util/StyleJson.js");
var context_js_1 = require("../../util/context.js");
exports.isDialog = !!((_a = context_js_1.context.window) === null || _a === void 0 ? void 0 : _a.HTMLDialogElement);
var DraggableDialog = (function () {
    function DraggableDialog(args) {
        this.minW = 200;
        this.minH = 80;
        this.tx = 0;
        this.ty = 0;
        this.dragging = false;
        this.events = [
            ['mousemove', this.MouseMove.bind(this)],
            ['mouseup', this.MouseUp.bind(this)],
        ];
        this.mode = '';
        this.actions = {
            down: {
                move: function (d) {
                    d.dialog.classList.add('mjx-moving');
                },
            },
            move: {
                move: function (dg, ev) { return [ev.x - dg.x, ev.y - dg.y, 0, 0]; },
                top: function (dg, ev) { return [0, (ev.y - dg.y) / 2, 0, dg.y - ev.y]; },
                bottom: function (dg, ev) { return [0, (ev.y - dg.y) / 2, 0, ev.y - dg.y]; },
                left: function (dg, ev) { return [(ev.x - dg.x) / 2, 0, dg.x - ev.x, 0]; },
                right: function (dg, ev) { return [(ev.x - dg.x) / 2, 0, ev.x - dg.x, 0]; },
                topleft: function (dg, ev) { return [
                    (ev.x - dg.x) / 2,
                    (ev.y - dg.y) / 2,
                    dg.x - ev.x,
                    dg.y - ev.y,
                ]; },
                topright: function (dg, ev) { return [
                    (ev.x - dg.x) / 2,
                    (ev.y - dg.y) / 2,
                    ev.x - dg.x,
                    dg.y - ev.y,
                ]; },
                botleft: function (dg, ev) { return [
                    (ev.x - dg.x) / 2,
                    (ev.y - dg.y) / 2,
                    dg.x - ev.x,
                    ev.y - dg.y,
                ]; },
                botright: function (dg, ev) { return [
                    (ev.x - dg.x) / 2,
                    (ev.y - dg.y) / 2,
                    ev.x - dg.x,
                    ev.y - dg.y,
                ]; },
            },
            up: {
                move: function (dg) {
                    dg.dialog.classList.remove('mjx-moving');
                },
            },
            keymove: {
                left: function () { return [-5, 0, 0, 0]; },
                right: function () { return [5, 0, 0, 0]; },
                up: function () { return [0, -5, 0, 0]; },
                down: function () { return [0, 5, 0, 0]; },
            },
            bigmove: {
                left: function () { return [-20, 0, 0, 0]; },
                right: function () { return [20, 0, 0, 0]; },
                up: function () { return [0, -20, 0, 0]; },
                down: function () { return [0, 20, 0, 0]; },
            },
            keysize: {
                left: function () { return [-3, 0, -6, 0]; },
                right: function () { return [3, 0, 6, 0]; },
                up: function () { return [0, -3, 0, -6]; },
                down: function () { return [0, 3, 0, 6]; },
            },
            bigsize: {
                left: function () { return [-10, 0, -20, 0]; },
                right: function () { return [10, 0, 20, 0]; },
                up: function () { return [0, -10, 0, -20]; },
                down: function () { return [0, 10, 0, 20]; },
            },
        };
        var adaptor = args.adaptor, _a = args.node, node = _a === void 0 ? null : _a;
        this.init(adaptor);
        this.node = node;
        this.background = exports.isDialog ? null : adaptor.node('mjx-dialog-background');
        this.x = this.y = 0;
        this.dragging = false;
        this.action = '';
        this.dialog = this.html(args);
        this.title = this.dialog.firstChild.firstChild.firstChild;
        this.content = this.dialog.firstChild.firstChild.nextSibling;
        var close = this.dialog.lastChild;
        close.addEventListener('click', this.closeDialog.bind(this));
        close.addEventListener('keydown', this.actionKey.bind(this, this.closeDialog.bind(this)));
        var help = this.dialog.lastChild.previousSibling;
        help.addEventListener('click', this.helpDialog.bind(this, adaptor));
        help.addEventListener('keydown', this.actionKey.bind(this, this.helpDialog.bind(this, adaptor)));
        this.noDrag = Array.from(this.dialog.querySelectorAll('[data-drag="none"]'));
    }
    DraggableDialog.prototype.init = function (adaptor) {
        var CLASS = this.constructor;
        var head = adaptor.document.head;
        if (!head.querySelector('#' + CLASS.styleId)) {
            var style = adaptor.node('style', { id: CLASS.styleId });
            style.textContent = new StyleJson_js_1.StyleJsonSheet(CLASS.styles).cssText;
            adaptor.document.head.append(style);
        }
    };
    DraggableDialog.prototype.html = function (args) {
        var title = args.title, message = args.message, adaptor = args.adaptor, _a = args.styles, styles = _a === void 0 ? null : _a, _b = args.extraNodes, extraNodes = _b === void 0 ? [] : _b, _c = args.className, className = _c === void 0 ? DraggableDialog.className : _c;
        if (styles) {
            var stylesheet = adaptor.node('style');
            stylesheet.textContent = new StyleJson_js_1.StyleJsonSheet(styles).cssText;
            extraNodes.unshift(stylesheet);
        }
        var label = 'mjx-dialog-label-' + DraggableDialog.id++;
        var dialog = adaptor.node('dialog', { closedby: 'any', class: ('mjx-dialog ' + className).trim() }, __spreadArray(__spreadArray([
            adaptor.node('mjx-dialog', { 'aria-labeledby': label }, [
                adaptor.node('mjx-title', {}, [
                    adaptor.node('h1', { id: label, tabIndex: 0 }),
                ]),
                adaptor.node('div', { 'data-drag': 'none', tabIndex: 0 }),
            ])
        ], __read(extraNodes), false), [
            adaptor.node('mjx-dialog-spacer', { 'aria-hidden': true }),
            adaptor.node('mjx-dialog-drag', {
                'data-drag': 'top',
                'aria-hidden': true,
            }),
            adaptor.node('mjx-dialog-drag', {
                'data-drag': 'bottom',
                'aria-hidden': true,
            }),
            adaptor.node('mjx-dialog-drag', {
                'data-drag': 'left',
                'aria-hidden': true,
            }),
            adaptor.node('mjx-dialog-drag', {
                'data-drag': 'right',
                'aria-hidden': true,
            }),
            adaptor.node('mjx-dialog-drag', {
                'data-drag': 'topleft',
                'aria-hidden': true,
            }),
            adaptor.node('mjx-dialog-drag', {
                'data-drag': 'topright',
                'aria-hidden': true,
            }),
            adaptor.node('mjx-dialog-drag', {
                'data-drag': 'botleft',
                'aria-hidden': true,
            }),
            adaptor.node('mjx-dialog-drag', {
                'data-drag': 'botright',
                'aria-hidden': true,
            }),
            adaptor.node('mjx-dialog-help', {
                class: 'mjx-dialog-button',
                'data-drag': 'none',
                tabIndex: 0,
                role: 'button',
                'aria-label': 'Dialog Help',
            }, [
                adaptor.node('mjx-dialog-icon', { 'aria-hidden': true }, [
                    adaptor.text('?'),
                ]),
            ]),
            adaptor.node('mjx-dialog-close', {
                class: 'mjx-dialog-button',
                'data-drag': 'none',
                tabIndex: 0,
                role: 'button',
                'aria-label': 'Close Dialog Box',
            }, [
                adaptor.node('mjx-dialog-icon', { 'aria-hidden': true }, [
                    adaptor.text('\u00d7'),
                ]),
            ]),
        ], false));
        dialog.firstChild.firstChild.firstChild.innerHTML = title;
        dialog.firstChild.childNodes[1].innerHTML = message;
        return dialog;
    };
    DraggableDialog.prototype.attach = function () {
        if (exports.isDialog) {
            this.dialog.addEventListener('mousedown', this.MouseDown.bind(this));
            this.dialog.addEventListener('keydown', this.KeyDown.bind(this), true);
            document.body.append(this.dialog);
            this.dialog.showModal();
        }
        else {
            this.background.addEventListener('mousedown', this.MouseDown.bind(this));
            this.background.addEventListener('keydown', this.KeyDown.bind(this), true);
            this.dialog.setAttribute('tabindex', '0');
            this.dialog.addEventListener('click', this.stop);
            this.background.append(this.dialog);
            document.body.append(this.background);
        }
        context_js_1.context.window.addEventListener('visibilitychange', this.Visibility.bind(this));
        this.minW = Math.min(this.minW, this.dialog.clientWidth - 8);
        this.minH = Math.min(this.minH, this.dialog.clientHeight - this.title.offsetHeight - 8);
        this.title.focus();
    };
    DraggableDialog.prototype.dragAction = function (type, event) {
        if (event === void 0) { event = null; }
        if (event) {
            this.stop(event);
        }
        var action = this.actions[type][this.action];
        var result = action ? action(this, event) : null;
        if (!result) {
            return;
        }
        var _a = __read(result, 4), dx = _a[0], dy = _a[1], dw = _a[2], dh = _a[3];
        if (dw) {
            var W = this.w + dw;
            if (W >= this.minW) {
                this.x = event === null || event === void 0 ? void 0 : event.x;
                this.w = W;
                this.dialog.style.maxWidth = this.dialog.style.width = W + 'px';
            }
            else {
                dx = 0;
            }
        }
        if (dh) {
            var H = this.h + dh;
            if (H >= this.minH + this.title.offsetHeight) {
                this.y = event === null || event === void 0 ? void 0 : event.y;
                this.h = H;
                this.dialog.style.maxHeight = this.dialog.style.height = H + 'px';
            }
            else {
                dy = 0;
            }
        }
        if (dx || dy) {
            if (dx) {
                this.x = event === null || event === void 0 ? void 0 : event.x;
                this.tx += dx || 0;
            }
            if (dy) {
                this.y = event === null || event === void 0 ? void 0 : event.y;
                this.ty += dy || 0;
            }
            this.dialog.style.transform = "translate(".concat(this.tx, "px, ").concat(this.ty, "px)");
        }
    };
    DraggableDialog.prototype.MouseDown = function (event) {
        var e_1, _a;
        var target = event.target;
        if (event.buttons !== 1 ||
            event.shiftKey ||
            event.metaKey ||
            event.altKey ||
            event.ctrlKey) {
            return;
        }
        if (!this.inDialog(event)) {
            this.closeDialog(event);
            return;
        }
        try {
            for (var _b = __values(this.noDrag), _c = _b.next(); !_c.done; _c = _b.next()) {
                var node = _c.value;
                if (target === node || node.contains(target)) {
                    return;
                }
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_1) throw e_1.error; }
        }
        this.action = target.getAttribute('data-drag') || 'move';
        this.startDrag(event);
        this.dragAction('down', event);
    };
    DraggableDialog.prototype.MouseMove = function (event) {
        if (event.buttons !== 1) {
            this.endDrag();
        }
        if (this.dragging) {
            this.dragAction('move', event);
        }
    };
    DraggableDialog.prototype.MouseUp = function (event) {
        if (this.dragging) {
            this.dragAction('up', event);
            this.endDrag();
        }
    };
    DraggableDialog.prototype.Visibility = function () {
        if (context_js_1.context.document.hidden) {
            this.closeDialog();
        }
    };
    DraggableDialog.prototype.KeyDown = function (event) {
        var CLASS = this.constructor;
        var action = CLASS.keyActions.get(event.key);
        if (action) {
            action(this, event);
        }
    };
    DraggableDialog.prototype.escKey = function (event) {
        this.closeDialog(event);
    };
    DraggableDialog.prototype.aKey = function (event) {
        if (event.ctrlKey || event.metaKey) {
            this.selectAll();
            this.stop(event);
        }
    };
    DraggableDialog.prototype.mKey = function (event) {
        this.mode = this.mode === 'move' ? '' : 'move';
        this.stop(event);
    };
    DraggableDialog.prototype.sKey = function (event) {
        this.mode = this.mode === 'size' ? '' : 'size';
        this.stop(event);
    };
    DraggableDialog.prototype.arrowKey = function (event, direction) {
        if (event.ctrlKey || this.dragging)
            return;
        this.action = direction;
        this.getWH();
        if (event.altKey || this.mode === 'move') {
            this.dragAction(event.shiftKey ? 'bigmove' : 'keymove');
            this.stop(event);
        }
        else if (event.metaKey || this.mode === 'size') {
            this.dragAction(event.shiftKey ? 'bigsize' : 'keysize');
            this.stop(event);
        }
        this.action = '';
    };
    DraggableDialog.prototype.actionKey = function (action, event) {
        if (event.code === 'Enter' || event.code === 'Space') {
            action(event);
        }
    };
    DraggableDialog.prototype.selectAll = function () {
        var selection = document.getSelection();
        selection.selectAllChildren(this.content);
    };
    DraggableDialog.prototype.copyToClipboard = function () {
        this.selectAll();
        try {
            document.execCommand('copy');
        }
        catch (err) {
            alert("Can't copy to clipboard: ".concat(err.message));
        }
        document.getSelection().removeAllRanges();
    };
    DraggableDialog.prototype.startDrag = function (event) {
        var e_2, _a;
        this.x = event.x;
        this.y = event.y;
        this.getWH();
        this.dragging = true;
        var node = this.background || this.dialog;
        try {
            for (var _b = __values(this.events), _c = _b.next(); !_c.done; _c = _b.next()) {
                var _d = __read(_c.value, 2), name_1 = _d[0], listener = _d[1];
                node.addEventListener(name_1, listener);
            }
        }
        catch (e_2_1) { e_2 = { error: e_2_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_2) throw e_2.error; }
        }
    };
    DraggableDialog.prototype.getWH = function () {
        this.w = this.dialog.clientWidth - 8;
        this.h = this.dialog.clientHeight - 8;
    };
    DraggableDialog.prototype.endDrag = function () {
        var e_3, _a;
        this.action = '';
        this.dragging = false;
        var node = this.background || this.dialog;
        try {
            for (var _b = __values(this.events), _c = _b.next(); !_c.done; _c = _b.next()) {
                var _d = __read(_c.value, 2), name_2 = _d[0], listener = _d[1];
                node.removeEventListener(name_2, listener);
            }
        }
        catch (e_3_1) { e_3 = { error: e_3_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_3) throw e_3.error; }
        }
    };
    DraggableDialog.prototype.closeDialog = function (event) {
        var _a;
        if (exports.isDialog) {
            this.dialog.close();
            this.dialog.remove();
        }
        else {
            this.background.remove();
        }
        (_a = this.node) === null || _a === void 0 ? void 0 : _a.focus();
        if (event) {
            this.stop(event);
        }
    };
    DraggableDialog.prototype.helpDialog = function (adaptor, event) {
        var help = new DraggableDialog({
            title: 'MathJax Dialog Help',
            message: this.constructor.helpMessage,
            adaptor: adaptor,
            className: 'mjx-dialog-help',
            styles: {
                '.mjx-dialog-help': {
                    'max-width': 'calc(min(50em, 80%))',
                },
            },
        });
        help.attach();
        this.stop(event);
    };
    DraggableDialog.prototype.inDialog = function (event) {
        if (!this.dialog.contains(event.target)) {
            return false;
        }
        var x = event.x, y = event.y;
        var _a = this.dialog.getBoundingClientRect(), left = _a.left, right = _a.right, top = _a.top, bottom = _a.bottom;
        return x >= left && x <= right && y >= top && y <= bottom;
    };
    DraggableDialog.prototype.stop = function (event) {
        if (event.preventDefault) {
            event.preventDefault();
        }
        if (event.stopImmediatePropagation) {
            event.stopImmediatePropagation();
        }
        else if (event.stopPropagation) {
            event.stopPropagation();
        }
    };
    DraggableDialog.keyActions = new Map([
        ['Escape', function (dialog, event) { return dialog.escKey(event); }],
        ['a', function (dialog, event) { return dialog.aKey(event); }],
        ['m', function (dialog, event) { return dialog.mKey(event); }],
        ['s', function (dialog, event) { return dialog.sKey(event); }],
        ['ArrowRight', function (dialog, event) { return dialog.arrowKey(event, 'right'); }],
        ['ArrowLeft', function (dialog, event) { return dialog.arrowKey(event, 'left'); }],
        ['ArrowUp', function (dialog, event) { return dialog.arrowKey(event, 'up'); }],
        ['ArrowDown', function (dialog, event) { return dialog.arrowKey(event, 'down'); }],
    ]);
    DraggableDialog.styleId = 'MJX-DIALOG-styles';
    DraggableDialog.className = '';
    DraggableDialog.id = 0;
    DraggableDialog.styles = {
        'mjx-dialog-background': {
            display: 'flex',
            'flex-direction': 'column',
            'justify-content': 'center',
            position: 'fixed',
            top: 0,
            left: 0,
            right: 0,
            bottom: 0,
            'z-index': 1001,
        },
        '.mjx-dialog': {
            'max-width': 'calc(min(60em, 90%))',
            'max-height': 'calc(min(50em, 85%))',
            border: '3px outset',
            'border-radius': '15px',
            color: 'black',
            'background-color': '#DDDDDD',
            'box-shadow': '0px 10px 20px #808080',
            padding: '4px 4px',
            cursor: 'grab',
            overflow: 'visible',
            display: 'flex',
            'flex-direction': 'column',
            'align-items': 'center',
            position: 'fixed',
            top: '-4%',
        },
        '.mjx-dialog.mjx-moving': {
            cursor: 'grabbing',
        },
        '.mjx-dialog > input[type="button"]': {
            width: 'fit-content',
        },
        '.mjx-dialog > mjx-dialog-spacer': {
            display: 'block',
            height: '.75em',
            'flex-shrink': 0,
        },
        '.mjx-dialog::backdrop': {
            opacity: 0,
            cursor: 'default',
        },
        'mjx-dialog': {
            all: 'initial',
            cursor: 'inherit',
            width: '100%',
            display: 'flex',
            'flex-direction': 'column',
            'flex-grow': 1,
            'flex-shrink': 1,
            overflow: 'hidden',
        },
        'mjx-dialog > mjx-title': {
            display: 'block',
            'text-align': 'center',
            margin: '.25em 1.75em',
            overflow: 'hidden',
            'white-space': 'nowrap',
            '-webkit-user-select': 'none',
            'user-select': 'none',
            'flex-shrink': 0,
        },
        'mjx-dialog > mjx-title > h1': {
            'font-size': '125%',
            margin: 0,
        },
        'mjx-dialog > div': {
            margin: '0 1em .5em',
            padding: '8px 18px',
            overflow: 'auto',
            border: '2px inset black',
            'background-color': 'white',
            'text-align': 'left',
            cursor: 'default',
            'flex-grow': 1,
            'flex-shrink': 1,
        },
        'mjx-dialog > div > pre': {
            margin: 0,
        },
        '.mjx-dialog-button': {
            position: 'absolute',
            top: '6px',
            height: '17px',
            width: '17px',
            cursor: 'default',
            display: 'block',
            border: '2px solid #AAA',
            'border-radius': '18px',
            'font-family': '"Courier New", Courier',
            'text-align': 'center',
            color: '#F0F0F0',
            '-webkit-user-select': 'none',
            'user-select': 'none',
        },
        '.mjx-dialog-button:hover': {
            color: 'white !important',
            border: '2px solid #CCC !important',
        },
        '.mjx-dialog-button > mjx-dialog-icon': {
            display: 'block',
            'background-color': '#AAA',
            border: '1.5px solid',
            'border-radius': '18px',
            'line-height': 0,
            padding: '8px 0 6px',
        },
        '.mjx-dialog-button > mjx-dialog-icon:hover': {
            'background-color': '#CCC !important',
        },
        'mjx-dialog-close': {
            right: '6px',
            'font-size': '20px;',
        },
        'mjx-dialog-help': {
            left: '6px',
            'font-size': '14px;',
            'font-weight': 'bold',
        },
        '.mjx-dialog-help mjx-dialog-help': {
            display: 'none',
        },
        'mjx-dialog kbd': {
            display: 'inline-block',
            padding: '3px 5px',
            'font-size': '11px',
            'line-height': '10px',
            color: '#444d56',
            'vertical-align': 'middle',
            'background-color': '#fafbfc',
            border: 'solid 1.5px #c6cbd1',
            'border-bottom-color': '#959da5',
            'border-radius': '3px',
            'box-shadow': 'inset -.5px -1px 0 #959da5',
        },
        'mjx-dialog-drag[data-drag="top"]': {
            height: '5px',
            position: 'absolute',
            top: '-3px',
            left: '10px',
            right: '10px',
            cursor: 'ns-resize',
        },
        'mjx-dialog-drag[data-drag="bottom"]': {
            height: '5px',
            position: 'absolute',
            bottom: '-3px',
            left: '10px',
            right: '10px',
            cursor: 'ns-resize',
        },
        'mjx-dialog-drag[data-drag="left"]': {
            width: '5px',
            position: 'absolute',
            left: '-3px',
            top: '10px',
            bottom: '10px',
            cursor: 'ew-resize',
        },
        'mjx-dialog-drag[data-drag="right"]': {
            width: '5px',
            position: 'absolute',
            right: '-3px',
            top: '10px',
            bottom: '10px',
            cursor: 'ew-resize',
        },
        'mjx-dialog-drag[data-drag="topleft"]': {
            width: '13px',
            height: '13px',
            position: 'absolute',
            left: '-3px',
            top: '-3px',
            cursor: 'nwse-resize',
        },
        'mjx-dialog-drag[data-drag="topright"]': {
            width: '13px',
            height: '13px',
            position: 'absolute',
            right: '-3px',
            top: '-3px',
            cursor: 'nesw-resize',
        },
        'mjx-dialog-drag[data-drag="botleft"]': {
            width: '13px',
            height: '13px',
            position: 'absolute',
            left: '-3px',
            bottom: '-3px',
            cursor: 'nesw-resize',
        },
        'mjx-dialog-drag[data-drag="botright"]': {
            width: '13px',
            height: '13px',
            position: 'absolute',
            right: '-3px',
            bottom: '-3px',
            cursor: 'nwse-resize',
        },
        '@media (prefers-color-scheme: dark)': {
            '.mjx-dialog': {
                'background-color': '#303030',
                'box-shadow': '0px 10px 20px #000',
                border: '3px outset #7C7C7C',
            },
            'mjx-dialog': {
                color: '#E0E0E0',
            },
            'mjx-dialog > div': {
                border: '2px inset #7C7C7C',
                'background-color': '#222025',
            },
            '.mjx-dialog a[href]': {
                color: '#86A7F5',
            },
            '.mjx-dialog a[href]:visited': {
                color: '#DD98E2',
            },
            'mjx-dialog kbd': {
                color: '#F8F8F8',
                'background-color': '#545454',
                border: 'solid 1.5px #7A7C7E',
                'border-bottom-color': '#707070',
                'box-shadow': 'inset -.5px -1px 0 #818589',
            },
            '.mjx-dialog-button': {
                border: '2px solid #686868',
                color: '#A4A4A4',
            },
            '.mjx-dialog-button:hover': {
                color: '#CBCBCB !important',
                border: '2px solid #888888 !important',
            },
            '.mjx-dialog-button > mjx-dialog-icon': {
                'background-color': '#646464',
            },
            '.mjx-dialog-button > mjx-dialog-icon:hover': {
                'background-color': '#888888 !important',
            },
        },
    };
    DraggableDialog.helpMessage = "\n    <p>The dialog boxes in MathJax are movable and sizeable.</p>\n\n    <p>For mouse users, dragging any of the edges will enlarge or shrink\n    the dialog box by moving that side.  Dragging any of the corners\n    changes the two sides that meet at that corner.  Dragging elsewhere on\n    the dialog frame will move the dialog without changing its size.</p>\n\n    <p>For keyboard users, there are two ways to adjust the position\n    and size of the dialog box.  The first is to hold the\n    <kbd>Alt</kbd> or <kbd>Option</kbd> key and press any of the arrow\n    keys to move the dialog box in the given direction.  Hold the\n    <kbd>Win</kbd> or <kbd>Command</kbd> key and press any of the\n    arrow keys to enlarge or shrink the dialog box.  Left and right\n    move the right-hand edge of the dialog, while up and down move the\n    bottom edge of the dialog.\n    </p>\n\n    <p>For some users, holding two keys down at once may be difficult,\n    so the second way is to press the <kbd>m</kbd> to start \"move\"\n    mode, then use the arrow keys to move the dialog box in the given\n    direction.  Press <kbd>m</kbd> again to stop moving the dialog.\n    Similarly, press <kbd>s</kbd> to start and stop \"sizing\" mode,\n    where the arrows will change the size of the dialog box.</p>\n\n    <p>Holding a <kbd>shift</kbd> key along with the arrow key will\n    make larger changes in the size or position, for either method\n    described above.</p>\n\n    <p>Use <kbd>Tab</kbd> to move among the text, buttons, and links\n    within the dialog.  The <kbd>Enter</kbd> or <kbd>Space</kbd> key\n    activates the focused item.  The <kbd>Escape</kbd> key closes the\n    dialog, as does clicking outside the dialog box, or clicking the\n    \"\u00D7\" icon in the upper right-hand corner of the dialog.</p>\n  ";
    return DraggableDialog;
}());
exports.DraggableDialog = DraggableDialog;
