// @mathjax/src/cjs/ui/dialog/SelectionDialog.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SelectionDialog = void 0;
var InfoDialog_js_1 = require("./InfoDialog.js");
var selection_box_js_1 = require("#menu/selection_box.js");
var SelectionDialog = (function (_super) {
    __extends(SelectionDialog, _super);
    function SelectionDialog(title, signature, selections, order, grid, menu) {
        var _this = _super.call(this, title, signature, order, grid) || this;
        _this.attachMenu(menu);
        var factory = menu.factory;
        _this.selections = selections.map(function (x) {
            return factory.get('selectionMenu')(factory, x, _this);
        });
        return _this;
    }
    SelectionDialog.prototype.post = function () {
        var jax = Array.from(Object.values(this.menu.jax)).filter(function (j) { return !!j; })[0];
        var dialog = new InfoDialog_js_1.InfoDialog({
            title: this.title,
            message: '',
            adaptor: jax.adaptor,
            styles: {
                'mjx-dialog > div': {
                    padding: 0,
                },
            },
        });
        dialog.attach();
        this.contentDiv = dialog.content;
        this.display();
    };
    SelectionDialog.prototype.display = function () {
        var THIS = this;
        THIS.order();
        if (!this.selections.length) {
            return;
        }
        var outerDivs = [];
        var maxWidth = 0;
        var balancedColumn = [];
        var chunks = THIS.getChunkSize(this.selections.length);
        var _loop_1 = function (i) {
            var sels = this_1.selections.slice(i, i + chunks);
            var _a = __read(THIS.rowDiv(sels), 4), div = _a[0], width = _a[1], height = _a[2], column = _a[3];
            outerDivs.push(div);
            maxWidth = Math.max(maxWidth, width);
            sels.forEach(function (sel) { return (sel.html.style.height = height + 'px'); });
            balancedColumn = THIS.combineColumn(balancedColumn, column);
        };
        var this_1 = this;
        for (var i = 0; i < this.selections.length; i += chunks) {
            _loop_1(i);
        }
        if (THIS._balanced) {
            THIS.balanceColumn(outerDivs, balancedColumn);
            maxWidth = balancedColumn.reduce(function (x, y) { return x + y - 2; }, 20);
        }
        outerDivs.forEach(function (div) { return (div.style.width = maxWidth + 'px'); });
    };
    return SelectionDialog;
}(selection_box_js_1.SelectionBox));
exports.SelectionDialog = SelectionDialog;
