// @mathjax/src/cjs/core/MmlTree/OperatorDictionary.js
"use strict";
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.OPTABLE = exports.MMLSPACING = exports.RANGES = exports.MO = void 0;
exports.OPDEF = OPDEF;
exports.getRange = getRange;
var MmlNode_js_1 = require("./MmlNode.js");
function OPDEF(lspace, rspace, texClass, properties) {
    if (texClass === void 0) { texClass = MmlNode_js_1.TEXCLASS.BIN; }
    if (properties === void 0) { properties = null; }
    return [lspace, rspace, texClass, properties];
}
exports.MO = {
    REL: OPDEF(5, 5, MmlNode_js_1.TEXCLASS.REL),
    WIDEREL: OPDEF(5, 5, MmlNode_js_1.TEXCLASS.REL, { accent: true, stretchy: true }),
    BIN4: OPDEF(4, 4, MmlNode_js_1.TEXCLASS.BIN),
    RELSTRETCH: OPDEF(5, 5, MmlNode_js_1.TEXCLASS.REL, { stretchy: true }),
    ORD: OPDEF(0, 0, MmlNode_js_1.TEXCLASS.ORD),
    BIN3: OPDEF(3, 3, MmlNode_js_1.TEXCLASS.BIN),
    OPEN: OPDEF(0, 0, MmlNode_js_1.TEXCLASS.OPEN, {
        fence: true,
        stretchy: true,
        symmetric: true,
    }),
    CLOSE: OPDEF(0, 0, MmlNode_js_1.TEXCLASS.CLOSE, {
        fence: true,
        stretchy: true,
        symmetric: true,
    }),
    INTEGRAL: OPDEF(3, 3, MmlNode_js_1.TEXCLASS.OP, { largeop: true, symmetric: true }),
    ACCENT: OPDEF(0, 0, MmlNode_js_1.TEXCLASS.ORD, { accent: true }),
    WIDEACCENT: OPDEF(0, 0, MmlNode_js_1.TEXCLASS.ORD, { accent: true, stretchy: true }),
    OP: OPDEF(3, 3, MmlNode_js_1.TEXCLASS.OP, {
        largeop: true,
        movablelimits: true,
        symmetric: true,
    }),
    RELACCENT: OPDEF(5, 5, MmlNode_js_1.TEXCLASS.REL, { accent: true }),
    BIN0: OPDEF(0, 0, MmlNode_js_1.TEXCLASS.BIN),
    BIN5: OPDEF(5, 5, MmlNode_js_1.TEXCLASS.BIN),
    FENCE: OPDEF(0, 0, MmlNode_js_1.TEXCLASS.ORD, {
        fence: true,
        stretchy: true,
        symmetric: true,
    }),
    INNER: OPDEF(1, 1, MmlNode_js_1.TEXCLASS.INNER),
    ORD30: OPDEF(3, 0, MmlNode_js_1.TEXCLASS.ORD),
    NONE: OPDEF(0, 0, MmlNode_js_1.TEXCLASS.NONE),
    ORDSTRETCH0: OPDEF(0, 0, MmlNode_js_1.TEXCLASS.ORD, { stretchy: true }),
    BINSTRETCH0: OPDEF(0, 0, MmlNode_js_1.TEXCLASS.BIN, { stretchy: true }),
    RELSTRETCH0: OPDEF(0, 0, MmlNode_js_1.TEXCLASS.REL, { stretchy: true }),
    CLOSE0: OPDEF(0, 0, MmlNode_js_1.TEXCLASS.CLOSE, { fence: true }),
    ORD3: OPDEF(3, 3, MmlNode_js_1.TEXCLASS.ORD),
    PUNCT03: OPDEF(0, 3, MmlNode_js_1.TEXCLASS.PUNCT, { linebreakstyle: 'after' }),
    OPEN0: OPDEF(0, 0, MmlNode_js_1.TEXCLASS.OPEN, { fence: true }),
    STRETCH4: OPDEF(4, 4, MmlNode_js_1.TEXCLASS.BIN, { stretchy: true }),
};
exports.RANGES = [
    [0x0020, 0x007f, MmlNode_js_1.TEXCLASS.REL, 'mo'],
    [0x00a0, 0x00bf, MmlNode_js_1.TEXCLASS.ORD, 'mo'],
    [0x00c0, 0x024f, MmlNode_js_1.TEXCLASS.ORD, 'mi'],
    [0x02b0, 0x036f, MmlNode_js_1.TEXCLASS.ORD, 'mo'],
    [0x0370, 0x1a20, MmlNode_js_1.TEXCLASS.ORD, 'mi'],
    [0x1ab0, 0x1aff, MmlNode_js_1.TEXCLASS.ORD, 'mo'],
    [0x1b00, 0x1dbf, MmlNode_js_1.TEXCLASS.ORD, 'mi'],
    [0x1dc0, 0x1dff, MmlNode_js_1.TEXCLASS.ORD, 'mo'],
    [0x1e00, 0x1fff, MmlNode_js_1.TEXCLASS.ORD, 'mi'],
    [0x2000, 0x206f, MmlNode_js_1.TEXCLASS.ORD, 'mo'],
    [0x2070, 0x209f, MmlNode_js_1.TEXCLASS.ORD, 'mo'],
    [0x2100, 0x214f, MmlNode_js_1.TEXCLASS.ORD, 'mi'],
    [0x2150, 0x218f, MmlNode_js_1.TEXCLASS.ORD, 'mn'],
    [0x2190, 0x21ff, MmlNode_js_1.TEXCLASS.REL, 'mo'],
    [0x2200, 0x22ff, MmlNode_js_1.TEXCLASS.BIN, 'mo'],
    [0x2300, 0x23ff, MmlNode_js_1.TEXCLASS.ORD, 'mo'],
    [0x2460, 0x24ff, MmlNode_js_1.TEXCLASS.ORD, 'mn'],
    [0x2500, 0x27ef, MmlNode_js_1.TEXCLASS.ORD, 'mo'],
    [0x27f0, 0x27ff, MmlNode_js_1.TEXCLASS.REL, 'mo'],
    [0x2800, 0x28ff, MmlNode_js_1.TEXCLASS.ORD, 'mtext'],
    [0x2900, 0x297f, MmlNode_js_1.TEXCLASS.REL, 'mo'],
    [0x2980, 0x29ff, MmlNode_js_1.TEXCLASS.ORD, 'mo'],
    [0x2a00, 0x2aff, MmlNode_js_1.TEXCLASS.BIN, 'mo'],
    [0x2b00, 0x2b2f, MmlNode_js_1.TEXCLASS.ORD, 'mo'],
    [0x2b30, 0x2b4f, MmlNode_js_1.TEXCLASS.REL, 'mo'],
    [0x2b50, 0x2bff, MmlNode_js_1.TEXCLASS.ORD, 'mo'],
    [0x2c00, 0x2de0, MmlNode_js_1.TEXCLASS.ORD, 'mi'],
    [0x2e00, 0x2e7f, MmlNode_js_1.TEXCLASS.ORD, 'mo'],
    [0x2e80, 0x2fdf, MmlNode_js_1.TEXCLASS.ORD, 'mi', 'normal'],
    [0x2ff0, 0x303f, MmlNode_js_1.TEXCLASS.ORD, 'mo'],
    [0x3040, 0xa49f, MmlNode_js_1.TEXCLASS.ORD, 'mi', 'normal'],
    [0xa4d0, 0xa82f, MmlNode_js_1.TEXCLASS.ORD, 'mi'],
    [0xa830, 0xa83f, MmlNode_js_1.TEXCLASS.ORD, 'mn'],
    [0xa840, 0xd7ff, MmlNode_js_1.TEXCLASS.ORD, 'mi'],
    [0xf900, 0xfaff, MmlNode_js_1.TEXCLASS.ORD, 'mi', 'normal'],
    [0xfb00, 0xfdff, MmlNode_js_1.TEXCLASS.ORD, 'mi'],
    [0xfe00, 0xfe6f, MmlNode_js_1.TEXCLASS.ORD, 'mo'],
    [0xfe70, 0x100ff, MmlNode_js_1.TEXCLASS.ORD, 'mi'],
    [0x10100, 0x1018f, MmlNode_js_1.TEXCLASS.ORD, 'mn'],
    [0x10190, 0x123ff, MmlNode_js_1.TEXCLASS.ORD, 'mi', 'normal'],
    [0x12400, 0x1247f, MmlNode_js_1.TEXCLASS.ORD, 'mn'],
    [0x12480, 0x1bc9f, MmlNode_js_1.TEXCLASS.ORD, 'mi', 'normal'],
    [0x1bca0, 0x1d25f, MmlNode_js_1.TEXCLASS.ORD, 'mo'],
    [0x1d360, 0x1d37f, MmlNode_js_1.TEXCLASS.ORD, 'mn'],
    [0x1d400, 0x1d7cd, MmlNode_js_1.TEXCLASS.ORD, 'mi'],
    [0x1d7ce, 0x1d7ff, MmlNode_js_1.TEXCLASS.ORD, 'mn'],
    [0x1df00, 0x1f7ff, MmlNode_js_1.TEXCLASS.ORD, 'mo'],
    [0x1f800, 0x1f8ff, MmlNode_js_1.TEXCLASS.REL, 'mo'],
    [0x1f900, 0x1f9ff, MmlNode_js_1.TEXCLASS.ORD, 'mo'],
    [0x20000, 0x2fa1f, MmlNode_js_1.TEXCLASS.ORD, 'mi', 'normal'],
];
function getRange(text) {
    var e_1, _a;
    var def = exports.OPTABLE.infix[text] || exports.OPTABLE.prefix[text] || exports.OPTABLE.postfix[text];
    if (def) {
        return [0, 0, def[2], 'mo'];
    }
    var n = text.codePointAt(0);
    try {
        for (var RANGES_1 = __values(exports.RANGES), RANGES_1_1 = RANGES_1.next(); !RANGES_1_1.done; RANGES_1_1 = RANGES_1.next()) {
            var range = RANGES_1_1.value;
            if (n <= range[1]) {
                if (n >= range[0]) {
                    return range;
                }
                break;
            }
        }
    }
    catch (e_1_1) { e_1 = { error: e_1_1 }; }
    finally {
        try {
            if (RANGES_1_1 && !RANGES_1_1.done && (_a = RANGES_1.return)) _a.call(RANGES_1);
        }
        finally { if (e_1) throw e_1.error; }
    }
    return [0, 0, MmlNode_js_1.TEXCLASS.REL, 'mo'];
}
exports.MMLSPACING = [
    [0, 0],
    [1, 2],
    [3, 3],
    [4, 4],
    [0, 0],
    [0, 0],
    [0, 3],
    [1, 1],
];
exports.OPTABLE = {
    prefix: {
        '!': exports.MO.ORD,
        '(': exports.MO.OPEN,
        '+': exports.MO.BIN0,
        '-': exports.MO.BIN0,
        '[': exports.MO.OPEN,
        '{': exports.MO.OPEN,
        '|': exports.MO.OPEN,
        '||': exports.MO.BIN0,
        '\u00AC': exports.MO.ORD,
        '\u00B1': exports.MO.BIN0,
        '\u2016': exports.MO.FENCE,
        '\u2018': exports.MO.OPEN0,
        '\u201C': exports.MO.OPEN0,
        '\u2145': exports.MO.ORD30,
        '\u2146': exports.MO.ORD30,
        '\u2200': exports.MO.ORD,
        '\u2201': exports.MO.ORD,
        '\u2202': exports.MO.ORD30,
        '\u2203': exports.MO.ORD,
        '\u2204': exports.MO.ORD,
        '\u2207': exports.MO.ORD,
        '\u220F': exports.MO.OP,
        '\u2210': exports.MO.OP,
        '\u2211': exports.MO.OP,
        '\u2212': exports.MO.BIN0,
        '\u2213': exports.MO.BIN0,
        '\u221A': [3, 0, MmlNode_js_1.TEXCLASS.ORD, { stretchy: true }],
        '\u221B': exports.MO.ORD30,
        '\u221C': exports.MO.ORD30,
        '\u221F': exports.MO.ORD,
        '\u2220': exports.MO.ORD,
        '\u2221': exports.MO.ORD,
        '\u2222': exports.MO.ORD,
        '\u222B': exports.MO.INTEGRAL,
        '\u222C': exports.MO.INTEGRAL,
        '\u222D': exports.MO.INTEGRAL,
        '\u222E': exports.MO.INTEGRAL,
        '\u222F': exports.MO.INTEGRAL,
        '\u2230': exports.MO.INTEGRAL,
        '\u2231': exports.MO.INTEGRAL,
        '\u2232': exports.MO.INTEGRAL,
        '\u2233': exports.MO.INTEGRAL,
        '\u2234': exports.MO.REL,
        '\u2235': exports.MO.REL,
        '\u223C': [0, 0, MmlNode_js_1.TEXCLASS.REL, {}],
        '\u22BE': exports.MO.ORD,
        '\u22BF': exports.MO.ORD,
        '\u22C0': exports.MO.OP,
        '\u22C1': exports.MO.OP,
        '\u22C2': exports.MO.OP,
        '\u22C3': exports.MO.OP,
        '\u2308': exports.MO.OPEN,
        '\u230A': exports.MO.OPEN,
        '\u2310': exports.MO.ORD,
        '\u2319': exports.MO.ORD,
        '\u2772': exports.MO.OPEN,
        '\u2795': exports.MO.ORD,
        '\u2796': exports.MO.ORD,
        '\u27C0': exports.MO.ORD,
        '\u27E6': exports.MO.OPEN,
        '\u27E8': exports.MO.OPEN,
        '\u27EA': exports.MO.OPEN,
        '\u27EC': exports.MO.OPEN,
        '\u27EE': exports.MO.OPEN,
        '\u2980': exports.MO.FENCE,
        '\u2983': exports.MO.OPEN,
        '\u2985': exports.MO.OPEN,
        '\u2987': exports.MO.OPEN,
        '\u2989': exports.MO.OPEN,
        '\u298B': exports.MO.OPEN,
        '\u298D': exports.MO.OPEN,
        '\u298F': exports.MO.OPEN,
        '\u2991': exports.MO.OPEN,
        '\u2993': exports.MO.OPEN,
        '\u2995': exports.MO.OPEN,
        '\u2997': exports.MO.OPEN,
        '\u2999': exports.MO.FENCE,
        '\u299B': exports.MO.ORD,
        '\u299C': exports.MO.ORD,
        '\u299D': exports.MO.ORD,
        '\u299E': exports.MO.ORD,
        '\u299F': exports.MO.ORD,
        '\u29A0': exports.MO.ORD,
        '\u29A1': exports.MO.ORD,
        '\u29A2': exports.MO.ORD,
        '\u29A3': exports.MO.ORD,
        '\u29A4': exports.MO.ORD,
        '\u29A5': exports.MO.ORD,
        '\u29A6': exports.MO.ORD,
        '\u29A7': exports.MO.ORD,
        '\u29A8': exports.MO.ORD,
        '\u29A9': exports.MO.ORD,
        '\u29AA': exports.MO.ORD,
        '\u29AB': exports.MO.ORD,
        '\u29AC': exports.MO.ORD,
        '\u29AD': exports.MO.ORD,
        '\u29AE': exports.MO.ORD,
        '\u29AF': exports.MO.ORD,
        '\u29D8': exports.MO.OPEN,
        '\u29DA': exports.MO.OPEN,
        '\u29FC': exports.MO.OPEN,
        '\u2A00': exports.MO.OP,
        '\u2A01': exports.MO.OP,
        '\u2A02': exports.MO.OP,
        '\u2A03': exports.MO.OP,
        '\u2A04': exports.MO.OP,
        '\u2A05': exports.MO.OP,
        '\u2A06': exports.MO.OP,
        '\u2A07': exports.MO.OP,
        '\u2A08': exports.MO.OP,
        '\u2A09': exports.MO.OP,
        '\u2A0A': exports.MO.OP,
        '\u2A0B': exports.MO.INTEGRAL,
        '\u2A0C': exports.MO.INTEGRAL,
        '\u2A0D': exports.MO.INTEGRAL,
        '\u2A0E': exports.MO.INTEGRAL,
        '\u2A0F': exports.MO.INTEGRAL,
        '\u2A10': exports.MO.INTEGRAL,
        '\u2A11': exports.MO.INTEGRAL,
        '\u2A12': exports.MO.INTEGRAL,
        '\u2A13': exports.MO.INTEGRAL,
        '\u2A14': exports.MO.INTEGRAL,
        '\u2A15': exports.MO.INTEGRAL,
        '\u2A16': exports.MO.INTEGRAL,
        '\u2A17': exports.MO.INTEGRAL,
        '\u2A18': exports.MO.INTEGRAL,
        '\u2A19': exports.MO.INTEGRAL,
        '\u2A1A': exports.MO.INTEGRAL,
        '\u2A1B': exports.MO.INTEGRAL,
        '\u2A1C': exports.MO.INTEGRAL,
        '\u2A1D': exports.MO.OP,
        '\u2A1E': exports.MO.OP,
        '\u2AEC': exports.MO.ORD,
        '\u2AED': exports.MO.ORD,
        '\u2AFC': exports.MO.OP,
        '\u2AFF': exports.MO.OP,
        '\u3008': exports.MO.OPEN,
    },
    postfix: {
        '!!': exports.MO.BIN0,
        '!': exports.MO.CLOSE0,
        '"': exports.MO.ORD,
        '%': exports.MO.ORD,
        '&': exports.MO.ORD,
        '\'': exports.MO.ACCENT,
        ')': exports.MO.CLOSE,
        '++': exports.MO.BIN0,
        '--': exports.MO.BIN0,
        ']': exports.MO.CLOSE,
        '^': exports.MO.WIDEACCENT,
        '_': exports.MO.WIDEACCENT,
        '`': exports.MO.ACCENT,
        '|': exports.MO.CLOSE,
        '||': exports.MO.BIN0,
        '}': exports.MO.CLOSE,
        '~': exports.MO.WIDEACCENT,
        '\u00A8': exports.MO.ACCENT,
        '\u00AF': exports.MO.WIDEACCENT,
        '\u00B0': exports.MO.ACCENT,
        '\u00B2': exports.MO.ORD,
        '\u00B3': exports.MO.ORD,
        '\u00B4': exports.MO.ACCENT,
        '\u00B8': exports.MO.ACCENT,
        '\u00B9': exports.MO.ORD,
        '\u02C6': exports.MO.WIDEACCENT,
        '\u02C7': exports.MO.WIDEACCENT,
        '\u02C9': exports.MO.WIDEACCENT,
        '\u02CA': exports.MO.ACCENT,
        '\u02CB': exports.MO.ACCENT,
        '\u02CD': exports.MO.WIDEACCENT,
        '\u02D8': exports.MO.ACCENT,
        '\u02D9': exports.MO.ACCENT,
        '\u02DA': exports.MO.ACCENT,
        '\u02DC': exports.MO.WIDEACCENT,
        '\u02DD': exports.MO.ACCENT,
        '\u02F7': exports.MO.WIDEACCENT,
        '\u0302': exports.MO.WIDEACCENT,
        '\u0311': exports.MO.ACCENT,
        '\u2016': exports.MO.FENCE,
        '\u2019': exports.MO.CLOSE0,
        '\u201A': exports.MO.ORD,
        '\u201B': exports.MO.ORD,
        '\u201D': exports.MO.CLOSE0,
        '\u201E': exports.MO.ORD,
        '\u201F': exports.MO.ORD,
        '\u2032': exports.MO.ORD,
        '\u2033': exports.MO.ORD,
        '\u2034': exports.MO.ORD,
        '\u2035': exports.MO.ORD,
        '\u2036': exports.MO.ORD,
        '\u2037': exports.MO.ORD,
        '\u203E': exports.MO.WIDEACCENT,
        '\u2057': exports.MO.ORD,
        '\u20DB': exports.MO.ACCENT,
        '\u20DC': exports.MO.ACCENT,
        '\u2309': exports.MO.CLOSE,
        '\u230B': exports.MO.CLOSE,
        '\u2322': exports.MO.RELSTRETCH0,
        '\u2323': exports.MO.RELSTRETCH0,
        '\u23B4': exports.MO.WIDEACCENT,
        '\u23B5': exports.MO.WIDEACCENT,
        '\u23CD': exports.MO.ORD,
        '\u23DC': exports.MO.WIDEACCENT,
        '\u23DD': exports.MO.WIDEACCENT,
        '\u23DE': exports.MO.WIDEACCENT,
        '\u23DF': exports.MO.WIDEACCENT,
        '\u23E0': exports.MO.WIDEACCENT,
        '\u23E1': exports.MO.WIDEACCENT,
        '\u2773': exports.MO.CLOSE,
        '\u27E7': exports.MO.CLOSE,
        '\u27E9': exports.MO.CLOSE,
        '\u27EB': exports.MO.CLOSE,
        '\u27ED': exports.MO.CLOSE,
        '\u27EF': exports.MO.CLOSE,
        '\u2980': exports.MO.FENCE,
        '\u2984': exports.MO.CLOSE,
        '\u2986': exports.MO.CLOSE,
        '\u2988': exports.MO.CLOSE,
        '\u298A': exports.MO.CLOSE,
        '\u298C': exports.MO.CLOSE,
        '\u298E': exports.MO.CLOSE,
        '\u2990': exports.MO.CLOSE,
        '\u2992': exports.MO.CLOSE,
        '\u2994': exports.MO.CLOSE,
        '\u2996': exports.MO.CLOSE,
        '\u2998': exports.MO.CLOSE,
        '\u2999': exports.MO.FENCE,
        '\u29D9': exports.MO.CLOSE,
        '\u29DB': exports.MO.CLOSE,
        '\u29FD': exports.MO.CLOSE,
        '\u3009': exports.MO.CLOSE,
        "\uD83B\uDEF0": exports.MO.BINSTRETCH0,
        "\uD83B\uDEF1": exports.MO.BINSTRETCH0,
    },
    infix: {
        '!': exports.MO.ORD,
        '!=': exports.MO.BIN5,
        '#': exports.MO.ORD,
        '$': exports.MO.ORD,
        '%': exports.MO.ORD3,
        '&&': exports.MO.BIN4,
        '**': exports.MO.BIN3,
        '*': exports.MO.BIN3,
        '*=': exports.MO.BIN5,
        '+': exports.MO.BIN4,
        '+=': exports.MO.BIN5,
        ',': exports.MO.PUNCT03,
        '': exports.MO.ORD,
        '-': exports.MO.BIN4,
        '-=': exports.MO.BIN5,
        '->': exports.MO.BIN5,
        '.': exports.MO.ORD3,
        '..': exports.MO.BIN3,
        '...': exports.MO.INNER,
        '/': [4, 4, MmlNode_js_1.TEXCLASS.ORD, {}],
        '//': exports.MO.BIN5,
        '/=': exports.MO.BIN5,
        ':': [0, 3, MmlNode_js_1.TEXCLASS.REL, {}],
        ':=': exports.MO.BIN5,
        ';': exports.MO.PUNCT03,
        '<': exports.MO.REL,
        '<=': exports.MO.REL,
        '<>': [3, 3, MmlNode_js_1.TEXCLASS.REL, {}],
        '=': exports.MO.REL,
        '==': exports.MO.REL,
        '>': exports.MO.REL,
        '>=': exports.MO.REL,
        '?': [3, 3, MmlNode_js_1.TEXCLASS.CLOSE, { fence: true }],
        '@': exports.MO.ORD3,
        '\\': exports.MO.ORD,
        '^': [3, 3, MmlNode_js_1.TEXCLASS.ORD, { accent: true, stretchy: true }],
        '_': exports.MO.WIDEACCENT,
        '|': [5, 5, MmlNode_js_1.TEXCLASS.ORD, {}],
        '||': exports.MO.BIN5,
        '\u00B1': exports.MO.BIN4,
        '\u00B7': exports.MO.BIN3,
        '\u00D7': exports.MO.BIN3,
        '\u00F7': exports.MO.BIN4,
        '\u02B9': exports.MO.ORD,
        '\u0300': exports.MO.ACCENT,
        '\u0301': exports.MO.ACCENT,
        '\u0303': exports.MO.WIDEACCENT,
        '\u0304': exports.MO.ACCENT,
        '\u0306': exports.MO.ACCENT,
        '\u0307': exports.MO.ACCENT,
        '\u0308': exports.MO.ACCENT,
        '\u030C': exports.MO.ACCENT,
        '\u0332': exports.MO.WIDEACCENT,
        '\u0338': exports.MO.REL,
        '\u03F6': exports.MO.REL,
        '\u2015': exports.MO.ORDSTRETCH0,
        '\u2017': exports.MO.ORDSTRETCH0,
        '\u2020': exports.MO.BIN3,
        '\u2021': exports.MO.BIN3,
        '\u2022': exports.MO.BIN3,
        '\u2026': exports.MO.INNER,
        '\u2043': exports.MO.BIN3,
        '\u2044': exports.MO.STRETCH4,
        '\u2061': exports.MO.NONE,
        '\u2062': exports.MO.NONE,
        '\u2063': [0, 0, MmlNode_js_1.TEXCLASS.NONE, { linebreakstyle: 'after' }],
        '\u2064': exports.MO.NONE,
        '\u20D7': exports.MO.ACCENT,
        '\u2111': exports.MO.ORD,
        '\u2113': exports.MO.ORD,
        '\u2118': exports.MO.ORD,
        '\u211C': exports.MO.ORD,
        '\u2190': exports.MO.WIDEREL,
        '\u2191': exports.MO.RELSTRETCH,
        '\u2192': exports.MO.WIDEREL,
        '\u2193': exports.MO.RELSTRETCH,
        '\u2194': exports.MO.WIDEREL,
        '\u2195': exports.MO.RELSTRETCH,
        '\u2196': exports.MO.REL,
        '\u2197': exports.MO.REL,
        '\u2198': exports.MO.REL,
        '\u2199': exports.MO.REL,
        '\u219A': exports.MO.WIDEREL,
        '\u219B': exports.MO.WIDEREL,
        '\u219C': exports.MO.WIDEREL,
        '\u219D': exports.MO.WIDEREL,
        '\u219E': exports.MO.WIDEREL,
        '\u219F': exports.MO.RELSTRETCH,
        '\u21A0': exports.MO.WIDEREL,
        '\u21A1': exports.MO.RELSTRETCH,
        '\u21A2': exports.MO.WIDEREL,
        '\u21A3': exports.MO.WIDEREL,
        '\u21A4': exports.MO.WIDEREL,
        '\u21A5': exports.MO.RELSTRETCH,
        '\u21A6': exports.MO.WIDEREL,
        '\u21A7': exports.MO.RELSTRETCH,
        '\u21A8': exports.MO.RELSTRETCH,
        '\u21A9': exports.MO.WIDEREL,
        '\u21AA': exports.MO.WIDEREL,
        '\u21AB': exports.MO.WIDEREL,
        '\u21AC': exports.MO.WIDEREL,
        '\u21AD': exports.MO.WIDEREL,
        '\u21AE': exports.MO.WIDEREL,
        '\u21AF': exports.MO.REL,
        '\u21B0': exports.MO.RELSTRETCH,
        '\u21B1': exports.MO.RELSTRETCH,
        '\u21B2': exports.MO.RELSTRETCH,
        '\u21B3': exports.MO.RELSTRETCH,
        '\u21B4': exports.MO.RELSTRETCH,
        '\u21B5': exports.MO.RELSTRETCH,
        '\u21B6': exports.MO.REL,
        '\u21B7': exports.MO.REL,
        '\u21B8': exports.MO.REL,
        '\u21B9': exports.MO.WIDEREL,
        '\u21BA': exports.MO.REL,
        '\u21BB': exports.MO.REL,
        '\u21BC': exports.MO.WIDEREL,
        '\u21BD': exports.MO.WIDEREL,
        '\u21BE': exports.MO.RELSTRETCH,
        '\u21BF': exports.MO.RELSTRETCH,
        '\u21C0': exports.MO.WIDEREL,
        '\u21C1': exports.MO.WIDEREL,
        '\u21C2': exports.MO.RELSTRETCH,
        '\u21C3': exports.MO.RELSTRETCH,
        '\u21C4': exports.MO.WIDEREL,
        '\u21C5': exports.MO.RELSTRETCH,
        '\u21C6': exports.MO.WIDEREL,
        '\u21C7': exports.MO.WIDEREL,
        '\u21C8': exports.MO.RELSTRETCH,
        '\u21C9': exports.MO.WIDEREL,
        '\u21CA': exports.MO.RELSTRETCH,
        '\u21CB': exports.MO.WIDEREL,
        '\u21CC': exports.MO.WIDEREL,
        '\u21CD': exports.MO.WIDEREL,
        '\u21CE': exports.MO.WIDEREL,
        '\u21CF': exports.MO.WIDEREL,
        '\u21D0': exports.MO.WIDEREL,
        '\u21D1': exports.MO.RELSTRETCH,
        '\u21D2': exports.MO.WIDEREL,
        '\u21D3': exports.MO.RELSTRETCH,
        '\u21D4': exports.MO.WIDEREL,
        '\u21D5': exports.MO.RELSTRETCH,
        '\u21D6': exports.MO.REL,
        '\u21D7': exports.MO.REL,
        '\u21D8': exports.MO.REL,
        '\u21D9': exports.MO.REL,
        '\u21DA': exports.MO.WIDEREL,
        '\u21DB': exports.MO.WIDEREL,
        '\u21DC': exports.MO.WIDEREL,
        '\u21DD': exports.MO.WIDEREL,
        '\u21DE': exports.MO.RELSTRETCH,
        '\u21DF': exports.MO.RELSTRETCH,
        '\u21E0': exports.MO.WIDEREL,
        '\u21E1': exports.MO.RELSTRETCH,
        '\u21E2': exports.MO.WIDEREL,
        '\u21E3': exports.MO.RELSTRETCH,
        '\u21E4': exports.MO.WIDEREL,
        '\u21E5': exports.MO.WIDEREL,
        '\u21E6': exports.MO.WIDEREL,
        '\u21E7': exports.MO.RELSTRETCH,
        '\u21E8': exports.MO.WIDEREL,
        '\u21E9': exports.MO.RELSTRETCH,
        '\u21EA': exports.MO.RELSTRETCH,
        '\u21EB': exports.MO.RELSTRETCH,
        '\u21EC': exports.MO.RELSTRETCH,
        '\u21ED': exports.MO.RELSTRETCH,
        '\u21EE': exports.MO.RELSTRETCH,
        '\u21EF': exports.MO.RELSTRETCH,
        '\u21F0': exports.MO.WIDEREL,
        '\u21F1': exports.MO.REL,
        '\u21F2': exports.MO.REL,
        '\u21F3': exports.MO.RELSTRETCH,
        '\u21F4': exports.MO.WIDEREL,
        '\u21F5': exports.MO.RELSTRETCH,
        '\u21F6': exports.MO.WIDEREL,
        '\u21F7': exports.MO.WIDEREL,
        '\u21F8': exports.MO.WIDEREL,
        '\u21F9': exports.MO.WIDEREL,
        '\u21FA': exports.MO.WIDEREL,
        '\u21FB': exports.MO.WIDEREL,
        '\u21FC': exports.MO.WIDEREL,
        '\u21FD': exports.MO.WIDEREL,
        '\u21FE': exports.MO.WIDEREL,
        '\u21FF': exports.MO.WIDEREL,
        '\u2205': exports.MO.ORD,
        '\u2206': exports.MO.ORD,
        '\u2208': exports.MO.REL,
        '\u2209': exports.MO.REL,
        '\u220A': exports.MO.REL,
        '\u220B': exports.MO.REL,
        '\u220C': exports.MO.REL,
        '\u220D': exports.MO.REL,
        '\u2212': exports.MO.BIN4,
        '\u2213': exports.MO.BIN4,
        '\u2214': exports.MO.BIN4,
        '\u2215': exports.MO.STRETCH4,
        '\u2216': exports.MO.BIN4,
        '\u2217': exports.MO.BIN3,
        '\u2218': exports.MO.BIN3,
        '\u2219': exports.MO.BIN3,
        '\u221D': exports.MO.REL,
        '\u221E': exports.MO.ORD,
        '\u2223': exports.MO.REL,
        '\u2224': exports.MO.REL,
        '\u2225': exports.MO.REL,
        '\u2226': exports.MO.REL,
        '\u2227': exports.MO.BIN4,
        '\u2228': exports.MO.BIN4,
        '\u2229': exports.MO.BIN4,
        '\u222A': exports.MO.BIN4,
        '\u2236': exports.MO.BIN4,
        '\u2237': exports.MO.REL,
        '\u2238': exports.MO.BIN4,
        '\u2239': exports.MO.REL,
        '\u223A': exports.MO.REL,
        '\u223B': exports.MO.REL,
        '\u223C': exports.MO.REL,
        '\u223D': exports.MO.REL,
        '\u223E': exports.MO.REL,
        '\u2240': exports.MO.BIN3,
        '\u2241': exports.MO.REL,
        '\u2242': exports.MO.REL,
        '\u2242\u0338': exports.MO.REL,
        '\u2243': exports.MO.REL,
        '\u2244': exports.MO.REL,
        '\u2245': exports.MO.REL,
        '\u2246': exports.MO.REL,
        '\u2247': exports.MO.REL,
        '\u2248': exports.MO.REL,
        '\u2249': exports.MO.REL,
        '\u224A': exports.MO.REL,
        '\u224B': exports.MO.REL,
        '\u224C': exports.MO.REL,
        '\u224D': exports.MO.REL,
        '\u224E': exports.MO.REL,
        '\u224F': exports.MO.REL,
        '\u2250': exports.MO.REL,
        '\u2251': exports.MO.REL,
        '\u2252': exports.MO.REL,
        '\u2253': exports.MO.REL,
        '\u2254': exports.MO.REL,
        '\u2255': exports.MO.REL,
        '\u2256': exports.MO.REL,
        '\u2257': exports.MO.REL,
        '\u2258': exports.MO.REL,
        '\u2259': exports.MO.REL,
        '\u225A': exports.MO.REL,
        '\u225B': exports.MO.REL,
        '\u225C': exports.MO.REL,
        '\u225D': exports.MO.REL,
        '\u225E': exports.MO.REL,
        '\u225F': exports.MO.REL,
        '\u2260': exports.MO.REL,
        '\u2261': exports.MO.REL,
        '\u2262': exports.MO.REL,
        '\u2263': exports.MO.REL,
        '\u2264': exports.MO.REL,
        '\u2265': exports.MO.REL,
        '\u2266': exports.MO.REL,
        '\u2266\u0338': exports.MO.REL,
        '\u2267': exports.MO.REL,
        '\u2267\u0338': exports.MO.REL,
        '\u2268': exports.MO.REL,
        '\u2269': exports.MO.REL,
        '\u226A': exports.MO.REL,
        '\u226A\u0338': exports.MO.REL,
        '\u226B': exports.MO.REL,
        '\u226B\u0338': exports.MO.REL,
        '\u226C': exports.MO.REL,
        '\u226D': exports.MO.REL,
        '\u226E': exports.MO.REL,
        '\u226F': exports.MO.REL,
        '\u2270': exports.MO.REL,
        '\u2271': exports.MO.REL,
        '\u2272': exports.MO.REL,
        '\u2273': exports.MO.REL,
        '\u2274': exports.MO.REL,
        '\u2275': exports.MO.REL,
        '\u2276': exports.MO.REL,
        '\u2277': exports.MO.REL,
        '\u2278': exports.MO.REL,
        '\u2279': exports.MO.REL,
        '\u227A': exports.MO.REL,
        '\u227B': exports.MO.REL,
        '\u227C': exports.MO.REL,
        '\u227D': exports.MO.REL,
        '\u227E': exports.MO.REL,
        '\u227E\u0338': exports.MO.REL,
        '\u227F': exports.MO.REL,
        '\u227F\u0338': exports.MO.REL,
        '\u2280': exports.MO.REL,
        '\u2281': exports.MO.REL,
        '\u2282': exports.MO.REL,
        '\u2283': exports.MO.REL,
        '\u2284': exports.MO.REL,
        '\u2285': exports.MO.REL,
        '\u2286': exports.MO.REL,
        '\u2287': exports.MO.REL,
        '\u2288': exports.MO.REL,
        '\u2289': exports.MO.REL,
        '\u228A': exports.MO.REL,
        '\u228B': exports.MO.REL,
        '\u228C': exports.MO.BIN4,
        '\u228D': exports.MO.BIN4,
        '\u228E': exports.MO.BIN4,
        '\u228F': exports.MO.REL,
        '\u228F\u0338': exports.MO.REL,
        '\u2290': exports.MO.REL,
        '\u2290\u0338': exports.MO.REL,
        '\u2291': exports.MO.REL,
        '\u2292': exports.MO.REL,
        '\u2293': exports.MO.BIN4,
        '\u2294': exports.MO.BIN4,
        '\u2295': exports.MO.BIN4,
        '\u2296': exports.MO.BIN4,
        '\u2297': exports.MO.BIN3,
        '\u2298': exports.MO.BIN4,
        '\u2299': exports.MO.BIN3,
        '\u229A': exports.MO.BIN3,
        '\u229B': exports.MO.BIN3,
        '\u229C': exports.MO.REL,
        '\u229D': exports.MO.BIN4,
        '\u229E': exports.MO.BIN4,
        '\u229F': exports.MO.BIN4,
        '\u22A0': exports.MO.BIN3,
        '\u22A1': exports.MO.BIN3,
        '\u22A2': exports.MO.REL,
        '\u22A3': exports.MO.REL,
        '\u22A4': exports.MO.ORD,
        '\u22A5': exports.MO.ORD,
        '\u22A6': exports.MO.REL,
        '\u22A7': exports.MO.REL,
        '\u22A8': exports.MO.REL,
        '\u22A9': exports.MO.REL,
        '\u22AA': exports.MO.REL,
        '\u22AB': exports.MO.REL,
        '\u22AC': exports.MO.REL,
        '\u22AD': exports.MO.REL,
        '\u22AE': exports.MO.REL,
        '\u22AF': exports.MO.REL,
        '\u22B0': exports.MO.REL,
        '\u22B1': exports.MO.REL,
        '\u22B2': exports.MO.REL,
        '\u22B3': exports.MO.REL,
        '\u22B4': exports.MO.REL,
        '\u22B5': exports.MO.REL,
        '\u22B6': exports.MO.REL,
        '\u22B7': exports.MO.REL,
        '\u22B8': exports.MO.REL,
        '\u22BA': exports.MO.BIN3,
        '\u22BB': exports.MO.BIN4,
        '\u22BC': exports.MO.BIN4,
        '\u22BD': exports.MO.BIN4,
        '\u22C4': exports.MO.BIN3,
        '\u22C5': exports.MO.BIN3,
        '\u22C6': exports.MO.BIN3,
        '\u22C7': exports.MO.BIN3,
        '\u22C8': exports.MO.REL,
        '\u22C9': exports.MO.BIN3,
        '\u22CA': exports.MO.BIN3,
        '\u22CB': exports.MO.BIN3,
        '\u22CC': exports.MO.BIN3,
        '\u22CD': exports.MO.REL,
        '\u22CE': exports.MO.BIN4,
        '\u22CF': exports.MO.BIN4,
        '\u22D0': exports.MO.REL,
        '\u22D1': exports.MO.REL,
        '\u22D2': exports.MO.BIN4,
        '\u22D3': exports.MO.BIN4,
        '\u22D4': exports.MO.REL,
        '\u22D5': exports.MO.REL,
        '\u22D6': exports.MO.REL,
        '\u22D7': exports.MO.REL,
        '\u22D8': exports.MO.REL,
        '\u22D9': exports.MO.REL,
        '\u22DA': exports.MO.REL,
        '\u22DB': exports.MO.REL,
        '\u22DC': exports.MO.REL,
        '\u22DD': exports.MO.REL,
        '\u22DE': exports.MO.REL,
        '\u22DF': exports.MO.REL,
        '\u22E0': exports.MO.REL,
        '\u22E1': exports.MO.REL,
        '\u22E2': exports.MO.REL,
        '\u22E3': exports.MO.REL,
        '\u22E4': exports.MO.REL,
        '\u22E5': exports.MO.REL,
        '\u22E6': exports.MO.REL,
        '\u22E7': exports.MO.REL,
        '\u22E8': exports.MO.REL,
        '\u22E9': exports.MO.REL,
        '\u22EA': exports.MO.REL,
        '\u22EB': exports.MO.REL,
        '\u22EC': exports.MO.REL,
        '\u22ED': exports.MO.REL,
        '\u22EE': exports.MO.ORD,
        '\u22EF': exports.MO.INNER,
        '\u22F0': exports.MO.INNER,
        '\u22F1': exports.MO.INNER,
        '\u22F2': exports.MO.REL,
        '\u22F3': exports.MO.REL,
        '\u22F4': exports.MO.REL,
        '\u22F5': exports.MO.REL,
        '\u22F6': exports.MO.REL,
        '\u22F7': exports.MO.REL,
        '\u22F8': exports.MO.REL,
        '\u22F9': exports.MO.REL,
        '\u22FA': exports.MO.REL,
        '\u22FB': exports.MO.REL,
        '\u22FC': exports.MO.REL,
        '\u22FD': exports.MO.REL,
        '\u22FE': exports.MO.REL,
        '\u22FF': exports.MO.REL,
        '\u2301': exports.MO.REL,
        '\u2305': exports.MO.BIN3,
        '\u2306': exports.MO.BIN3,
        '\u2329': exports.MO.OPEN,
        '\u232A': exports.MO.CLOSE,
        '\u237C': exports.MO.REL,
        '\u238B': exports.MO.REL,
        '\u23AA': exports.MO.ORD,
        '\u23AF': exports.MO.ORDSTRETCH0,
        '\u23B0': exports.MO.OPEN,
        '\u23B1': exports.MO.CLOSE,
        '\u2500': exports.MO.ORD,
        '\u25B3': exports.MO.BIN3,
        '\u25B5': exports.MO.BIN3,
        '\u25B9': exports.MO.BIN3,
        '\u25BD': exports.MO.BIN3,
        '\u25BF': exports.MO.BIN3,
        '\u25C3': exports.MO.BIN3,
        '\u25EF': exports.MO.BIN3,
        '\u2660': exports.MO.ORD,
        '\u2661': exports.MO.ORD,
        '\u2662': exports.MO.ORD,
        '\u2663': exports.MO.ORD,
        '\u266D': exports.MO.ORD,
        '\u266E': exports.MO.ORD,
        '\u266F': exports.MO.ORD,
        '\u2758': [5, 5, MmlNode_js_1.TEXCLASS.REL, { stretchy: true, symmetric: true }],
        '\u2794': exports.MO.WIDEREL,
        '\u2795': exports.MO.BIN4,
        '\u2796': exports.MO.BIN4,
        '\u2797': exports.MO.BIN4,
        '\u2798': exports.MO.REL,
        '\u2799': exports.MO.WIDEREL,
        '\u279A': exports.MO.REL,
        '\u279B': exports.MO.WIDEREL,
        '\u279C': exports.MO.WIDEREL,
        '\u279D': exports.MO.WIDEREL,
        '\u279E': exports.MO.WIDEREL,
        '\u279F': exports.MO.WIDEREL,
        '\u27A0': exports.MO.WIDEREL,
        '\u27A1': exports.MO.WIDEREL,
        '\u27A5': exports.MO.WIDEREL,
        '\u27A6': exports.MO.WIDEREL,
        '\u27A7': exports.MO.RELACCENT,
        '\u27A8': exports.MO.WIDEREL,
        '\u27A9': exports.MO.WIDEREL,
        '\u27AA': exports.MO.WIDEREL,
        '\u27AB': exports.MO.WIDEREL,
        '\u27AC': exports.MO.WIDEREL,
        '\u27AD': exports.MO.WIDEREL,
        '\u27AE': exports.MO.WIDEREL,
        '\u27AF': exports.MO.WIDEREL,
        '\u27B1': exports.MO.WIDEREL,
        '\u27B2': exports.MO.RELACCENT,
        '\u27B3': exports.MO.WIDEREL,
        '\u27B4': exports.MO.REL,
        '\u27B5': exports.MO.WIDEREL,
        '\u27B6': exports.MO.REL,
        '\u27B7': exports.MO.REL,
        '\u27B8': exports.MO.WIDEREL,
        '\u27B9': exports.MO.REL,
        '\u27BA': exports.MO.WIDEREL,
        '\u27BB': exports.MO.WIDEREL,
        '\u27BC': exports.MO.WIDEREL,
        '\u27BD': exports.MO.WIDEREL,
        '\u27BE': exports.MO.WIDEREL,
        '\u27C2': exports.MO.REL,
        '\u27C2\u0338': exports.MO.REL,
        '\u27CB': exports.MO.BIN3,
        '\u27CD': exports.MO.BIN3,
        '\u27F0': exports.MO.RELSTRETCH,
        '\u27F1': exports.MO.RELSTRETCH,
        '\u27F2': exports.MO.REL,
        '\u27F3': exports.MO.REL,
        '\u27F4': exports.MO.RELSTRETCH,
        '\u27F5': exports.MO.WIDEREL,
        '\u27F6': exports.MO.WIDEREL,
        '\u27F7': exports.MO.WIDEREL,
        '\u27F8': exports.MO.WIDEREL,
        '\u27F9': exports.MO.WIDEREL,
        '\u27FA': exports.MO.WIDEREL,
        '\u27FB': exports.MO.WIDEREL,
        '\u27FC': exports.MO.WIDEREL,
        '\u27FD': exports.MO.WIDEREL,
        '\u27FE': exports.MO.WIDEREL,
        '\u27FF': exports.MO.WIDEREL,
        '\u2900': exports.MO.WIDEREL,
        '\u2901': exports.MO.WIDEREL,
        '\u2902': exports.MO.WIDEREL,
        '\u2903': exports.MO.WIDEREL,
        '\u2904': exports.MO.WIDEREL,
        '\u2905': exports.MO.WIDEREL,
        '\u2906': exports.MO.WIDEREL,
        '\u2907': exports.MO.WIDEREL,
        '\u2908': exports.MO.RELSTRETCH,
        '\u2909': exports.MO.RELSTRETCH,
        '\u290A': exports.MO.RELSTRETCH,
        '\u290B': exports.MO.RELSTRETCH,
        '\u290C': exports.MO.WIDEREL,
        '\u290D': exports.MO.WIDEREL,
        '\u290E': exports.MO.WIDEREL,
        '\u290F': exports.MO.WIDEREL,
        '\u2910': exports.MO.WIDEREL,
        '\u2911': exports.MO.WIDEREL,
        '\u2912': exports.MO.RELSTRETCH,
        '\u2913': exports.MO.RELSTRETCH,
        '\u2914': exports.MO.WIDEREL,
        '\u2915': exports.MO.WIDEREL,
        '\u2916': exports.MO.WIDEREL,
        '\u2917': exports.MO.WIDEREL,
        '\u2918': exports.MO.WIDEREL,
        '\u2919': exports.MO.WIDEREL,
        '\u291A': exports.MO.WIDEREL,
        '\u291B': exports.MO.WIDEREL,
        '\u291C': exports.MO.WIDEREL,
        '\u291D': exports.MO.WIDEREL,
        '\u291E': exports.MO.WIDEREL,
        '\u291F': exports.MO.WIDEREL,
        '\u2920': exports.MO.WIDEREL,
        '\u2921': exports.MO.REL,
        '\u2922': exports.MO.REL,
        '\u2923': exports.MO.REL,
        '\u2924': exports.MO.REL,
        '\u2925': exports.MO.REL,
        '\u2926': exports.MO.REL,
        '\u2927': exports.MO.REL,
        '\u2928': exports.MO.REL,
        '\u2929': exports.MO.REL,
        '\u292A': exports.MO.REL,
        '\u292B': exports.MO.REL,
        '\u292C': exports.MO.REL,
        '\u292D': exports.MO.REL,
        '\u292E': exports.MO.REL,
        '\u292F': exports.MO.REL,
        '\u2930': exports.MO.REL,
        '\u2931': exports.MO.REL,
        '\u2932': exports.MO.REL,
        '\u2933': exports.MO.RELACCENT,
        '\u2934': exports.MO.RELSTRETCH,
        '\u2935': exports.MO.RELSTRETCH,
        '\u2936': exports.MO.RELSTRETCH,
        '\u2937': exports.MO.RELSTRETCH,
        '\u2938': exports.MO.REL,
        '\u2939': exports.MO.REL,
        '\u293A': exports.MO.RELACCENT,
        '\u293B': exports.MO.RELACCENT,
        '\u293C': exports.MO.RELACCENT,
        '\u293D': exports.MO.RELACCENT,
        '\u293E': exports.MO.REL,
        '\u293F': exports.MO.REL,
        '\u2940': exports.MO.REL,
        '\u2941': exports.MO.REL,
        '\u2942': exports.MO.WIDEREL,
        '\u2943': exports.MO.WIDEREL,
        '\u2944': exports.MO.WIDEREL,
        '\u2945': exports.MO.RELSTRETCH,
        '\u2946': exports.MO.RELSTRETCH,
        '\u2947': exports.MO.WIDEREL,
        '\u2948': exports.MO.WIDEREL,
        '\u2949': exports.MO.RELSTRETCH,
        '\u294A': exports.MO.WIDEREL,
        '\u294B': exports.MO.WIDEREL,
        '\u294C': exports.MO.RELSTRETCH,
        '\u294D': exports.MO.RELSTRETCH,
        '\u294E': exports.MO.WIDEREL,
        '\u294F': exports.MO.RELSTRETCH,
        '\u2950': exports.MO.WIDEREL,
        '\u2951': exports.MO.RELSTRETCH,
        '\u2952': exports.MO.WIDEREL,
        '\u2953': exports.MO.WIDEREL,
        '\u2954': exports.MO.RELSTRETCH,
        '\u2955': exports.MO.RELSTRETCH,
        '\u2956': exports.MO.WIDEREL,
        '\u2957': exports.MO.WIDEREL,
        '\u2958': exports.MO.RELSTRETCH,
        '\u2959': exports.MO.RELSTRETCH,
        '\u295A': exports.MO.WIDEREL,
        '\u295B': exports.MO.WIDEREL,
        '\u295C': exports.MO.RELSTRETCH,
        '\u295D': exports.MO.RELSTRETCH,
        '\u295E': exports.MO.WIDEREL,
        '\u295F': exports.MO.WIDEREL,
        '\u2960': exports.MO.RELSTRETCH,
        '\u2961': exports.MO.RELSTRETCH,
        '\u2962': exports.MO.WIDEREL,
        '\u2963': exports.MO.RELSTRETCH,
        '\u2964': exports.MO.WIDEREL,
        '\u2965': exports.MO.RELSTRETCH,
        '\u2966': exports.MO.WIDEREL,
        '\u2967': exports.MO.WIDEREL,
        '\u2968': exports.MO.WIDEREL,
        '\u2969': exports.MO.WIDEREL,
        '\u296A': exports.MO.WIDEREL,
        '\u296B': exports.MO.WIDEREL,
        '\u296C': exports.MO.WIDEREL,
        '\u296D': exports.MO.WIDEREL,
        '\u296E': exports.MO.RELSTRETCH,
        '\u296F': exports.MO.RELSTRETCH,
        '\u2970': exports.MO.WIDEREL,
        '\u2971': exports.MO.WIDEREL,
        '\u2972': exports.MO.WIDEREL,
        '\u2973': exports.MO.WIDEREL,
        '\u2974': exports.MO.WIDEREL,
        '\u2975': exports.MO.WIDEREL,
        '\u2976': exports.MO.RELACCENT,
        '\u2977': exports.MO.RELACCENT,
        '\u2978': exports.MO.RELACCENT,
        '\u2979': exports.MO.RELACCENT,
        '\u297A': exports.MO.RELACCENT,
        '\u297B': exports.MO.RELACCENT,
        '\u297C': exports.MO.WIDEREL,
        '\u297D': exports.MO.WIDEREL,
        '\u297E': exports.MO.RELSTRETCH,
        '\u297F': exports.MO.RELSTRETCH,
        '\u2981': exports.MO.REL,
        '\u2982': exports.MO.REL,
        '\u29B6': exports.MO.REL,
        '\u29B7': exports.MO.REL,
        '\u29B8': exports.MO.BIN4,
        '\u29B9': exports.MO.REL,
        '\u29BC': exports.MO.BIN4,
        '\u29C0': exports.MO.REL,
        '\u29C1': exports.MO.REL,
        '\u29C4': exports.MO.BIN4,
        '\u29C5': exports.MO.BIN4,
        '\u29C6': exports.MO.BIN3,
        '\u29C7': exports.MO.BIN3,
        '\u29C8': exports.MO.BIN3,
        '\u29CE': exports.MO.REL,
        '\u29CF': exports.MO.REL,
        '\u29D0': exports.MO.REL,
        '\u29D1': exports.MO.REL,
        '\u29D2': exports.MO.REL,
        '\u29D3': exports.MO.REL,
        '\u29D4': exports.MO.BIN3,
        '\u29D5': exports.MO.BIN3,
        '\u29D6': exports.MO.BIN3,
        '\u29D7': exports.MO.BIN3,
        '\u29DF': exports.MO.REL,
        '\u29E1': exports.MO.REL,
        '\u29E2': exports.MO.BIN3,
        '\u29E3': exports.MO.REL,
        '\u29E4': exports.MO.REL,
        '\u29E5': exports.MO.REL,
        '\u29E6': exports.MO.REL,
        '\u29F4': exports.MO.REL,
        '\u29F5': exports.MO.BIN4,
        '\u29F6': exports.MO.BIN4,
        '\u29F7': exports.MO.BIN4,
        '\u29F8': exports.MO.BIN4,
        '\u29F9': exports.MO.BIN4,
        '\u29FA': exports.MO.BIN4,
        '\u29FB': exports.MO.BIN4,
        '\u2A1D': exports.MO.BIN3,
        '\u2A1E': exports.MO.BIN3,
        '\u2A1F': exports.MO.BIN4,
        '\u2A20': exports.MO.BIN4,
        '\u2A21': exports.MO.BIN4,
        '\u2A22': exports.MO.BIN4,
        '\u2A23': exports.MO.BIN4,
        '\u2A24': exports.MO.BIN4,
        '\u2A25': exports.MO.BIN4,
        '\u2A26': exports.MO.BIN4,
        '\u2A27': exports.MO.BIN4,
        '\u2A28': exports.MO.BIN4,
        '\u2A29': exports.MO.BIN4,
        '\u2A2A': exports.MO.BIN4,
        '\u2A2B': exports.MO.BIN4,
        '\u2A2C': exports.MO.BIN4,
        '\u2A2D': exports.MO.BIN4,
        '\u2A2E': exports.MO.BIN4,
        '\u2A2F': exports.MO.BIN3,
        '\u2A30': exports.MO.BIN3,
        '\u2A31': exports.MO.BIN3,
        '\u2A32': exports.MO.BIN3,
        '\u2A33': exports.MO.BIN3,
        '\u2A34': exports.MO.BIN3,
        '\u2A35': exports.MO.BIN3,
        '\u2A36': exports.MO.BIN3,
        '\u2A37': exports.MO.BIN3,
        '\u2A38': exports.MO.BIN4,
        '\u2A39': exports.MO.BIN4,
        '\u2A3A': exports.MO.BIN4,
        '\u2A3B': exports.MO.BIN3,
        '\u2A3C': exports.MO.BIN3,
        '\u2A3D': exports.MO.BIN3,
        '\u2A3E': exports.MO.BIN4,
        '\u2A3F': exports.MO.BIN3,
        '\u2A40': exports.MO.BIN4,
        '\u2A41': exports.MO.BIN4,
        '\u2A42': exports.MO.BIN4,
        '\u2A43': exports.MO.BIN4,
        '\u2A44': exports.MO.BIN4,
        '\u2A45': exports.MO.BIN4,
        '\u2A46': exports.MO.BIN4,
        '\u2A47': exports.MO.BIN4,
        '\u2A48': exports.MO.BIN4,
        '\u2A49': exports.MO.BIN4,
        '\u2A4A': exports.MO.BIN4,
        '\u2A4B': exports.MO.BIN4,
        '\u2A4C': exports.MO.BIN4,
        '\u2A4D': exports.MO.BIN4,
        '\u2A4E': exports.MO.BIN4,
        '\u2A4F': exports.MO.BIN4,
        '\u2A50': exports.MO.BIN3,
        '\u2A51': exports.MO.BIN4,
        '\u2A52': exports.MO.BIN4,
        '\u2A53': exports.MO.BIN4,
        '\u2A54': exports.MO.BIN4,
        '\u2A55': exports.MO.BIN4,
        '\u2A56': exports.MO.BIN4,
        '\u2A57': exports.MO.BIN4,
        '\u2A58': exports.MO.BIN4,
        '\u2A59': exports.MO.BIN4,
        '\u2A5A': exports.MO.BIN4,
        '\u2A5B': exports.MO.BIN4,
        '\u2A5C': exports.MO.BIN4,
        '\u2A5D': exports.MO.BIN4,
        '\u2A5E': exports.MO.BIN4,
        '\u2A5F': exports.MO.BIN4,
        '\u2A60': exports.MO.BIN4,
        '\u2A61': exports.MO.BIN4,
        '\u2A62': exports.MO.BIN4,
        '\u2A63': exports.MO.BIN4,
        '\u2A64': exports.MO.BIN3,
        '\u2A65': exports.MO.BIN3,
        '\u2A66': exports.MO.REL,
        '\u2A67': exports.MO.REL,
        '\u2A68': exports.MO.REL,
        '\u2A69': exports.MO.REL,
        '\u2A6A': exports.MO.REL,
        '\u2A6B': exports.MO.REL,
        '\u2A6C': exports.MO.REL,
        '\u2A6D': exports.MO.REL,
        '\u2A6E': exports.MO.REL,
        '\u2A6F': exports.MO.REL,
        '\u2A70': exports.MO.REL,
        '\u2A71': exports.MO.REL,
        '\u2A72': exports.MO.REL,
        '\u2A73': exports.MO.REL,
        '\u2A74': exports.MO.REL,
        '\u2A75': exports.MO.REL,
        '\u2A76': exports.MO.REL,
        '\u2A77': exports.MO.REL,
        '\u2A78': exports.MO.REL,
        '\u2A79': exports.MO.REL,
        '\u2A7A': exports.MO.REL,
        '\u2A7B': exports.MO.REL,
        '\u2A7C': exports.MO.REL,
        '\u2A7D': exports.MO.REL,
        '\u2A7D\u0338': exports.MO.REL,
        '\u2A7E': exports.MO.REL,
        '\u2A7E\u0338': exports.MO.REL,
        '\u2A7F': exports.MO.REL,
        '\u2A80': exports.MO.REL,
        '\u2A81': exports.MO.REL,
        '\u2A82': exports.MO.REL,
        '\u2A83': exports.MO.REL,
        '\u2A84': exports.MO.REL,
        '\u2A85': exports.MO.REL,
        '\u2A86': exports.MO.REL,
        '\u2A87': exports.MO.REL,
        '\u2A88': exports.MO.REL,
        '\u2A89': exports.MO.REL,
        '\u2A8A': exports.MO.REL,
        '\u2A8B': exports.MO.REL,
        '\u2A8C': exports.MO.REL,
        '\u2A8D': exports.MO.REL,
        '\u2A8E': exports.MO.REL,
        '\u2A8F': exports.MO.REL,
        '\u2A90': exports.MO.REL,
        '\u2A91': exports.MO.REL,
        '\u2A92': exports.MO.REL,
        '\u2A93': exports.MO.REL,
        '\u2A94': exports.MO.REL,
        '\u2A95': exports.MO.REL,
        '\u2A96': exports.MO.REL,
        '\u2A97': exports.MO.REL,
        '\u2A98': exports.MO.REL,
        '\u2A99': exports.MO.REL,
        '\u2A9A': exports.MO.REL,
        '\u2A9B': exports.MO.REL,
        '\u2A9C': exports.MO.REL,
        '\u2A9D': exports.MO.REL,
        '\u2A9E': exports.MO.REL,
        '\u2A9F': exports.MO.REL,
        '\u2AA0': exports.MO.REL,
        '\u2AA1': exports.MO.REL,
        '\u2AA2': exports.MO.REL,
        '\u2AA3': exports.MO.REL,
        '\u2AA4': exports.MO.REL,
        '\u2AA5': exports.MO.REL,
        '\u2AA6': exports.MO.REL,
        '\u2AA7': exports.MO.REL,
        '\u2AA8': exports.MO.REL,
        '\u2AA9': exports.MO.REL,
        '\u2AAA': exports.MO.REL,
        '\u2AAB': exports.MO.REL,
        '\u2AAC': exports.MO.REL,
        '\u2AAD': exports.MO.REL,
        '\u2AAE': exports.MO.REL,
        '\u2AAF': exports.MO.REL,
        '\u2AAF\u0338': exports.MO.REL,
        '\u2AB0': exports.MO.REL,
        '\u2AB0\u0338': exports.MO.REL,
        '\u2AB1': exports.MO.REL,
        '\u2AB2': exports.MO.REL,
        '\u2AB3': exports.MO.REL,
        '\u2AB4': exports.MO.REL,
        '\u2AB5': exports.MO.REL,
        '\u2AB6': exports.MO.REL,
        '\u2AB7': exports.MO.REL,
        '\u2AB8': exports.MO.REL,
        '\u2AB9': exports.MO.REL,
        '\u2ABA': exports.MO.REL,
        '\u2ABB': exports.MO.REL,
        '\u2ABC': exports.MO.REL,
        '\u2ABD': exports.MO.REL,
        '\u2ABE': exports.MO.REL,
        '\u2ABF': exports.MO.REL,
        '\u2AC0': exports.MO.REL,
        '\u2AC1': exports.MO.REL,
        '\u2AC2': exports.MO.REL,
        '\u2AC3': exports.MO.REL,
        '\u2AC4': exports.MO.REL,
        '\u2AC5': exports.MO.REL,
        '\u2AC6': exports.MO.REL,
        '\u2AC7': exports.MO.REL,
        '\u2AC8': exports.MO.REL,
        '\u2AC9': exports.MO.REL,
        '\u2ACA': exports.MO.REL,
        '\u2ACB': exports.MO.REL,
        '\u2ACC': exports.MO.REL,
        '\u2ACD': exports.MO.REL,
        '\u2ACE': exports.MO.REL,
        '\u2ACF': exports.MO.REL,
        '\u2AD0': exports.MO.REL,
        '\u2AD1': exports.MO.REL,
        '\u2AD2': exports.MO.REL,
        '\u2AD3': exports.MO.REL,
        '\u2AD4': exports.MO.REL,
        '\u2AD5': exports.MO.REL,
        '\u2AD6': exports.MO.REL,
        '\u2AD7': exports.MO.REL,
        '\u2AD8': exports.MO.REL,
        '\u2AD9': exports.MO.REL,
        '\u2ADA': exports.MO.REL,
        '\u2ADB': exports.MO.BIN4,
        '\u2ADD': exports.MO.BIN3,
        '\u2ADD\u0338': exports.MO.REL,
        '\u2ADE': exports.MO.REL,
        '\u2ADF': exports.MO.REL,
        '\u2AE0': exports.MO.REL,
        '\u2AE1': exports.MO.REL,
        '\u2AE2': exports.MO.REL,
        '\u2AE3': exports.MO.REL,
        '\u2AE4': exports.MO.REL,
        '\u2AE5': exports.MO.REL,
        '\u2AE6': exports.MO.REL,
        '\u2AE7': exports.MO.REL,
        '\u2AE8': exports.MO.REL,
        '\u2AE9': exports.MO.REL,
        '\u2AEA': exports.MO.REL,
        '\u2AEB': exports.MO.REL,
        '\u2AEE': exports.MO.REL,
        '\u2AF2': exports.MO.REL,
        '\u2AF3': exports.MO.REL,
        '\u2AF4': exports.MO.REL,
        '\u2AF5': exports.MO.REL,
        '\u2AF6': exports.MO.BIN4,
        '\u2AF7': exports.MO.REL,
        '\u2AF8': exports.MO.REL,
        '\u2AF9': exports.MO.REL,
        '\u2AFA': exports.MO.REL,
        '\u2AFB': exports.MO.BIN4,
        '\u2AFD': exports.MO.BIN4,
        '\u2AFE': exports.MO.BIN3,
        '\u2B00': exports.MO.REL,
        '\u2B01': exports.MO.REL,
        '\u2B02': exports.MO.REL,
        '\u2B03': exports.MO.REL,
        '\u2B04': exports.MO.WIDEREL,
        '\u2B05': exports.MO.WIDEREL,
        '\u2B06': exports.MO.RELSTRETCH,
        '\u2B07': exports.MO.RELSTRETCH,
        '\u2B08': exports.MO.REL,
        '\u2B09': exports.MO.REL,
        '\u2B0A': exports.MO.REL,
        '\u2B0B': exports.MO.REL,
        '\u2B0C': exports.MO.WIDEREL,
        '\u2B0D': exports.MO.RELSTRETCH,
        '\u2B0E': exports.MO.RELSTRETCH,
        '\u2B0F': exports.MO.RELSTRETCH,
        '\u2B10': exports.MO.RELSTRETCH,
        '\u2B11': exports.MO.RELSTRETCH,
        '\u2B30': exports.MO.WIDEREL,
        '\u2B31': exports.MO.WIDEREL,
        '\u2B32': exports.MO.RELSTRETCH,
        '\u2B33': exports.MO.WIDEREL,
        '\u2B34': exports.MO.WIDEREL,
        '\u2B35': exports.MO.WIDEREL,
        '\u2B36': exports.MO.WIDEREL,
        '\u2B37': exports.MO.WIDEREL,
        '\u2B38': exports.MO.WIDEREL,
        '\u2B39': exports.MO.WIDEREL,
        '\u2B3A': exports.MO.WIDEREL,
        '\u2B3B': exports.MO.WIDEREL,
        '\u2B3C': exports.MO.WIDEREL,
        '\u2B3D': exports.MO.WIDEREL,
        '\u2B3E': exports.MO.WIDEREL,
        '\u2B3F': exports.MO.RELACCENT,
        '\u2B40': exports.MO.WIDEREL,
        '\u2B41': exports.MO.WIDEREL,
        '\u2B42': exports.MO.WIDEREL,
        '\u2B43': exports.MO.WIDEREL,
        '\u2B44': exports.MO.WIDEREL,
        '\u2B45': exports.MO.WIDEREL,
        '\u2B46': exports.MO.WIDEREL,
        '\u2B47': exports.MO.WIDEREL,
        '\u2B48': exports.MO.WIDEREL,
        '\u2B49': exports.MO.WIDEREL,
        '\u2B4A': exports.MO.WIDEREL,
        '\u2B4B': exports.MO.WIDEREL,
        '\u2B4C': exports.MO.WIDEREL,
        '\u2B4D': exports.MO.REL,
        '\u2B4E': exports.MO.REL,
        '\u2B4F': exports.MO.REL,
        '\u2B5A': exports.MO.REL,
        '\u2B5B': exports.MO.REL,
        '\u2B5C': exports.MO.REL,
        '\u2B5D': exports.MO.REL,
        '\u2B5E': exports.MO.REL,
        '\u2B5F': exports.MO.REL,
        '\u2B60': exports.MO.WIDEREL,
        '\u2B61': exports.MO.RELSTRETCH,
        '\u2B62': exports.MO.WIDEREL,
        '\u2B63': exports.MO.RELSTRETCH,
        '\u2B64': exports.MO.WIDEREL,
        '\u2B65': exports.MO.RELSTRETCH,
        '\u2B66': exports.MO.REL,
        '\u2B67': exports.MO.REL,
        '\u2B68': exports.MO.REL,
        '\u2B69': exports.MO.REL,
        '\u2B6A': exports.MO.WIDEREL,
        '\u2B6B': exports.MO.RELSTRETCH,
        '\u2B6C': exports.MO.WIDEREL,
        '\u2B6D': exports.MO.RELSTRETCH,
        '\u2B6E': exports.MO.REL,
        '\u2B6F': exports.MO.REL,
        '\u2B70': exports.MO.WIDEREL,
        '\u2B71': exports.MO.RELSTRETCH,
        '\u2B72': exports.MO.WIDEREL,
        '\u2B73': exports.MO.RELSTRETCH,
        '\u2B76': exports.MO.REL,
        '\u2B77': exports.MO.REL,
        '\u2B78': exports.MO.REL,
        '\u2B79': exports.MO.REL,
        '\u2B7A': exports.MO.WIDEREL,
        '\u2B7B': exports.MO.RELSTRETCH,
        '\u2B7C': exports.MO.WIDEREL,
        '\u2B7D': exports.MO.RELSTRETCH,
        '\u2B80': exports.MO.WIDEREL,
        '\u2B81': exports.MO.RELSTRETCH,
        '\u2B82': exports.MO.WIDEREL,
        '\u2B83': exports.MO.RELSTRETCH,
        '\u2B84': exports.MO.WIDEREL,
        '\u2B85': exports.MO.RELSTRETCH,
        '\u2B86': exports.MO.WIDEREL,
        '\u2B87': exports.MO.RELSTRETCH,
        '\u2B88': exports.MO.RELACCENT,
        '\u2B89': exports.MO.REL,
        '\u2B8A': exports.MO.RELACCENT,
        '\u2B8B': exports.MO.REL,
        '\u2B8C': exports.MO.REL,
        '\u2B8D': exports.MO.REL,
        '\u2B8E': exports.MO.REL,
        '\u2B8F': exports.MO.REL,
        '\u2B94': exports.MO.REL,
        '\u2B95': exports.MO.WIDEREL,
        '\u2BA0': exports.MO.RELSTRETCH,
        '\u2BA1': exports.MO.RELSTRETCH,
        '\u2BA2': exports.MO.RELSTRETCH,
        '\u2BA3': exports.MO.RELSTRETCH,
        '\u2BA4': exports.MO.RELSTRETCH,
        '\u2BA5': exports.MO.RELSTRETCH,
        '\u2BA6': exports.MO.RELSTRETCH,
        '\u2BA7': exports.MO.RELSTRETCH,
        '\u2BA8': exports.MO.WIDEREL,
        '\u2BA9': exports.MO.WIDEREL,
        '\u2BAA': exports.MO.WIDEREL,
        '\u2BAB': exports.MO.WIDEREL,
        '\u2BAC': exports.MO.RELSTRETCH,
        '\u2BAD': exports.MO.RELSTRETCH,
        '\u2BAE': exports.MO.RELSTRETCH,
        '\u2BAF': exports.MO.RELSTRETCH,
        '\u2BB0': exports.MO.REL,
        '\u2BB1': exports.MO.REL,
        '\u2BB2': exports.MO.REL,
        '\u2BB3': exports.MO.REL,
        '\u2BB4': exports.MO.REL,
        '\u2BB5': exports.MO.REL,
        '\u2BB6': exports.MO.REL,
        '\u2BB7': exports.MO.REL,
        '\u2BB8': exports.MO.RELSTRETCH,
        '\u2BD1': exports.MO.REL,
        '\u3ADC': exports.MO.BIN3,
        '\uFE37': exports.MO.WIDEACCENT,
        '\uFE38': exports.MO.WIDEACCENT,
    },
};
