// @mathjax/src/mjs/util/context.js
export const hasWindow = typeof window !== 'undefined';
export const context = {
    window: hasWindow ? window : null,
    document: hasWindow ? window.document : null,
    os: (() => {
        if (hasWindow && window.navigator) {
            const app = window.navigator.appVersion;
            const osNames = [
                ['Win', 'Windows'],
                ['Mac', 'MacOS'],
                ['X11', 'Unix'],
                ['Linux', 'Unix'],
            ];
            for (const [key, os] of osNames) {
                if (app.includes(key)) {
                    return os;
                }
            }
            if (window.navigator.userAgent.includes('Android')) {
                return 'Unix';
            }
        }
        else if (typeof process !== 'undefined') {
            return ({
                linux: 'Unix',
                android: 'Unix',
                aix: 'Unix',
                freebsd: 'Unix',
                netbsd: 'Unix',
                openbsd: 'Unix',
                sunos: 'Unix',
                darwin: 'MacOS',
                win32: 'Windows',
                cygwin: 'Windows',
            }[process.platform] || process.platform);
        }
        return 'unknown';
    })(),
    path: (file) => file,
};
if (context.os === 'Windows') {
    context.path = (file) => file.match(/^[/\\]?[a-zA-Z]:[/\\]/)
        ? 'file://' + file.replace(/\\/g, '/').replace(/^\//, '')
        : file.replace(/^\//, 'file:///');
}
