// speech-rule-engine/cjs/speech_rules/clearspeak_util.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.nodeCounter = nodeCounter;
exports.allCellsSimple = allCellsSimple;
exports.isSmallVulgarFraction = isSmallVulgarFraction;
exports.ordinalExponent = ordinalExponent;
exports.nestingDepth = nestingDepth;
exports.matchingFences = matchingFences;
exports.fencedArguments = fencedArguments;
exports.simpleArguments = simpleArguments;
exports.wordOrdinal = wordOrdinal;
exports.isUnit = isUnit;
const span_js_1 = require("../audio/span.js");
const DomUtil = __importStar(require("../common/dom_util.js"));
const XpathUtil = __importStar(require("../common/xpath_util.js"));
const locale_js_1 = require("../l10n/locale.js");
const transformers_js_1 = require("../l10n/transformers.js");
const grammar_js_1 = require("../rule_engine/grammar.js");
const StoreUtil = __importStar(require("../rule_engine/store_util.js"));
const semantic_attr_js_1 = require("../semantic_tree/semantic_attr.js");
const semantic_meaning_js_1 = require("../semantic_tree/semantic_meaning.js");
const math_compound_store_js_1 = require("../rule_engine/math_compound_store.js");
function nodeCounter(nodes, context) {
    const split = context.split('-');
    const func = StoreUtil.nodeCounter(nodes, split[0] || '');
    const sep = split[1] || '';
    const init = split[2] || '';
    let first = true;
    return function () {
        const result = func();
        if (first) {
            first = false;
            return init + result + sep;
        }
        else {
            return result + sep;
        }
    };
}
function simpleNode(node) {
    if (!node.hasAttribute('annotation')) {
        return false;
    }
    const annotation = node.getAttribute('annotation');
    return !!/clearspeak:simple$|clearspeak:simple;/.exec(annotation);
}
function simpleCell_(node) {
    if (simpleNode(node)) {
        return true;
    }
    if (node.tagName !== semantic_meaning_js_1.SemanticType.SUBSCRIPT) {
        return false;
    }
    const children = node.childNodes[0].childNodes;
    const index = children[1];
    return (children[0].tagName === semantic_meaning_js_1.SemanticType.IDENTIFIER &&
        (isInteger_(index) ||
            (index.tagName === semantic_meaning_js_1.SemanticType.INFIXOP &&
                index.hasAttribute('role') &&
                index.getAttribute('role') === semantic_meaning_js_1.SemanticRole.IMPLICIT &&
                allIndices_(index))));
}
function isInteger_(node) {
    return (node.tagName === semantic_meaning_js_1.SemanticType.NUMBER &&
        node.hasAttribute('role') &&
        node.getAttribute('role') === semantic_meaning_js_1.SemanticRole.INTEGER);
}
function allIndices_(node) {
    const nodes = XpathUtil.evalXPath('children/*', node);
    return nodes.every((x) => isInteger_(x) || x.tagName === semantic_meaning_js_1.SemanticType.IDENTIFIER);
}
function allCellsSimple(node) {
    const xpath = node.tagName === semantic_meaning_js_1.SemanticType.MATRIX
        ? 'children/row/children/cell/children/*'
        : 'children/line/children/*';
    const nodes = XpathUtil.evalXPath(xpath, node);
    const result = nodes.every(simpleCell_);
    return result ? [node] : [];
}
function isSmallVulgarFraction(node) {
    return (0, transformers_js_1.vulgarFractionSmall)(node, 20, 11) ? [node] : [];
}
function ordinalExponent(node) {
    const num = parseInt(node.textContent, 10);
    return [
        span_js_1.Span.stringEmpty(isNaN(num)
            ? node.textContent
            : num > 10
                ? locale_js_1.LOCALE.NUMBERS.numericOrdinal(num)
                : locale_js_1.LOCALE.NUMBERS.wordOrdinal(num))
    ];
}
let NESTING_DEPTH = null;
function nestingDepth(node) {
    let count = 0;
    const fence = node.textContent;
    const index = node.getAttribute('role') === 'open' ? 0 : 1;
    let parent = node.parentNode;
    while (parent) {
        if (parent.tagName === semantic_meaning_js_1.SemanticType.FENCED &&
            parent.childNodes[0].childNodes[index].textContent === fence) {
            count++;
        }
        parent = parent.parentNode;
    }
    NESTING_DEPTH = count > 1 ? locale_js_1.LOCALE.NUMBERS.wordOrdinal(count) : '';
    return [span_js_1.Span.stringEmpty(NESTING_DEPTH)];
}
function matchingFences(node) {
    const sibling = node.previousSibling;
    let left, right;
    if (sibling) {
        left = sibling;
        right = node;
    }
    else {
        left = node;
        right = node.nextSibling;
    }
    if (!right) {
        return [];
    }
    return (0, semantic_attr_js_1.isMatchingFence)(left.textContent, right.textContent) ? [node] : [];
}
function insertNesting(text, correction) {
    if (!correction || !text) {
        return text;
    }
    const start = text.match(/^(open|close) /);
    if (!start) {
        return correction + ' ' + text;
    }
    return start[0] + correction + ' ' + text.substring(start[0].length);
}
grammar_js_1.Grammar.getInstance().setCorrection('insertNesting', insertNesting);
function fencedArguments(node) {
    const content = DomUtil.toArray(node.parentNode.childNodes);
    const children = XpathUtil.evalXPath('../../children/*', node);
    const index = content.indexOf(node);
    return fencedFactor_(children[index]) || fencedFactor_(children[index + 1])
        ? [node]
        : [];
}
function simpleArguments(node) {
    const content = DomUtil.toArray(node.parentNode.childNodes);
    const children = XpathUtil.evalXPath('../../children/*', node);
    const index = content.indexOf(node);
    return simpleFactor_(children[index]) &&
        children[index + 1] &&
        (simpleFactor_(children[index + 1]) ||
            children[index + 1].tagName === semantic_meaning_js_1.SemanticType.ROOT ||
            children[index + 1].tagName === semantic_meaning_js_1.SemanticType.SQRT ||
            (children[index + 1].tagName === semantic_meaning_js_1.SemanticType.SUPERSCRIPT &&
                children[index + 1].childNodes[0].childNodes[0] &&
                (children[index + 1].childNodes[0].childNodes[0]
                    .tagName === semantic_meaning_js_1.SemanticType.NUMBER ||
                    children[index + 1].childNodes[0].childNodes[0]
                        .tagName === semantic_meaning_js_1.SemanticType.IDENTIFIER) &&
                (children[index + 1].childNodes[0].childNodes[1].textContent === '2' ||
                    children[index + 1].childNodes[0].childNodes[1].textContent === '3')))
        ? [node]
        : [];
}
function simpleFactor_(node) {
    return (!!node &&
        (node.tagName === semantic_meaning_js_1.SemanticType.NUMBER ||
            node.tagName === semantic_meaning_js_1.SemanticType.IDENTIFIER ||
            node.tagName === semantic_meaning_js_1.SemanticType.FUNCTION ||
            node.tagName === semantic_meaning_js_1.SemanticType.APPL ||
            node.tagName === semantic_meaning_js_1.SemanticType.FRACTION));
}
function fencedFactor_(node) {
    return (node &&
        (node.tagName === semantic_meaning_js_1.SemanticType.FENCED ||
            (node.hasAttribute('role') &&
                node.getAttribute('role') === semantic_meaning_js_1.SemanticRole.LEFTRIGHT) ||
            layoutFactor_(node)));
}
function layoutFactor_(node) {
    return (!!node &&
        (node.tagName === semantic_meaning_js_1.SemanticType.MATRIX ||
            node.tagName === semantic_meaning_js_1.SemanticType.VECTOR));
}
function wordOrdinal(node) {
    return [
        span_js_1.Span.stringEmpty(locale_js_1.LOCALE.NUMBERS.wordOrdinal(parseInt(node.textContent, 10)))
    ];
}
function isUnit(node) {
    return (0, math_compound_store_js_1.lookupCategory)(node.textContent + ':unit') ? [node] : [];
}
